{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "import GEOparse\n",
    "import sys\n",
    "import pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21-Dec-2017 12:32:04 INFO GEOparse - Parsing _data/genomic_data/melanoma/GSE8401_family.soft.gz: \n",
      "21-Dec-2017 12:32:04 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "21-Dec-2017 12:32:04 DEBUG GEOparse - SERIES: GSE8401\n",
      "21-Dec-2017 12:32:05 DEBUG GEOparse - PLATFORM: GPL96\n",
      "/usr/local/lib/python3.5/dist-packages/GEOparse/GEOparse.py:330: FutureWarning: from_csv is deprecated. Please use read_csv(...) instead. Note that some of the default arguments are different, so please refer to the documentation for from_csv when changing your function calls\n",
      "  return DataFrame.from_csv(StringIO(data), index_col=None, sep=\"\\t\")\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207929\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207930\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207931\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207932\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207933\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207934\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207935\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207936\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207937\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207938\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207939\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207940\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207941\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207942\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207943\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207944\n",
      "21-Dec-2017 12:32:06 DEBUG GEOparse - SAMPLE: GSM207945\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207946\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207947\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207948\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207949\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207950\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207951\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207952\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207953\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207954\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207955\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207956\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207957\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207958\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207959\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207960\n",
      "21-Dec-2017 12:32:07 DEBUG GEOparse - SAMPLE: GSM207961\n",
      "21-Dec-2017 12:32:08 DEBUG GEOparse - SAMPLE: GSM207962\n",
      "21-Dec-2017 12:32:08 DEBUG GEOparse - SAMPLE: GSM207963\n",
      "21-Dec-2017 12:32:08 DEBUG GEOparse - SAMPLE: GSM207964\n",
      "21-Dec-2017 12:32:08 DEBUG GEOparse - SAMPLE: GSM207965\n",
      "21-Dec-2017 12:32:08 DEBUG GEOparse - SAMPLE: GSM207966\n",
      "21-Dec-2017 12:32:08 DEBUG GEOparse - SAMPLE: GSM207967\n",
      "21-Dec-2017 12:32:08 DEBUG GEOparse - SAMPLE: GSM207968\n",
      "21-Dec-2017 12:32:08 DEBUG GEOparse - SAMPLE: GSM207969\n",
      "21-Dec-2017 12:32:08 DEBUG GEOparse - SAMPLE: GSM207970\n",
      "21-Dec-2017 12:32:08 DEBUG GEOparse - SAMPLE: GSM207971\n",
      "21-Dec-2017 12:32:08 DEBUG GEOparse - SAMPLE: GSM207972\n",
      "21-Dec-2017 12:32:08 DEBUG GEOparse - SAMPLE: GSM207973\n",
      "21-Dec-2017 12:32:08 DEBUG GEOparse - SAMPLE: GSM207974\n",
      "21-Dec-2017 12:32:08 DEBUG GEOparse - SAMPLE: GSM207975\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207976\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207977\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207978\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207979\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207980\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207981\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207982\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207983\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207984\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207985\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207986\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207987\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207988\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207989\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207990\n",
      "21-Dec-2017 12:32:09 DEBUG GEOparse - SAMPLE: GSM207991\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM207992\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM207993\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM207994\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM207995\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM207996\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM207997\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM207998\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM207999\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM208000\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM208001\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM208002\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM208003\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM208004\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM208005\n",
      "21-Dec-2017 12:32:10 DEBUG GEOparse - SAMPLE: GSM208006\n",
      "21-Dec-2017 12:32:11 DEBUG GEOparse - SAMPLE: GSM208007\n",
      "21-Dec-2017 12:32:11 DEBUG GEOparse - SAMPLE: GSM208008\n",
      "21-Dec-2017 12:32:11 DEBUG GEOparse - SAMPLE: GSM208009\n",
      "21-Dec-2017 12:32:11 DEBUG GEOparse - SAMPLE: GSM208010\n",
      "21-Dec-2017 12:32:11 DEBUG GEOparse - SAMPLE: GSM208011\n"
     ]
    }
   ],
   "source": [
    "dataSet = GEOparse.get_GEO(filepath=\"_data/genomic_data/melanoma/GSE8401_family.soft.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'email': ['geo@ncbi.nlm.nih.gov'],\n",
       " 'institute': ['NCBI NLM NIH'],\n",
       " 'name': ['Gene Expression Omnibus (GEO)'],\n",
       " 'web_link': ['http://www.ncbi.nlm.nih.gov/geo']}"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataSet.database.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset name:GSE8401 \n",
      "++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "!Series_type = Expression profiling by array\n",
      "!Series_contact_zip/postal_code = 02139\n",
      "!Series_platform_id = GPL96\n",
      "!Series_contributor = Lei,,Xu\n",
      "!Series_contributor = Steven,,Shen\n",
      "!Series_contributor = Richard,,Hynes\n",
      "!Series_sample_taxid = 9606\n",
      "!Series_submission_date = Jul 06 2007\n",
      "!Series_contact_email = ssshen@mit.edu\n",
      "!Series_platform_taxid = 9606\n",
      "!Series_contact_country = USA\n",
      "!Series_overall_design = test signature genereated from experiments of parental cell line vs xenograft model\n",
      "!Series_relation = BioProject: https://www.ncbi.nlm.nih.gov/bioproject/PRJNA101453\n",
      "!Series_pubmed_id = 18505921\n",
      "!Series_contact_state = MA\n",
      "!Series_contact_address = 77 mass ave\n",
      "!Series_sample_id = GSM207929\n",
      "!Series_sample_id = GSM207930\n",
      "!Series_sample_id = GSM207931\n",
      "!Series_sample_id = GSM207932\n",
      "!Series_sample_id = GSM207933\n",
      "!Series_sample_id = GSM207934\n",
      "!Series_sample_id = GSM207935\n",
      "!Series_sample_id = GSM207936\n",
      "!Series_sample_id = GSM207937\n",
      "!Series_sample_id = GSM207938\n",
      "!Series_sample_id = GSM207939\n",
      "!Series_sample_id = GSM207940\n",
      "!Series_sample_id = GSM207941\n",
      "!Series_sample_id = GSM207942\n",
      "!Series_sample_id = GSM207943\n",
      "!Series_sample_id = GSM207944\n",
      "!Series_sample_id = GSM207945\n",
      "!Series_sample_id = GSM207946\n",
      "!Series_sample_id = GSM207947\n",
      "!Series_sample_id = GSM207948\n",
      "!Series_sample_id = GSM207949\n",
      "!Series_sample_id = GSM207950\n",
      "!Series_sample_id = GSM207951\n",
      "!Series_sample_id = GSM207952\n",
      "!Series_sample_id = GSM207953\n",
      "!Series_sample_id = GSM207954\n",
      "!Series_sample_id = GSM207955\n",
      "!Series_sample_id = GSM207956\n",
      "!Series_sample_id = GSM207957\n",
      "!Series_sample_id = GSM207958\n",
      "!Series_sample_id = GSM207959\n",
      "!Series_sample_id = GSM207960\n",
      "!Series_sample_id = GSM207961\n",
      "!Series_sample_id = GSM207962\n",
      "!Series_sample_id = GSM207963\n",
      "!Series_sample_id = GSM207964\n",
      "!Series_sample_id = GSM207965\n",
      "!Series_sample_id = GSM207966\n",
      "!Series_sample_id = GSM207967\n",
      "!Series_sample_id = GSM207968\n",
      "!Series_sample_id = GSM207969\n",
      "!Series_sample_id = GSM207970\n",
      "!Series_sample_id = GSM207971\n",
      "!Series_sample_id = GSM207972\n",
      "!Series_sample_id = GSM207973\n",
      "!Series_sample_id = GSM207974\n",
      "!Series_sample_id = GSM207975\n",
      "!Series_sample_id = GSM207976\n",
      "!Series_sample_id = GSM207977\n",
      "!Series_sample_id = GSM207978\n",
      "!Series_sample_id = GSM207979\n",
      "!Series_sample_id = GSM207980\n",
      "!Series_sample_id = GSM207981\n",
      "!Series_sample_id = GSM207982\n",
      "!Series_sample_id = GSM207983\n",
      "!Series_sample_id = GSM207984\n",
      "!Series_sample_id = GSM207985\n",
      "!Series_sample_id = GSM207986\n",
      "!Series_sample_id = GSM207987\n",
      "!Series_sample_id = GSM207988\n",
      "!Series_sample_id = GSM207989\n",
      "!Series_sample_id = GSM207990\n",
      "!Series_sample_id = GSM207991\n",
      "!Series_sample_id = GSM207992\n",
      "!Series_sample_id = GSM207993\n",
      "!Series_sample_id = GSM207994\n",
      "!Series_sample_id = GSM207995\n",
      "!Series_sample_id = GSM207996\n",
      "!Series_sample_id = GSM207997\n",
      "!Series_sample_id = GSM207998\n",
      "!Series_sample_id = GSM207999\n",
      "!Series_sample_id = GSM208000\n",
      "!Series_sample_id = GSM208001\n",
      "!Series_sample_id = GSM208002\n",
      "!Series_sample_id = GSM208003\n",
      "!Series_sample_id = GSM208004\n",
      "!Series_sample_id = GSM208005\n",
      "!Series_sample_id = GSM208006\n",
      "!Series_sample_id = GSM208007\n",
      "!Series_sample_id = GSM208008\n",
      "!Series_sample_id = GSM208009\n",
      "!Series_sample_id = GSM208010\n",
      "!Series_sample_id = GSM208011\n",
      "!Series_contact_name = steve,,shen\n",
      "!Series_contact_institute = MIT\n",
      "!Series_last_update_date = Jul 01 2016\n",
      "!Series_supplementary_file = ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE8nnn/GSE8401/suppl/GSE8401_RAW.tar\n",
      "!Series_title = Gene Signature for Aggression of Melanoma Metastases - Melanoma Metastasis\n",
      "!Series_contact_city = cambridge\n",
      "!Series_status = Public on Oct 02 2007\n",
      "!Series_geo_accession = GSE8401\n",
      "!Series_summary = Metastasis is the deadliest phase of cancer progression. Experimental models using immunodeficient mice have been used to gain insights into the mechanisms of metastasis. We report here the identification of a “metastasis aggressiveness gene expression signature” derived using human melanoma cells selected based on their metastatic potentials in a xenotransplant metastasis model. Comparison with expression data from human melanoma patients shows that this metastasis gene signature correlates with the aggressiveness of melanoma metastases in human patients. Many genes encoding secreted and membrane proteins are included in the signature, suggesting the importance of tumor-microenvironment interactions during metastasis.\n",
      "!Series_summary = Keywords: disease state\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(\"Dataset name:{} \".format(dataSet.name))\n",
    "print(\"+\"*50)\n",
    "print(dataSet.show_metadata())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_table = dataSet.pivot_samples('VALUE')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <td>Affymetrix Probe Set ID LINK_PRE:\"https://www....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_ACC</th>\n",
       "      <td>GenBank Accession Number LINK_PRE:\"http://www....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPOT_ID</th>\n",
       "      <td>identifies controls</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Species Scientific Name</th>\n",
       "      <td>The genus and species of the organism represen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Annotation Date</th>\n",
       "      <td>The date that the annotations for this probe a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sequence Type</th>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sequence Source</th>\n",
       "      <td>The database from which the sequence used to d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Target Description</th>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Representative Public ID</th>\n",
       "      <td>The accession number of a representative seque...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Title</th>\n",
       "      <td>Title of Gene represented by the probe set.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Symbol</th>\n",
       "      <td>A gene symbol, when one is available (from Uni...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <td>Entrez Gene Database UID LINK_PRE:\"http://www....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RefSeq Transcript ID</th>\n",
       "      <td>References to multiple sequences in RefSeq. Th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Ontology Biological Process</th>\n",
       "      <td>Gene Ontology Consortium Biological Process de...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Ontology Cellular Component</th>\n",
       "      <td>Gene Ontology Consortium Cellular Component de...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Ontology Molecular Function</th>\n",
       "      <td>Gene Ontology Consortium Molecular Function de...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                        description\n",
       "ID                                Affymetrix Probe Set ID LINK_PRE:\"https://www....\n",
       "GB_ACC                            GenBank Accession Number LINK_PRE:\"http://www....\n",
       "SPOT_ID                                                         identifies controls\n",
       "Species Scientific Name           The genus and species of the organism represen...\n",
       "Annotation Date                   The date that the annotations for this probe a...\n",
       "Sequence Type                                                                      \n",
       "Sequence Source                   The database from which the sequence used to d...\n",
       "Target Description                                                                 \n",
       "Representative Public ID          The accession number of a representative seque...\n",
       "Gene Title                              Title of Gene represented by the probe set.\n",
       "Gene Symbol                       A gene symbol, when one is available (from Uni...\n",
       "ENTREZ_GENE_ID                    Entrez Gene Database UID LINK_PRE:\"http://www....\n",
       "RefSeq Transcript ID              References to multiple sequences in RefSeq. Th...\n",
       "Gene Ontology Biological Process  Gene Ontology Consortium Biological Process de...\n",
       "Gene Ontology Cellular Component  Gene Ontology Consortium Cellular Component de...\n",
       "Gene Ontology Molecular Function  Gene Ontology Consortium Molecular Function de..."
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataSet.gpls[list(dataSet.gpls.keys())[0]].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [],
   "source": [
    "# here 'characteristics_ch1.0.Genotype' is the target variable!\n",
    "df_type = pandas.DataFrame(data=dataSet.phenotype_data['characteristics_ch1.0.Genotype'], \n",
    "                           columns=['sample', 'type'])\n",
    "#df_type.index = df_type['sample']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_type = []\n",
    "for idx, row in dataSet.phenotype_data.iterrows():\n",
    "    list_type.append({\"sample\":idx, \"type\":row['characteristics_ch1.0.Genotype']})\n",
    "df_type = pandas.DataFrame(list_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_df = genome_table.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df = genome_df.merge(df_type, how='inner', left_index=True,right_on='sample')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
