{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "import GEOparse\n",
    "import sys\n",
    "import pandas \n",
    "pandas.set_option('display.max_colwidth', -1)\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filepath = \"_data/genomic_data/melanoma/GSE8401_family.soft.gz\" \n",
    "#filepath = \"_data/genomic_data/melanoma_2/GSE7553_family.soft.gz\"\n",
    "filepath = \"_data/genomic_data/melanoma_2/GSE15605_family.soft.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "21-May-2018 21:33:49 INFO GEOparse - Parsing _data/genomic_data/melanoma_2/GSE15605_family.soft.gz: \n",
      "21-May-2018 21:33:49 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "21-May-2018 21:33:49 DEBUG GEOparse - SERIES: GSE15605\n",
      "21-May-2018 21:33:49 DEBUG GEOparse - PLATFORM: GPL570\n",
      "/usr/local/lib/python3.6/dist-packages/GEOparse/GEOparse.py:332: FutureWarning: from_csv is deprecated. Please use read_csv(...) instead. Note that some of the default arguments are different, so please refer to the documentation for from_csv when changing your function calls\n",
      "  return DataFrame.from_csv(StringIO(data), index_col=None, sep=\"\\t\")\n",
      "/usr/local/lib/python3.6/dist-packages/GEOparse/GEOparse.py:521: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gpls[entry_name] = parse_GPL(data_group, entry_name)\n",
      "21-May-2018 21:33:51 DEBUG GEOparse - SAMPLE: GSM390208\n",
      "21-May-2018 21:33:51 DEBUG GEOparse - SAMPLE: GSM390209\n",
      "21-May-2018 21:33:51 DEBUG GEOparse - SAMPLE: GSM390210\n",
      "21-May-2018 21:33:51 DEBUG GEOparse - SAMPLE: GSM390211\n",
      "21-May-2018 21:33:51 DEBUG GEOparse - SAMPLE: GSM390212\n",
      "21-May-2018 21:33:52 DEBUG GEOparse - SAMPLE: GSM390213\n",
      "21-May-2018 21:33:52 DEBUG GEOparse - SAMPLE: GSM390214\n",
      "21-May-2018 21:33:52 DEBUG GEOparse - SAMPLE: GSM390215\n",
      "21-May-2018 21:33:52 DEBUG GEOparse - SAMPLE: GSM390216\n",
      "21-May-2018 21:33:52 DEBUG GEOparse - SAMPLE: GSM390217\n",
      "21-May-2018 21:33:52 DEBUG GEOparse - SAMPLE: GSM390218\n",
      "21-May-2018 21:33:52 DEBUG GEOparse - SAMPLE: GSM390219\n",
      "21-May-2018 21:33:52 DEBUG GEOparse - SAMPLE: GSM390220\n",
      "21-May-2018 21:33:52 DEBUG GEOparse - SAMPLE: GSM390221\n",
      "21-May-2018 21:33:53 DEBUG GEOparse - SAMPLE: GSM390222\n",
      "21-May-2018 21:33:53 DEBUG GEOparse - SAMPLE: GSM390223\n",
      "21-May-2018 21:33:53 DEBUG GEOparse - SAMPLE: GSM390224\n",
      "21-May-2018 21:33:53 DEBUG GEOparse - SAMPLE: GSM390225\n",
      "21-May-2018 21:33:53 DEBUG GEOparse - SAMPLE: GSM390226\n",
      "21-May-2018 21:33:53 DEBUG GEOparse - SAMPLE: GSM390227\n",
      "21-May-2018 21:33:53 DEBUG GEOparse - SAMPLE: GSM390228\n",
      "21-May-2018 21:33:53 DEBUG GEOparse - SAMPLE: GSM390229\n",
      "21-May-2018 21:33:53 DEBUG GEOparse - SAMPLE: GSM390230\n",
      "21-May-2018 21:33:54 DEBUG GEOparse - SAMPLE: GSM390231\n",
      "21-May-2018 21:33:54 DEBUG GEOparse - SAMPLE: GSM390232\n",
      "21-May-2018 21:33:54 DEBUG GEOparse - SAMPLE: GSM390233\n",
      "21-May-2018 21:33:54 DEBUG GEOparse - SAMPLE: GSM390234\n",
      "21-May-2018 21:33:54 DEBUG GEOparse - SAMPLE: GSM390235\n",
      "21-May-2018 21:33:54 DEBUG GEOparse - SAMPLE: GSM390236\n",
      "21-May-2018 21:33:54 DEBUG GEOparse - SAMPLE: GSM390237\n",
      "21-May-2018 21:33:54 DEBUG GEOparse - SAMPLE: GSM390238\n",
      "21-May-2018 21:33:54 DEBUG GEOparse - SAMPLE: GSM390239\n",
      "21-May-2018 21:33:55 DEBUG GEOparse - SAMPLE: GSM390240\n",
      "21-May-2018 21:33:55 DEBUG GEOparse - SAMPLE: GSM390241\n",
      "21-May-2018 21:33:55 DEBUG GEOparse - SAMPLE: GSM390242\n",
      "21-May-2018 21:33:55 DEBUG GEOparse - SAMPLE: GSM390243\n",
      "21-May-2018 21:33:55 DEBUG GEOparse - SAMPLE: GSM390244\n",
      "21-May-2018 21:33:55 DEBUG GEOparse - SAMPLE: GSM390245\n",
      "21-May-2018 21:33:55 DEBUG GEOparse - SAMPLE: GSM390246\n",
      "21-May-2018 21:33:55 DEBUG GEOparse - SAMPLE: GSM390247\n",
      "21-May-2018 21:33:55 DEBUG GEOparse - SAMPLE: GSM390248\n",
      "21-May-2018 21:33:56 DEBUG GEOparse - SAMPLE: GSM390249\n",
      "21-May-2018 21:33:56 DEBUG GEOparse - SAMPLE: GSM390250\n",
      "21-May-2018 21:33:56 DEBUG GEOparse - SAMPLE: GSM390251\n",
      "21-May-2018 21:33:56 DEBUG GEOparse - SAMPLE: GSM390252\n",
      "21-May-2018 21:33:56 DEBUG GEOparse - SAMPLE: GSM390253\n",
      "21-May-2018 21:33:56 DEBUG GEOparse - SAMPLE: GSM390254\n",
      "21-May-2018 21:33:56 DEBUG GEOparse - SAMPLE: GSM390255\n",
      "21-May-2018 21:33:56 DEBUG GEOparse - SAMPLE: GSM390256\n",
      "21-May-2018 21:33:56 DEBUG GEOparse - SAMPLE: GSM390257\n",
      "21-May-2018 21:33:57 DEBUG GEOparse - SAMPLE: GSM390258\n",
      "21-May-2018 21:33:57 DEBUG GEOparse - SAMPLE: GSM390259\n",
      "21-May-2018 21:33:57 DEBUG GEOparse - SAMPLE: GSM390260\n",
      "21-May-2018 21:33:57 DEBUG GEOparse - SAMPLE: GSM390261\n",
      "21-May-2018 21:33:57 DEBUG GEOparse - SAMPLE: GSM390262\n",
      "21-May-2018 21:33:57 DEBUG GEOparse - SAMPLE: GSM390263\n",
      "21-May-2018 21:33:57 DEBUG GEOparse - SAMPLE: GSM390264\n",
      "21-May-2018 21:33:57 DEBUG GEOparse - SAMPLE: GSM390265\n",
      "21-May-2018 21:33:58 DEBUG GEOparse - SAMPLE: GSM390266\n",
      "21-May-2018 21:33:58 DEBUG GEOparse - SAMPLE: GSM390267\n",
      "21-May-2018 21:33:58 DEBUG GEOparse - SAMPLE: GSM390268\n",
      "21-May-2018 21:33:58 DEBUG GEOparse - SAMPLE: GSM390269\n",
      "21-May-2018 21:33:58 DEBUG GEOparse - SAMPLE: GSM390270\n",
      "21-May-2018 21:33:58 DEBUG GEOparse - SAMPLE: GSM390271\n",
      "21-May-2018 21:33:58 DEBUG GEOparse - SAMPLE: GSM390272\n",
      "21-May-2018 21:33:58 DEBUG GEOparse - SAMPLE: GSM390273\n",
      "21-May-2018 21:33:59 DEBUG GEOparse - SAMPLE: GSM390274\n",
      "21-May-2018 21:33:59 DEBUG GEOparse - SAMPLE: GSM390275\n",
      "21-May-2018 21:33:59 DEBUG GEOparse - SAMPLE: GSM390276\n",
      "21-May-2018 21:33:59 DEBUG GEOparse - SAMPLE: GSM390277\n",
      "21-May-2018 21:33:59 DEBUG GEOparse - SAMPLE: GSM390278\n",
      "21-May-2018 21:33:59 DEBUG GEOparse - SAMPLE: GSM390279\n",
      "21-May-2018 21:33:59 DEBUG GEOparse - SAMPLE: GSM390280\n",
      "21-May-2018 21:33:59 DEBUG GEOparse - SAMPLE: GSM390281\n"
     ]
    }
   ],
   "source": [
    "dataSet = GEOparse.get_GEO(filepath=filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': ['Gene Expression Omnibus (GEO)'],\n",
       " 'institute': ['NCBI NLM NIH'],\n",
       " 'web_link': ['http://www.ncbi.nlm.nih.gov/geo'],\n",
       " 'email': ['geo@ncbi.nlm.nih.gov']}"
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataSet.database.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset name:GSE15605 \n",
      "++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "!Series_title = Transcriptome profiling identifies HMGA2 as a novel gene in melanoma progression\n",
      "!Series_geo_accession = GSE15605\n",
      "!Series_status = Public on Jun 08 2012\n",
      "!Series_submission_date = Apr 08 2009\n",
      "!Series_last_update_date = Apr 23 2018\n",
      "!Series_pubmed_id = 23633021\n",
      "!Series_summary = The identification of novel tumor-specific markers may improve understanding of melanoma progression and prognostic accuracy. Whole genome expression profiling of 46 primary melanomas, 12 metastases, and 16 normal skin samples using Affymetrix U133 PLUS 2.0 array generated gene lists including both known and new melanoma genes.\n",
      "!Series_summary = The molecular genetic alterations contributing to the pathogenesis of melanoma are incompletely defined and few independent prognostic features have been identified beyond the pathologic characteristics of the primary tumor. Early stage melanoma is frequently curable, in contrast to the inferior prognosis of melanomas with regional lymph nodes involvement and the incurability of distant metastatic disease. The identification of novel tumor-specific markers may improve our understanding of melanoma progression and prognostic accuracy. To find differentially expressed genes that can distinguish melanoma from normal tissue, we performed a whole genome expression profiling of 46 primary melanoma samples, 12 regional or distant metastases, and 16 normal skin samples using Affymetrix U133 2.0 Plus array. Our study generated lists of differentially expressed genes in melanoma and identified novel prognostic marker HMGA2. It is  a novel, highly overexpressed melanoma gene associated with poor prognosis. The overexpression of HMGA2 is strongly associated with regional and distant metastases and serves as an independent predictor of disease-free survival and overall survival in melanoma.\n",
      "!Series_overall_design = Melanoma samples were obtained through an IRB-approved protocol using informed consent at the University of Michigan Multidisciplinary Melanoma Clinic.  A portion of pigmented lesions clinically suspicious for melanoma and known melanoma were sampled by punch biopsy at the time of excision. The punch biopsy was bisected; half was sent to for clinical diagnosis and the other half along with adjacent normal skin when available immediately snap-frozen in liquid nitrogen. Snap-frozen tissue was embedded in OCT (TissueTek) followed by frozen sectioning and H&E slide preparation. The slides were evaluated by dermatopathologist who identified areas with greater than 70% tumor cellularity, which were sampled for RNA extraction. Primary melanomas and melanoma metastases were derived from different patients. Metastatic samples included lymph nodes (n=8), subcutaneous soft tissue (n=2), spleen (n=1), and small intestine (n=1).\n",
      "!Series_type = Expression profiling by array\n",
      "!Series_contributor = Timothy,M,Johnson\n",
      "!Series_contributor = Douglas,R,Fullen\n",
      "!Series_contributor = Thomas,J,Giordano\n",
      "!Series_contributor = Stephen,B,Gruber\n",
      "!Series_sample_id = GSM390208\n",
      "!Series_sample_id = GSM390209\n",
      "!Series_sample_id = GSM390210\n",
      "!Series_sample_id = GSM390211\n",
      "!Series_sample_id = GSM390212\n",
      "!Series_sample_id = GSM390213\n",
      "!Series_sample_id = GSM390214\n",
      "!Series_sample_id = GSM390215\n",
      "!Series_sample_id = GSM390216\n",
      "!Series_sample_id = GSM390217\n",
      "!Series_sample_id = GSM390218\n",
      "!Series_sample_id = GSM390219\n",
      "!Series_sample_id = GSM390220\n",
      "!Series_sample_id = GSM390221\n",
      "!Series_sample_id = GSM390222\n",
      "!Series_sample_id = GSM390223\n",
      "!Series_sample_id = GSM390224\n",
      "!Series_sample_id = GSM390225\n",
      "!Series_sample_id = GSM390226\n",
      "!Series_sample_id = GSM390227\n",
      "!Series_sample_id = GSM390228\n",
      "!Series_sample_id = GSM390229\n",
      "!Series_sample_id = GSM390230\n",
      "!Series_sample_id = GSM390231\n",
      "!Series_sample_id = GSM390232\n",
      "!Series_sample_id = GSM390233\n",
      "!Series_sample_id = GSM390234\n",
      "!Series_sample_id = GSM390235\n",
      "!Series_sample_id = GSM390236\n",
      "!Series_sample_id = GSM390237\n",
      "!Series_sample_id = GSM390238\n",
      "!Series_sample_id = GSM390239\n",
      "!Series_sample_id = GSM390240\n",
      "!Series_sample_id = GSM390241\n",
      "!Series_sample_id = GSM390242\n",
      "!Series_sample_id = GSM390243\n",
      "!Series_sample_id = GSM390244\n",
      "!Series_sample_id = GSM390245\n",
      "!Series_sample_id = GSM390246\n",
      "!Series_sample_id = GSM390247\n",
      "!Series_sample_id = GSM390248\n",
      "!Series_sample_id = GSM390249\n",
      "!Series_sample_id = GSM390250\n",
      "!Series_sample_id = GSM390251\n",
      "!Series_sample_id = GSM390252\n",
      "!Series_sample_id = GSM390253\n",
      "!Series_sample_id = GSM390254\n",
      "!Series_sample_id = GSM390255\n",
      "!Series_sample_id = GSM390256\n",
      "!Series_sample_id = GSM390257\n",
      "!Series_sample_id = GSM390258\n",
      "!Series_sample_id = GSM390259\n",
      "!Series_sample_id = GSM390260\n",
      "!Series_sample_id = GSM390261\n",
      "!Series_sample_id = GSM390262\n",
      "!Series_sample_id = GSM390263\n",
      "!Series_sample_id = GSM390264\n",
      "!Series_sample_id = GSM390265\n",
      "!Series_sample_id = GSM390266\n",
      "!Series_sample_id = GSM390267\n",
      "!Series_sample_id = GSM390268\n",
      "!Series_sample_id = GSM390269\n",
      "!Series_sample_id = GSM390270\n",
      "!Series_sample_id = GSM390271\n",
      "!Series_sample_id = GSM390272\n",
      "!Series_sample_id = GSM390273\n",
      "!Series_sample_id = GSM390274\n",
      "!Series_sample_id = GSM390275\n",
      "!Series_sample_id = GSM390276\n",
      "!Series_sample_id = GSM390277\n",
      "!Series_sample_id = GSM390278\n",
      "!Series_sample_id = GSM390279\n",
      "!Series_sample_id = GSM390280\n",
      "!Series_sample_id = GSM390281\n",
      "!Series_contact_name = Leon,,Raskin\n",
      "!Series_contact_email = leonid.raskin@vanderbilt.edu\n",
      "!Series_contact_phone = (615) 343-5378\n",
      "!Series_contact_fax = (615) 343-5938\n",
      "!Series_contact_laboratory = Raskin lab\n",
      "!Series_contact_department = Medicine\n",
      "!Series_contact_institute = Vanderbilt University\n",
      "!Series_contact_address = 2525 West End Avenue, Suite 300\n",
      "!Series_contact_city = Nashville\n",
      "!Series_contact_state = TN\n",
      "!Series_contact_zip/postal_code = 37203\n",
      "!Series_contact_country = USA\n",
      "!Series_supplementary_file = ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE15nnn/GSE15605/suppl/GSE15605_RAW.tar\n",
      "!Series_platform_id = GPL570\n",
      "!Series_platform_taxid = 9606\n",
      "!Series_sample_taxid = 9606\n",
      "!Series_relation = BioProject: https://www.ncbi.nlm.nih.gov/bioproject/PRJNA115729\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(\"Dataset name:{} \".format(dataSet.name))\n",
    "print(\"+\"*50)\n",
    "print(dataSet.show_metadata())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_table = dataSet.pivot_samples('VALUE')\n",
    "genome_df = genome_table.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <td>Affymetrix Probe Set ID LINK_PRE:\"https://www.affymetrix.com/LinkServlet?array=U133PLUS&amp;probeset=\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_ACC</th>\n",
       "      <td>GenBank Accession Number LINK_PRE:\"http://www.ncbi.nlm.nih.gov/nuccore/?term=\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPOT_ID</th>\n",
       "      <td>identifies controls</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Species Scientific Name</th>\n",
       "      <td>The genus and species of the organism represented by the probe set.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Annotation Date</th>\n",
       "      <td>The date that the annotations for this probe array were last updated. It will generally be earlier than the date when the annotations were posted on the Affymetrix web site.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sequence Type</th>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sequence Source</th>\n",
       "      <td>The database from which the sequence used to design this probe set was taken.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Target Description</th>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Representative Public ID</th>\n",
       "      <td>The accession number of a representative sequence. Note that for consensus-based probe sets, the representative sequence is only one of several sequences (sequence sub-clusters) used to build the consensus sequence and it is not directly used to derive the probe sequences. The representative sequence is chosen during array design as a sequence that is best associated with the transcribed region being interrogated by the probe set. Refer to the \"Sequence Source\" field to determine the database used.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Title</th>\n",
       "      <td>Title of Gene represented by the probe set.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Symbol</th>\n",
       "      <td>A gene symbol, when one is available (from UniGene).</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <td>Entrez Gene Database UID LINK_PRE:\"http://www.ncbi.nlm.nih.gov/gene/\" DELIMIT:\" /// \"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RefSeq Transcript ID</th>\n",
       "      <td>References to multiple sequences in RefSeq. The field contains the ID and Description for each entry, and there can be multiple entries per ProbeSet.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Ontology Biological Process</th>\n",
       "      <td>Gene Ontology Consortium Biological Process derived from LocusLink.  Each annotation consists of three parts: \"Accession Number // Description // Evidence\". The description corresponds directly to the GO ID. The evidence can be \"direct\", or \"extended\".</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Ontology Cellular Component</th>\n",
       "      <td>Gene Ontology Consortium Cellular Component derived from LocusLink.  Each annotation consists of three parts: \"Accession Number // Description // Evidence\". The description corresponds directly to the GO ID. The evidence can be \"direct\", or \"extended\".</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Ontology Molecular Function</th>\n",
       "      <td>Gene Ontology Consortium Molecular Function derived from LocusLink.  Each annotation consists of three parts: \"Accession Number // Description // Evidence\". The description corresponds directly to the GO ID. The evidence can be \"direct\", or \"extended\".</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              description\n",
       "ID                                Affymetrix Probe Set ID LINK_PRE:\"https://www.affymetrix.com/LinkServlet?array=U133PLUS&probeset=\"                                                                                                                                                                                                                                                                                                                                                                                                                     \n",
       "GB_ACC                            GenBank Accession Number LINK_PRE:\"http://www.ncbi.nlm.nih.gov/nuccore/?term=\"                                                                                                                                                                                                                                                                                                                                                                                                                                         \n",
       "SPOT_ID                           identifies controls                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \n",
       "Species Scientific Name           The genus and species of the organism represented by the probe set.                                                                                                                                                                                                                                                                                                                                                                                                                                                    \n",
       "Annotation Date                   The date that the annotations for this probe array were last updated. It will generally be earlier than the date when the annotations were posted on the Affymetrix web site.                                                                                                                                                                                                                                                                                                                                          \n",
       "Sequence Type                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            \n",
       "Sequence Source                   The database from which the sequence used to design this probe set was taken.                                                                                                                                                                                                                                                                                                                                                                                                                                          \n",
       "Target Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       \n",
       "Representative Public ID          The accession number of a representative sequence. Note that for consensus-based probe sets, the representative sequence is only one of several sequences (sequence sub-clusters) used to build the consensus sequence and it is not directly used to derive the probe sequences. The representative sequence is chosen during array design as a sequence that is best associated with the transcribed region being interrogated by the probe set. Refer to the \"Sequence Source\" field to determine the database used.\n",
       "Gene Title                        Title of Gene represented by the probe set.                                                                                                                                                                                                                                                                                                                                                                                                                                                                            \n",
       "Gene Symbol                       A gene symbol, when one is available (from UniGene).                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \n",
       "ENTREZ_GENE_ID                    Entrez Gene Database UID LINK_PRE:\"http://www.ncbi.nlm.nih.gov/gene/\" DELIMIT:\" /// \"                                                                                                                                                                                                                                                                                                                                                                                                                                  \n",
       "RefSeq Transcript ID              References to multiple sequences in RefSeq. The field contains the ID and Description for each entry, and there can be multiple entries per ProbeSet.                                                                                                                                                                                                                                                                                                                                                                  \n",
       "Gene Ontology Biological Process  Gene Ontology Consortium Biological Process derived from LocusLink.  Each annotation consists of three parts: \"Accession Number // Description // Evidence\". The description corresponds directly to the GO ID. The evidence can be \"direct\", or \"extended\".                                                                                                                                                                                                                                                           \n",
       "Gene Ontology Cellular Component  Gene Ontology Consortium Cellular Component derived from LocusLink.  Each annotation consists of three parts: \"Accession Number // Description // Evidence\". The description corresponds directly to the GO ID. The evidence can be \"direct\", or \"extended\".                                                                                                                                                                                                                                                           \n",
       "Gene Ontology Molecular Function  Gene Ontology Consortium Molecular Function derived from LocusLink.  Each annotation consists of three parts: \"Accession Number // Description // Evidence\". The description corresponds directly to the GO ID. The evidence can be \"direct\", or \"extended\".                                                                                                                                                                                                                                                           "
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataSet.gpls[list(dataSet.gpls.keys())[0]].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [],
   "source": [
    "characteristics =[col_name for col_name in dataSet.phenotype_data.columns \n",
    "                  if any(target in col_name for target in ['characteristics',\n",
    "                                                           'label', \n",
    "                                                           'type', \n",
    "                                                           'source_name',\n",
    "                                                           'organisme',\n",
    "                                                           'molecule',\n",
    "                                                           'title'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "characteristics_ch1.0.gender : (2,) : ['m' 'f']\n",
      "characteristics_ch1.1.age at diagnosis : (44,) : ['79' '47' '44' '85' '39' '41' '50' '74' '56' '73' '58' '20' '62' '55'\n",
      " '52' '66' '60' '76' '59' '63' '42' '81' '83' '80' '72' '61' '33' '45'\n",
      " '51' '70' '67' '43' '48' '36' '87' '92' '64' '40' '27' '71' '77' '90'\n",
      " '53' '89']\n",
      "characteristics_ch1.2.diagnosis : (13,) : ['NORMAL SKIN' 'DESMOPLASTIC MELANOMA' 'SSM' 'NODULAR MELANOMA' 'MIS'\n",
      " 'SPINDLE CELL MELANOMA' 'ACRAL MELANOMA' 'DERMAL  MELANOMA'\n",
      " 'POLYPOID MELANOMA' 'LMM' 'REGIONAL LYMPH NODE METASTASIS'\n",
      " 'DISTANT METASTASIS' 'SUBCUTANIOUS SOFT TISSUE METASTASIS']\n",
      "characteristics_ch1.3.braf mutation status : (4,) : [nan 'V600R' 'WT' 'V600K']\n",
      "characteristics_ch1.3.distant metastases : (2,) : [nan 'YES']\n",
      "characteristics_ch1.3.regional metastases : (4,) : ['not applicable' 'NO' 'YES' nan]\n",
      "characteristics_ch1.4.braf mutation status : (4,) : [nan 'V600E' 'WT' 'V600K']\n",
      "characteristics_ch1.4.distant metastases : (4,) : ['not applicable' 'NO' 'YES' nan]\n",
      "characteristics_ch1.4.nras mutation status : (2,) : [nan 'WT']\n",
      "characteristics_ch1.5.braf mutation status : (7,) : ['not applicable' 'WT' nan 'V600E' 'L597S' 'V600K' 'L597R']\n",
      "characteristics_ch1.5.nras mutation status : (3,) : [nan 'WT' 'Q61K']\n",
      "characteristics_ch1.6.nras mutation status : (6,) : ['not applicable' 'WT' 'Q61R' nan 'Q61K' 'Q61L']\n",
      "source_name_ch1 : (3,) : ['Frozen tip of melanoma excision'\n",
      " 'Frozen half of the melanoma punch biopsy' 'Frozen melanoma metastasis']\n",
      "title : (74,) : ['Normal_skin N118' 'Normal_skin N157' 'Normal_skin N173'\n",
      " 'Normal_skin N201' 'Normal_skin N206' 'Normal_skin N236'\n",
      " 'Normal_skin N243' 'Normal_skin N273' 'Normal_skin N275'\n",
      " 'Normal_skin N280' 'Normal_skin N287' 'Normal_skin N288'\n",
      " 'Normal_skin N290' 'Normal_skin N298' 'Normal_skin N307'\n",
      " 'Normal_skin N356' 'Primary_melanoma MEL101' 'Primary_melanoma MEL128'\n",
      " 'Primary_melanoma MEL131' 'Primary_melanoma MEL135'\n",
      " 'Primary_melanoma MEL142' 'Primary_melanoma MEL145'\n",
      " 'Primary_melanoma MEL157' 'Primary_melanoma MEL173'\n",
      " 'Primary_melanoma MEL176' 'Primary_melanoma MEL185'\n",
      " 'Primary_melanoma MEL190' 'Primary_melanoma MEL197'\n",
      " 'Primary_melanoma MEL209' 'Primary_melanoma MEL213'\n",
      " 'Primary_melanoma MEL233' 'Primary_melanoma MEL236'\n",
      " 'Primary_melanoma MEL243' 'Primary_melanoma MEL244'\n",
      " 'Primary_melanoma MEL250' 'Primary_melanoma MEL257'\n",
      " 'Primary_melanoma MEL258' 'Primary_melanoma MEL272'\n",
      " 'Primary_melanoma MEL275' 'Primary_melanoma MEL276'\n",
      " 'Primary_melanoma MEL280' 'Primary_melanoma MEL282'\n",
      " 'Primary_melanoma MEL283' 'Primary_melanoma MEL287'\n",
      " 'Primary_melanoma MEL290' 'Primary_melanoma MEL294'\n",
      " 'Primary_melanoma MEL298' 'Primary_melanoma MEL307'\n",
      " 'Primary_melanoma MEL310' 'Primary_melanoma MEL326'\n",
      " 'Primary_melanoma MEL339' 'Primary_melanoma MEL340'\n",
      " 'Primary_melanoma MEL356' 'Primary_melanoma MEL362'\n",
      " 'Primary_melanoma MEL364' 'Primary_melanoma MEL375'\n",
      " 'Primary_melanoma MEL380' 'Primary_melanoma MEL385'\n",
      " 'Primary_melanoma MEL395' 'Primary_melanoma MEL420'\n",
      " 'Primary_melanoma MEL429' 'Primary_melanoma MEL430'\n",
      " 'Melanoma_metastasis METMEL195' 'Melanoma_metastasis METMEL200701'\n",
      " 'Melanoma_metastasis METMEL200703' 'Melanoma_metastasis METMEL200705'\n",
      " 'Melanoma_metastasis METMEL200707' 'Melanoma_metastasis METMEL200708'\n",
      " 'Melanoma_metastasis METMEL240' 'Melanoma_metastasis METMEL261'\n",
      " 'Melanoma_metastasis METMEL291' 'Melanoma_metastasis METMEL308'\n",
      " 'Melanoma_metastasis METMEL361' 'Melanoma_metastasis METMEL381']\n"
     ]
    }
   ],
   "source": [
    "for char in characteristics:\n",
    "    if dataSet.phenotype_data[char].unique().shape[0] > 1:\n",
    "        print(char,':',dataSet.phenotype_data[char].unique().shape,':', dataSet.phenotype_data[char].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_variable = 'source_name_ch1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'basal cell carcinoma '"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = re.compile(r'([0-9]{4}\\_[0-9]{2})')\n",
    "test.sub('', 'basal cell carcinoma 7845_17')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "filters = [re.compile(r'(group [0-9])|\\,'), \n",
    "           re.compile(r'([0-9]{4}\\_[0-9]{1,2})'),\n",
    "           re.compile(r'([0-9]{4}[a-z]?\\-[0-9]{1,2})'),\n",
    "           re.compile(r'([0-9]{4}\\s[a-z][0-9]?)'),\n",
    "           re.compile(r'([a-z][0-9]{4}[a-z])'),\n",
    "           re.compile(r'[0-9]'),\n",
    "           re.compile(r'(^\\s)|(\\s$)')\n",
    "          ]\n",
    "\n",
    "def filter_string(x):\n",
    "    for re_filter in filters:\n",
    "        x = re_filter.sub('', x.lower())\n",
    "    return x\n",
    "df_type = dataSet.phenotype_data[['source_name_ch1', 'title']].applymap(lambda x: filter_string(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "#list_type = []\n",
    "#for idx, row in dataSet.phenotype_data.iterrows():\n",
    "#    list_type.append({\"sample\":idx, \"type\":row[target_variable]})\n",
    "#df_type = pandas.DataFrame(list_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df = genome_df.merge(df_type, how='inner', left_index=True, right_index=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
