{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/h5py/__init__.py:34: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.\n",
      "  from ._conv import register_converters as _register_converters\n",
      "Using TensorFlow backend.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "++++++++++++++++++++++++++++++ Firing up RexR! ++++++++++++++++++++++++++++++\n"
     ]
    }
   ],
   "source": [
    "import pandas\n",
    "import numpy\n",
    "from RexR import *\n",
    "import _helpers\n",
    "Rocket = RexR(datalocation = None, #'_data/genomic_data/data.pkl', \n",
    "              seed = 3123, \n",
    "              debug = False, \n",
    "              write_out=True,\n",
    "              set_name = 'ALL_10') # data to read in ALL_10, or MELA\n",
    "Rocket.load_probeset_data();\n",
    "\n",
    "%matplotlib inline\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "METHOD_LIST = ['RF','XGB', 'LGBM', 'ExtraTrees','SVM', 'LR', 'MLNN', 'RVM', 'DNN']#, 'CNN'] #, 'NaiveBayes','MLNN', 'XGB'] # \n",
    "Runs = []\n",
    "nruns = 1\n",
    "SCALER = \"standard\" # minmax, standard, normaliser\n",
    "GROUPING = \"mean\"\n",
    "FEAT_SELECTOR = \"low_variance\" # \"low_variance\" None\n",
    "SELECTOR_METHOD = \"FDR\" # mannwhitney, FDR\n",
    "SELECTOR_ALPHA = 0.01 # see this as the maximum p-value to classify \n",
    "DIM_TYPE =  None #\"PCA\" #\"LDA\", \"PCA\", \"PLS\" \n",
    "DIM_NUM = 1000\n",
    "Results = None\n",
    "ACC = pd.DataFrame()\n",
    "Rocket.VIZ = False\n",
    "Rocket.DATA_merged_processed = None\n",
    "PREPROC_DICT = {\"patient_grouping\": GROUPING, \"bias_removal\": False, \"noise\": True}\n",
    "FSELECT_DICT = {\"type\": FEAT_SELECTOR, \"pvalue\": SELECTOR_ALPHA, \"method\": SELECTOR_METHOD}\n",
    "DIMRED_DICT = {\"type\": DIM_TYPE, \"n_comp\": DIM_NUM}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+ + + + + + + + + + + + + + + + + + + + + + + + + + + + + +  Prepping data, this may take a while..\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  Grouping probesets\n",
      "+ + + + + + + + + + + + + + + + + + + + + + + + + + + + + +  Creating X,y\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  Selecting features using a low_variance filter\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  Kept 4 of 54613 features using FDR with p = 0.01\n",
      "++++++++++++++++++++++++++++++  RESULTS FOR CLASSIFICATION WITH GENOMIC DATA ++++++++++++++++++++++++++++++\n",
      "++++++++++++++++++++ ..processing feature array (54, 4) and class vector (54,)\n",
      "MODEL: RF accuracy:  0.8703703703703702 +/-: 0.001883027808953732\n",
      "++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "+ + + + + + + + + + + + + + + + + + + + + + + + + + + + + +  Creating X,y\n",
      "++++++++++++++++++++++++++++++  RESULTS FOR CLASSIFICATION WITH GENOMIC DATA ++++++++++++++++++++++++++++++\n",
      "++++++++++++++++++++ ..processing feature array (54, 4) and class vector (54,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.\n",
      "  if diff:\n",
      "/usr/local/lib/python3.5/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.\n",
      "  if diff:\n",
      "/usr/local/lib/python3.5/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.\n",
      "  if diff:\n",
      "/usr/local/lib/python3.5/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.\n",
      "  if diff:\n",
      "/usr/local/lib/python3.5/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.\n",
      "  if diff:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MODEL: XGB accuracy:  0.9074074074074074 +/-: 0.010113480483850853\n",
      "++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "+ + + + + + + + + + + + + + + + + + + + + + + + + + + + + +  Creating X,y\n",
      "++++++++++++++++++++++++++++++  RESULTS FOR CLASSIFICATION WITH GENOMIC DATA ++++++++++++++++++++++++++++++\n",
      "++++++++++++++++++++ ..processing feature array (54, 4) and class vector (54,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/dist-packages/lightgbm/basic.py:642: UserWarning: max_bin keyword has been found in `params` and will be ignored. Please use max_bin argument of the Dataset constructor to pass this parameter.\n",
      "  'Please use {0} argument of the Dataset constructor to pass this parameter.'.format(key))\n",
      "/usr/local/lib/python3.5/dist-packages/lightgbm/basic.py:648: LGBMDeprecationWarning: The `max_bin` parameter is deprecated and will be removed in 2.0.12 version. Please use `params` to pass this parameter.\n",
      "  'Please use `params` to pass this parameter.', LGBMDeprecationWarning)\n",
      "/usr/local/lib/python3.5/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.\n",
      "  if diff:\n",
      "/usr/local/lib/python3.5/dist-packages/lightgbm/basic.py:642: UserWarning: max_bin keyword has been found in `params` and will be ignored. Please use max_bin argument of the Dataset constructor to pass this parameter.\n",
      "  'Please use {0} argument of the Dataset constructor to pass this parameter.'.format(key))\n",
      "/usr/local/lib/python3.5/dist-packages/lightgbm/basic.py:648: LGBMDeprecationWarning: The `max_bin` parameter is deprecated and will be removed in 2.0.12 version. Please use `params` to pass this parameter.\n",
      "  'Please use `params` to pass this parameter.', LGBMDeprecationWarning)\n",
      "/usr/local/lib/python3.5/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.\n",
      "  if diff:\n",
      "/usr/local/lib/python3.5/dist-packages/lightgbm/basic.py:642: UserWarning: max_bin keyword has been found in `params` and will be ignored. Please use max_bin argument of the Dataset constructor to pass this parameter.\n",
      "  'Please use {0} argument of the Dataset constructor to pass this parameter.'.format(key))\n",
      "/usr/local/lib/python3.5/dist-packages/lightgbm/basic.py:648: LGBMDeprecationWarning: The `max_bin` parameter is deprecated and will be removed in 2.0.12 version. Please use `params` to pass this parameter.\n",
      "  'Please use `params` to pass this parameter.', LGBMDeprecationWarning)\n",
      "/usr/local/lib/python3.5/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.\n",
      "  if diff:\n",
      "/usr/local/lib/python3.5/dist-packages/lightgbm/basic.py:642: UserWarning: max_bin keyword has been found in `params` and will be ignored. Please use max_bin argument of the Dataset constructor to pass this parameter.\n",
      "  'Please use {0} argument of the Dataset constructor to pass this parameter.'.format(key))\n",
      "/usr/local/lib/python3.5/dist-packages/lightgbm/basic.py:648: LGBMDeprecationWarning: The `max_bin` parameter is deprecated and will be removed in 2.0.12 version. Please use `params` to pass this parameter.\n",
      "  'Please use `params` to pass this parameter.', LGBMDeprecationWarning)\n",
      "/usr/local/lib/python3.5/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.\n",
      "  if diff:\n",
      "/usr/local/lib/python3.5/dist-packages/lightgbm/basic.py:642: UserWarning: max_bin keyword has been found in `params` and will be ignored. Please use max_bin argument of the Dataset constructor to pass this parameter.\n",
      "  'Please use {0} argument of the Dataset constructor to pass this parameter.'.format(key))\n",
      "/usr/local/lib/python3.5/dist-packages/lightgbm/basic.py:648: LGBMDeprecationWarning: The `max_bin` parameter is deprecated and will be removed in 2.0.12 version. Please use `params` to pass this parameter.\n",
      "  'Please use `params` to pass this parameter.', LGBMDeprecationWarning)\n",
      "/usr/local/lib/python3.5/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.\n",
      "  if diff:\n",
      "/usr/local/lib/python3.5/dist-packages/lightgbm/basic.py:642: UserWarning: max_bin keyword has been found in `params` and will be ignored. Please use max_bin argument of the Dataset constructor to pass this parameter.\n",
      "  'Please use {0} argument of the Dataset constructor to pass this parameter.'.format(key))\n",
      "/usr/local/lib/python3.5/dist-packages/lightgbm/basic.py:648: LGBMDeprecationWarning: The `max_bin` parameter is deprecated and will be removed in 2.0.12 version. Please use `params` to pass this parameter.\n",
      "  'Please use `params` to pass this parameter.', LGBMDeprecationWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MODEL: LGBM accuracy:  0.9444444444444444 +/-: 0.005499438832772164\n",
      "++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "+ + + + + + + + + + + + + + + + + + + + + + + + + + + + + +  Creating X,y\n",
      "++++++++++++++++++++++++++++++  RESULTS FOR CLASSIFICATION WITH GENOMIC DATA ++++++++++++++++++++++++++++++\n",
      "++++++++++++++++++++ ..processing feature array (54, 4) and class vector (54,)\n",
      "MODEL: ET accuracy:  0.8888888888888887 +/-: 0.0012906846240179554\n",
      "++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "+ + + + + + + + + + + + + + + + + + + + + + + + + + + + + +  Creating X,y\n",
      "++++++++++++++++++++++++++++++  RESULTS FOR CLASSIFICATION WITH GENOMIC DATA ++++++++++++++++++++++++++++++\n",
      "++++++++++++++++++++ ..processing feature array (54, 4) and class vector (54,)\n",
      "MODEL: SVM accuracy:  0.8518518518518516 +/-: 0.0017894999376480837\n",
      "++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "+ + + + + + + + + + + + + + + + + + + + + + + + + + + + + +  Creating X,y\n",
      "++++++++++++++++++++++++++++++  RESULTS FOR CLASSIFICATION WITH GENOMIC DATA ++++++++++++++++++++++++++++++\n",
      "++++++++++++++++++++ ..processing feature array (54, 4) and class vector (54,)\n",
      "MODEL: LR accuracy:  0.8703703703703702 +/-: 0.001883027808953732\n",
      "++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "+ + + + + + + + + + + + + + + + + + + + + + + + + + + + + +  Creating X,y\n",
      "++++++++++++++++++++++++++++++  RESULTS FOR CLASSIFICATION WITH GENOMIC DATA ++++++++++++++++++++++++++++++\n",
      "++++++++++++++++++++ ..processing feature array (54, 4) and class vector (54,)\n",
      "MODEL: MLNN accuracy:  0.8703703703703702 +/-: 0.001883027808953732\n",
      "++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "+ + + + + + + + + + + + + + + + + + + + + + + + + + + + + +  Creating X,y\n",
      "++++++++++++++++++++++++++++++  RESULTS FOR CLASSIFICATION WITH GENOMIC DATA ++++++++++++++++++++++++++++++\n",
      "++++++++++++++++++++ ..processing feature array (54, 4) and class vector (54,)\n",
      "Initial alpha = [[0.12193607]]\n",
      "   1 - L=-1034.5359698 - Gamma= 1.9999299 (M=   2) - s=0.0100\n",
      "   2 - L=-1016.6849590 - Gamma= 2.9992814 (M=   3) - s=0.0100\n",
      "   3 - L=-1001.9606324 - Gamma= 3.9982781 (M=   4) - s=0.0100\n",
      "   4 - L=-1001.9554971 - Gamma= 3.9987483 (M=   4) - s=0.0100\n",
      "   5 - L=-1001.9533744 - Gamma= 3.9987314 (M=   4) - s=0.0100\n",
      "   6 - L=-1001.9528565 - Gamma= 3.9987039 (M=   4) - s=0.0100\n",
      "   7 - L=-1001.9528565 - Gamma= 3.9987039 (M=   4) - s=0.0100\n",
      "Stopping at iteration 7 - max_delta_ml=1.4813017004921799e-07\n",
      "L=-1001.9528565240158 - Gamma=3.998703910534795 (M=4) - s=0.01\n",
      "Initial alpha = [[0.1160043]]\n",
      "   1 - L=-920.3977204 - Gamma= 1.9999422 (M=   2) - s=0.0100\n",
      "   2 - L=-894.5231734 - Gamma= 2.9994894 (M=   3) - s=0.0100\n",
      "   3 - L=-870.6786157 - Gamma= 3.9988634 (M=   4) - s=0.0100\n",
      "   4 - L=-870.6726422 - Gamma= 3.9988156 (M=   4) - s=0.0100\n",
      "   5 - L=-870.6667971 - Gamma= 3.9991544 (M=   4) - s=0.0100\n",
      "   6 - L=-870.6666188 - Gamma= 3.9991459 (M=   4) - s=0.0100\n",
      "   7 - L=-870.6666188 - Gamma= 3.9991459 (M=   4) - s=0.0100\n",
      "Stopping at iteration 7 - max_delta_ml=6.111536944320437e-08\n",
      "L=-870.6666188469153 - Gamma=3.999145878568824 (M=4) - s=0.01\n",
      "Initial alpha = [[0.11626696]]\n",
      "   1 - L=-964.6947399 - Gamma= 1.9999337 (M=   2) - s=0.0100\n",
      "   2 - L=-948.9595886 - Gamma= 2.9991922 (M=   3) - s=0.0100\n",
      "   3 - L=-919.9603358 - Gamma= 3.9986583 (M=   4) - s=0.0100\n",
      "   4 - L=-919.9535803 - Gamma= 3.9991788 (M=   4) - s=0.0100\n",
      "   5 - L=-919.9470966 - Gamma= 3.9991148 (M=   4) - s=0.0100\n",
      "   6 - L=-919.9470897 - Gamma= 3.9991131 (M=   4) - s=0.0100\n",
      "   7 - L=-919.9470897 - Gamma= 3.9991131 (M=   4) - s=0.0100\n",
      "Stopping at iteration 7 - max_delta_ml=5.2691283720864186e-08\n",
      "L=-919.9470896657706 - Gamma=3.9991130664467924 (M=4) - s=0.01\n",
      "Initial alpha = [[0.09859078]]\n",
      "   1 - L=-923.4016149 - Gamma= 1.9998721 (M=   2) - s=0.0100\n",
      "   2 - L=-898.8310245 - Gamma= 2.9993975 (M=   3) - s=0.0100\n",
      "   3 - L=-842.2782415 - Gamma= 3.9990704 (M=   4) - s=0.0100\n",
      "   4 - L=-842.2694660 - Gamma= 3.9994300 (M=   4) - s=0.0100\n",
      "   5 - L=-842.2670257 - Gamma= 3.9992729 (M=   4) - s=0.0100\n",
      "   6 - L=-842.2657739 - Gamma= 3.9992632 (M=   4) - s=0.0100\n",
      "   7 - L=-842.2657739 - Gamma= 3.9992632 (M=   4) - s=0.0100\n",
      "Stopping at iteration 7 - max_delta_ml=4.3932283736727255e-08\n",
      "L=-842.2657738882813 - Gamma=3.999263189384271 (M=4) - s=0.01\n",
      "Initial alpha = [[0.12139625]]\n",
      "   1 - L=-1090.6318254 - Gamma= 1.9999358 (M=   2) - s=0.0100\n",
      "   2 - L=-1064.3226397 - Gamma= 2.9995043 (M=   3) - s=0.0100\n",
      "   3 - L=-1039.2145261 - Gamma= 3.9989654 (M=   4) - s=0.0100\n",
      "   4 - L=-1039.2109781 - Gamma= 3.9992212 (M=   4) - s=0.0100\n",
      "   5 - L=-1039.2075004 - Gamma= 3.9991946 (M=   4) - s=0.0100\n",
      "   6 - L=-1039.2073387 - Gamma= 3.9991839 (M=   4) - s=0.0100\n",
      "   7 - L=-1039.2073387 - Gamma= 3.9991839 (M=   4) - s=0.0100\n",
      "Stopping at iteration 7 - max_delta_ml=2.7264867829483556e-08\n",
      "L=-1039.207338737163 - Gamma=3.999183903711974 (M=4) - s=0.01\n",
      "MODEL: RVM accuracy:  0.7037037037037039 +/-: 0.005811198403790997\n",
      "Initial alpha = [[0.09232197]]\n",
      "   1 - L=-995.7529945 - Gamma= 1.9999446 (M=   2) - s=0.0100\n",
      "   2 - L=-981.0103980 - Gamma= 2.9993190 (M=   3) - s=0.0100\n",
      "   3 - L=-946.6939503 - Gamma= 3.9988946 (M=   4) - s=0.0100\n",
      "   4 - L=-946.6856279 - Gamma= 3.9994066 (M=   4) - s=0.0100\n",
      "   5 - L=-946.6832583 - Gamma= 3.9993893 (M=   4) - s=0.0100\n",
      "   6 - L=-946.6826847 - Gamma= 3.9993648 (M=   4) - s=0.0100\n",
      "   7 - L=-946.6826847 - Gamma= 3.9993648 (M=   4) - s=0.0100\n",
      "Stopping at iteration 7 - max_delta_ml=6.593646949905373e-08\n",
      "L=-946.6826846532525 - Gamma=3.999364835464323 (M=4) - s=0.01\n",
      "++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      "+ + + + + + + + + + + + + + + + + + + + + + + + + + + + + +  Creating X,y\n",
      "++++++++++++++++++++++++++++++  RESULTS FOR CLASSIFICATION WITH GENOMIC DATA ++++++++++++++++++++++++++++++\n",
      "++++++++++++++++++++ ..processing feature array (54, 4) and class vector (54,)\n",
      "Train on 43 samples, validate on 11 samples\n",
      "Epoch 1/5\n",
      "43/43 [==============================] - 0s - loss: 0.6632 - acc: 0.5116 - val_loss: 0.4755 - val_acc: 0.4545\n",
      "Epoch 2/5\n",
      "43/43 [==============================] - 0s - loss: 0.5441 - acc: 0.7209 - val_loss: 0.3811 - val_acc: 0.9091\n",
      "Epoch 3/5\n",
      "43/43 [==============================] - 0s - loss: 0.5166 - acc: 0.8605 - val_loss: 0.2455 - val_acc: 1.0000\n",
      "Epoch 4/5\n",
      "43/43 [==============================] - 0s - loss: 0.3346 - acc: 0.8372 - val_loss: 0.1066 - val_acc: 1.0000\n",
      "Epoch 5/5\n",
      "43/43 [==============================] - 0s - loss: 0.2942 - acc: 0.8605 - val_loss: 0.1466 - val_acc: 1.0000\n",
      "Train on 43 samples, validate on 11 samples\n",
      "Epoch 1/5\n",
      "43/43 [==============================] - 0s - loss: 0.2062 - acc: 0.9302 - val_loss: 0.3167 - val_acc: 0.8182\n",
      "Epoch 2/5\n",
      "43/43 [==============================] - 0s - loss: 0.3006 - acc: 0.9302 - val_loss: 0.2693 - val_acc: 0.9091\n",
      "Epoch 3/5\n",
      "43/43 [==============================] - 0s - loss: 0.1933 - acc: 0.9302 - val_loss: 0.2197 - val_acc: 0.9091\n",
      "Epoch 4/5\n",
      "43/43 [==============================] - 0s - loss: 0.2137 - acc: 0.9302 - val_loss: 0.2168 - val_acc: 0.9091\n",
      "Epoch 5/5\n",
      "43/43 [==============================] - 0s - loss: 0.2136 - acc: 0.9302 - val_loss: 0.1891 - val_acc: 0.9091\n",
      "Train on 43 samples, validate on 11 samples\n",
      "Epoch 1/5\n",
      "43/43 [==============================] - 0s - loss: 0.3967 - acc: 0.8837 - val_loss: 0.1154 - val_acc: 0.9091\n",
      "Epoch 2/5\n",
      "43/43 [==============================] - 0s - loss: 0.2189 - acc: 0.9302 - val_loss: 0.1250 - val_acc: 0.9091\n",
      "Epoch 3/5\n",
      "43/43 [==============================] - 0s - loss: 0.3212 - acc: 0.9070 - val_loss: 0.1261 - val_acc: 0.9091\n",
      "Epoch 4/5\n",
      "43/43 [==============================] - 0s - loss: 0.2076 - acc: 0.9302 - val_loss: 0.1012 - val_acc: 1.0000\n",
      "Epoch 5/5\n",
      "43/43 [==============================] - 0s - loss: 0.2286 - acc: 0.9302 - val_loss: 0.1140 - val_acc: 0.9091\n",
      "Train on 43 samples, validate on 11 samples\n",
      "Epoch 1/5\n",
      "43/43 [==============================] - 0s - loss: 0.1040 - acc: 0.9767 - val_loss: 0.3339 - val_acc: 0.9091\n",
      "Epoch 2/5\n",
      "43/43 [==============================] - 0s - loss: 0.1026 - acc: 0.9767 - val_loss: 0.3822 - val_acc: 0.9091\n",
      "Epoch 3/5\n",
      "43/43 [==============================] - 0s - loss: 0.0974 - acc: 0.9535 - val_loss: 0.3939 - val_acc: 0.9091\n",
      "Epoch 4/5\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43/43 [==============================] - 0s - loss: 0.0780 - acc: 0.9767 - val_loss: 0.9136 - val_acc: 0.7273\n",
      "Epoch 5/5\n",
      "43/43 [==============================] - 0s - loss: 0.0700 - acc: 0.9767 - val_loss: 0.7547 - val_acc: 0.7273\n",
      "Train on 44 samples, validate on 10 samples\n",
      "Epoch 1/5\n",
      "44/44 [==============================] - 0s - loss: 0.3158 - acc: 0.9091 - val_loss: 0.2929 - val_acc: 0.8000\n",
      "Epoch 2/5\n",
      "44/44 [==============================] - 0s - loss: 0.1826 - acc: 0.9545 - val_loss: 0.1521 - val_acc: 0.9000\n",
      "Epoch 3/5\n",
      "44/44 [==============================] - 0s - loss: 0.1872 - acc: 0.9318 - val_loss: 0.1576 - val_acc: 0.9000\n",
      "Epoch 4/5\n",
      "44/44 [==============================] - 0s - loss: 0.1426 - acc: 0.9318 - val_loss: 0.1749 - val_acc: 0.9000\n",
      "Epoch 5/5\n",
      "44/44 [==============================] - 0s - loss: 0.1039 - acc: 0.9773 - val_loss: 0.1850 - val_acc: 0.9000\n",
      "MODEL: DNN accuracy:  0.8888888888888888 +/-: 0.008024691358024687\n",
      "++++++++++++++++++++++++++++++++++++++++++++++++++\n"
     ]
    }
   ],
   "source": [
    "Rocket.X_GENOME = None\n",
    "Rocket.Y_CLASS = None\n",
    "Rocket.PREP_HASH = None\n",
    "\n",
    "RUNS, MODELS, ACC = Rocket.run_classification(method_list = METHOD_LIST, \n",
    "                          num_run = nruns,\n",
    "                          pipeline = {\"scaler\": {\"type\": SCALER},\n",
    "                                      \"pre_processing\": PREPROC_DICT,\n",
    "                                      \"feature_selection\": FSELECT_DICT, # mannwhitney         \n",
    "                                      \"dim_reduction\": DIMRED_DICT},\n",
    "                          parameters = {}, \n",
    "                          features = 'genomic')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy on average: 0.8662551440329218 +- 0.004242008562996215, median: 0.8703703703703702+-0.001883027808953732\n",
      "++++++++++++++++++++++++++++++++++++++++\n",
      "        acc model       var\n",
      "0  0.870370    RF  0.001883\n",
      "1  0.907407   XGB  0.010113\n",
      "2  0.944444  LGBM  0.005499\n",
      "3  0.888889    ET  0.001291\n",
      "4  0.851852   SVM  0.001789\n",
      "5  0.870370    LR  0.001883\n",
      "6  0.870370  MLNN  0.001883\n",
      "7  0.703704   RVM  0.005811\n",
      "8  0.888889   DNN  0.008025\n"
     ]
    }
   ],
   "source": [
    "print(\"Accuracy on average: {} +- {}, median: {}+-{}\".format(ACC.mean()[0], ACC.mean()[1], ACC.median()[0], ACC.median()[1]))\n",
    "print(\"+\"*40)\n",
    "print(ACC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "###########\n",
    "###########\n",
    "##Runs.append(AllResults)\n",
    "final_df = pandas.DataFrame()\n",
    "\n",
    "for idx, df in enumerate(RUNS):    \n",
    "    df['run'] = idx\n",
    "    final_df = final_df.append(df, ignore_index = True)\n",
    "final_df[Rocket.MODEL_PARAMETERS['ID']] = final_df[Rocket.MODEL_PARAMETERS['ID']].astype(str)\n",
    "final_df = final_df.sort_values(by=Rocket.MODEL_PARAMETERS['ID'])\n",
    "final_df['pred']= pandas.to_numeric(final_df['pred'])\n",
    "final_df_agg = final_df.groupby([Rocket.MODEL_PARAMETERS['ID'], 'method']).agg({'pred': [numpy.mean, numpy.median, numpy.std]})\n",
    "final_df_agg = final_df_agg['pred'].groupby(by=Rocket.MODEL_PARAMETERS['ID']).agg({'mean': [numpy.mean, numpy.median, numpy.std]})['mean']\n",
    "final_df.to_csv(\"out/patient_results_\"+Rocket.SET_NAME+\"_FDR0025.csv\")\n",
    "final_df_agg.to_csv(\"out/patient_results_agg_\"+Rocket.SET_NAME+\"_FDR0025.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Top genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "top_weights, top_coeffs = _helpers.get_top_genes(MODELS=MODELS, n_max=5000, RexR=Rocket)\n",
    "#top_coeffs.to_csv(\"out/coeffs_\"+Rocket.SET_NAME+\"_FDR_alpha005.csv\")\n",
    "#top_weights.to_csv(\"out/weights_\"+Rocket.SET_NAME+\"_FDR_alpha0005.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0_RF</th>\n",
       "      <th>1_XGB</th>\n",
       "      <th>2_LGBM</th>\n",
       "      <th>3_ExtraTrees</th>\n",
       "      <th>MEAN</th>\n",
       "      <th>MEDIAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1569566_at</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.955882</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225791_at</th>\n",
       "      <td>0.567469</td>\n",
       "      <td>0.849057</td>\n",
       "      <td>0.764706</td>\n",
       "      <td>0.497780</td>\n",
       "      <td>0.669753</td>\n",
       "      <td>0.669753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230932_at</th>\n",
       "      <td>0.418336</td>\n",
       "      <td>0.344340</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.576257</td>\n",
       "      <td>0.584733</td>\n",
       "      <td>0.576257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227134_at</th>\n",
       "      <td>0.451434</td>\n",
       "      <td>0.551887</td>\n",
       "      <td>0.352941</td>\n",
       "      <td>0.537645</td>\n",
       "      <td>0.473477</td>\n",
       "      <td>0.473477</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                0_RF     1_XGB    2_LGBM  3_ExtraTrees      MEAN    MEDIAN\n",
       "1569566_at  1.000000  1.000000  0.823529      1.000000  0.955882  1.000000\n",
       "225791_at   0.567469  0.849057  0.764706      0.497780  0.669753  0.669753\n",
       "230932_at   0.418336  0.344340  1.000000      0.576257  0.584733  0.576257\n",
       "227134_at   0.451434  0.551887  0.352941      0.537645  0.473477  0.473477"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>8_SVM</th>\n",
       "      <th>8_LR</th>\n",
       "      <th>MEAN</th>\n",
       "      <th>MEDIAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1569566_at</th>\n",
       "      <td>-0.935915</td>\n",
       "      <td>-0.894093</td>\n",
       "      <td>-0.915004</td>\n",
       "      <td>-0.915004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1569566_at</th>\n",
       "      <td>-0.935915</td>\n",
       "      <td>-0.894093</td>\n",
       "      <td>-0.915004</td>\n",
       "      <td>-0.915004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230932_at</th>\n",
       "      <td>0.409112</td>\n",
       "      <td>0.388079</td>\n",
       "      <td>0.398596</td>\n",
       "      <td>0.398596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230932_at</th>\n",
       "      <td>0.409112</td>\n",
       "      <td>0.388079</td>\n",
       "      <td>0.398596</td>\n",
       "      <td>0.398596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225791_at</th>\n",
       "      <td>0.717438</td>\n",
       "      <td>0.758431</td>\n",
       "      <td>0.737934</td>\n",
       "      <td>0.737934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225791_at</th>\n",
       "      <td>0.717438</td>\n",
       "      <td>0.758431</td>\n",
       "      <td>0.737934</td>\n",
       "      <td>0.737934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227134_at</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227134_at</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               8_SVM      8_LR      MEAN    MEDIAN\n",
       "1569566_at -0.935915 -0.894093 -0.915004 -0.915004\n",
       "1569566_at -0.935915 -0.894093 -0.915004 -0.915004\n",
       "230932_at   0.409112  0.388079  0.398596  0.398596\n",
       "230932_at   0.409112  0.388079  0.398596  0.398596\n",
       "225791_at   0.717438  0.758431  0.737934  0.737934\n",
       "225791_at   0.717438  0.758431  0.737934  0.737934\n",
       "227134_at   1.000000  1.000000  1.000000  1.000000\n",
       "227134_at   1.000000  1.000000  1.000000  1.000000"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_coeffs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Connecting probesets to genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_helpers._probeset_mapper took 2.82 seconds to finish\n"
     ]
    }
   ],
   "source": [
    "probe_list = top_weights.index.tolist()\n",
    "probeset_col = 'Probe Set ID'\n",
    "desc_list = ['Pathway','Gene Title', 'Gene Symbol', 'Chromosomal Location', \n",
    "             'EC', 'gene ontology biological process', 'gene ontology molecular function']\n",
    "probeset_to_genome_map = _helpers._probeset_mapper(probeset_type = 'HG-U133',                                                    \n",
    "                                                   mapping_file = None,\n",
    "                                                   probeset_col = probeset_col,\n",
    "                                                   description_list = desc_list,\n",
    "                                                   probe_list = probe_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1569566_at': {'AGI': '---',\n",
       "  'Alignments': 'chr4:37892744-37954428 (+) // 68.11 // p14',\n",
       "  'Annotation Date': '30-Mar-16',\n",
       "  'Annotation Description': 'This probe set was annotated using the Matching Probes based pipeline to a Entrez Gene identifier using 2 transcripts. // false // Matching Probes // A',\n",
       "  'Annotation Notes': 'XR_945684 // refseq // 5 // Cross Hyb Matching Probes /// XR_945685 // refseq // 5 // Cross Hyb Matching Probes /// XR_925212 // refseq // 3 // Cross Hyb Matching Probes /// XR_925213 // refseq // 3 // Cross Hyb Matching Probes',\n",
       "  'Annotation Transcript Cluster': 'ENST00000402522(9),OTTHUMT00000317444(9)',\n",
       "  'Archival UniGene Cluster': 'Hs.203722',\n",
       "  'Chromosomal Location': 'chr4p14',\n",
       "  'EC': '---',\n",
       "  'Ensembl': 'ENSG00000065882 /// OTTHUMG00000150302',\n",
       "  'Entrez Gene': '23216',\n",
       "  'FlyBase': '---',\n",
       "  'Gene Ontology Biological Process': '0032880 // regulation of protein localization // inferred from electronic annotation /// 0043547 // positive regulation of GTPase activity // inferred from electronic annotation /// 0061024 // membrane organization // traceable author statement /// 1902017 // regulation of cilium assembly // inferred from mutant phenotype',\n",
       "  'Gene Ontology Cellular Component': '0005634 // nucleus // inferred from electronic annotation /// 0005829 // cytosol // traceable author statement',\n",
       "  'Gene Ontology Molecular Function': '0005096 // GTPase activator activity // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction',\n",
       "  'Gene Symbol': 'TBC1D1',\n",
       "  'Gene Title': 'TBC1 (tre-2/USP6, BUB2, cdc16) domain family, member 1',\n",
       "  'GeneChip Array': 'Human Genome U133 Plus 2.0 Array',\n",
       "  'Genome Version': 'February 2009 (Genome Reference Consortium GRCh37)',\n",
       "  'InterPro': '---',\n",
       "  'MGI Name': '---',\n",
       "  'OMIM': '609850',\n",
       "  'Pathway': '---',\n",
       "  'Probe Set ID': '1569566_at',\n",
       "  'QTL': '---',\n",
       "  'RGD Name': '---',\n",
       "  'RefSeq Protein ID': 'NP_001240841 /// NP_001240842 /// NP_001240843 /// NP_001240844 /// NP_055988 /// XP_005262703 /// XP_005262704 /// XP_005262706 /// XP_006714064 /// XP_011511961 /// XP_011511962 /// XP_011511963 /// XP_011511964 /// XP_011511965 /// XP_011511966 /// XP_011511967 /// XP_011511968 /// XP_011511969 /// XP_011511970 /// XP_011511971 /// XP_011511972',\n",
       "  'RefSeq Transcript ID': 'NM_001253912 /// NM_001253913 /// NM_001253914 /// NM_001253915 /// NM_015173 /// XM_005262646 /// XM_005262647 /// XM_005262649 /// XM_006714001 /// XM_011513659 /// XM_011513660 /// XM_011513661 /// XM_011513662 /// XM_011513663 /// XM_011513664 /// XM_011513665 /// XM_011513666 /// XM_011513667 /// XM_011513668 /// XM_011513669 /// XM_011513670',\n",
       "  'Representative Public ID': 'BC028196',\n",
       "  'SGD accession number': '---',\n",
       "  'Sequence Source': 'GenBank',\n",
       "  'Sequence Type': 'Consensus sequence',\n",
       "  'Species Scientific Name': 'Homo sapiens',\n",
       "  'SwissProt': 'Q86TI0 /// Q8NC59 /// B9A6J6 /// Q6PJJ8 /// Q7Z6H2 /// B4E0Y2 /// B5MCJ2',\n",
       "  'Target Description': 'gb:BC028196.1 /DB_XREF=gi:20380158 /TID=Hs2.203722.2 /CNT=3 /FEA=mRNA /TIER=ConsEnd /STK=0 /UG=Hs.203722 /UG_TITLE=Homo sapiens, clone IMAGE:5176169, mRNA /DEF=Homo sapiens, similar to tre-2USP6, BUB2, cdc16) domain family, member 1; tbc1, clone IMAGE:5211948, mRNA.',\n",
       "  'Trans Membrane': '---',\n",
       "  'Transcript Assignments': 'ENST00000402522 // havana:known chromosome:GRCh38:4:37891099:37952812:1 gene:ENSG00000065882 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 9 // --- /// OTTHUMT00000317444 // otter:putative chromosome:VEGA61:4:37891099:37952812:1 gene:OTTHUMG00000150302 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 9 // ---',\n",
       "  'Transcript ID(Array Design)': 'Hs2.203722.2',\n",
       "  'UniGene ID': '---',\n",
       "  'Unigene Cluster Type': '---',\n",
       "  'WormBase': '---'},\n",
       " '225791_at': {'AGI': '---',\n",
       "  'Alignments': 'chr2:238875684-238951236 (+) // 92.51 // q37.3',\n",
       "  'Annotation Date': '30-Mar-16',\n",
       "  'Annotation Description': 'This probe set was annotated using the Matching Probes based pipeline to a Entrez Gene identifier using 16 transcripts. // false // Matching Probes // A',\n",
       "  'Annotation Notes': 'AK290063 // gb // 5 // Cross Hyb Matching Probes /// AK294107 // gb // 8 // Cross Hyb Matching Probes /// AK297502 // gb // 8 // Cross Hyb Matching Probes /// AK303094 // gb // 4 // Cross Hyb Matching Probes',\n",
       "  'Annotation Transcript Cluster': 'ENST00000272930(11),ENST00000414443(11),ENST00000612130(11),NM_001278305(11),NM_001278306(11),NM_001278307(11),NM_001278308(11),NM_080678(11),NR_103498(11),NR_103499(11),NR_103500(11),OTTHUMT00000257171(11),uc002vxk.3,uc010znn.2,uc010zno.2,uc010znp.2',\n",
       "  'Archival UniGene Cluster': 'Hs.289051',\n",
       "  'Chromosomal Location': 'chr2q37.3',\n",
       "  'EC': 'EC:6.3.2.-',\n",
       "  'Ensembl': 'ENSG00000184182 /// OTTHUMG00000133341',\n",
       "  'Entrez Gene': '140739',\n",
       "  'FlyBase': '---',\n",
       "  'Gene Ontology Biological Process': '0016567 // protein ubiquitination // not recorded /// 0045116 // protein neddylation // inferred from direct assay /// 0045116 // protein neddylation // inferred from electronic annotation',\n",
       "  'Gene Ontology Cellular Component': '0005737 // cytoplasm // not recorded',\n",
       "  'Gene Ontology Molecular Function': '0000166 // nucleotide binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016874 // ligase activity // inferred from electronic annotation /// 0019788 // NEDD8 transferase activity // inferred from direct assay /// 0031625 // ubiquitin protein ligase binding // not recorded /// 0061630 // ubiquitin protein ligase activity // not recorded',\n",
       "  'Gene Symbol': 'UBE2F',\n",
       "  'Gene Title': 'ubiquitin-conjugating enzyme E2F (putative)',\n",
       "  'GeneChip Array': 'Human Genome U133 Plus 2.0 Array',\n",
       "  'Genome Version': 'February 2009 (Genome Reference Consortium GRCh37)',\n",
       "  'InterPro': 'IPR000608 // Ubiquitin-conjugating enzyme E2 // 1.1E-30 /// IPR000608 // Ubiquitin-conjugating enzyme E2 // 2.5E-27 /// IPR000608 // Ubiquitin-conjugating enzyme E2 // 1.6E-34',\n",
       "  'MGI Name': '---',\n",
       "  'OMIM': '---',\n",
       "  'Pathway': '---',\n",
       "  'Probe Set ID': '225791_at',\n",
       "  'QTL': '---',\n",
       "  'RGD Name': '---',\n",
       "  'RefSeq Protein ID': 'NP_001265234 /// NP_001265235 /// NP_001265236 /// NP_001265237 /// NP_542409',\n",
       "  'RefSeq Transcript ID': 'NM_001278305 /// NM_001278306 /// NM_001278307 /// NM_001278308 /// NM_080678 /// NR_103498 /// NR_103499 /// NR_103500',\n",
       "  'Representative Public ID': 'AI040343',\n",
       "  'SGD accession number': '---',\n",
       "  'Sequence Source': 'GenBank',\n",
       "  'Sequence Type': 'Consensus sequence',\n",
       "  'Species Scientific Name': 'Homo sapiens',\n",
       "  'SwissProt': 'Q969M7',\n",
       "  'Target Description': 'gb:AI040343 /DB_XREF=gi:3279537 /DB_XREF=oy33c08.x1 /CLONE=IMAGE:1667630 /FEA=EST /CNT=94 /TID=Hs.289051.0 /TIER=Stack /STK=10 /UG=Hs.289051 /UG_TITLE=Homo sapiens cDNA FLJ13857 fis, clone THYRO1001003, weakly similar to UBIQUITIN-CONJUGATING ENZYME E2-21.2 KD (EC 6.3.2.19)',\n",
       "  'Trans Membrane': '---',\n",
       "  'Transcript Assignments': 'ENST00000272930 // ensembl_havana_transcript:known chromosome:GRCh38:2:237966955:238042595:1 gene:ENSG00000184182 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// ENST00000414443 // ensembl:known chromosome:GRCh38:2:237966955:238042782:1 gene:ENSG00000184182 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// ENST00000612130 // ensembl:known chromosome:GRCh38:2:237968793:238042782:1 gene:ENSG00000184182 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// NM_001278305 // Homo sapiens ubiquitin-conjugating enzyme E2F (putative) (UBE2F), transcript variant 2, mRNA. // refseq // 11 // --- /// NM_001278306 // Homo sapiens ubiquitin-conjugating enzyme E2F (putative) (UBE2F), transcript variant 3, mRNA. // refseq // 11 // --- /// NM_001278307 // Homo sapiens ubiquitin-conjugating enzyme E2F (putative) (UBE2F), transcript variant 4, mRNA. // refseq // 11 // --- /// NM_001278308 // Homo sapiens ubiquitin-conjugating enzyme E2F (putative) (UBE2F), transcript variant 5, mRNA. // refseq // 11 // --- /// NM_080678 // Homo sapiens ubiquitin-conjugating enzyme E2F (putative) (UBE2F), transcript variant 1, mRNA. // refseq // 11 // --- /// NR_103498 // Homo sapiens ubiquitin-conjugating enzyme E2F (putative) (UBE2F), transcript variant 6, non-coding RNA. // refseq // 11 // --- /// NR_103499 // Homo sapiens ubiquitin-conjugating enzyme E2F (putative) (UBE2F), transcript variant 7, non-coding RNA. // refseq // 11 // --- /// NR_103500 // Homo sapiens ubiquitin-conjugating enzyme E2F (putative) (UBE2F), transcript variant 8, non-coding RNA. // refseq // 11 // --- /// OTTHUMT00000257171 // otter:known chromosome:VEGA61:2:237966955:238042595:1 gene:OTTHUMG00000133341 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 11 // --- /// uc002vxk.3 // --- // ucsc_genes // 11 // --- /// uc010znn.2 // --- // ucsc_genes // 11 // --- /// uc010zno.2 // --- // ucsc_genes // 11 // --- /// uc010znp.2 // --- // ucsc_genes // 11 // ---',\n",
       "  'Transcript ID(Array Design)': 'Hs.289051.0',\n",
       "  'UniGene ID': 'Hs.471785',\n",
       "  'Unigene Cluster Type': 'full length',\n",
       "  'WormBase': '---'},\n",
       " '227134_at': {'AGI': '---',\n",
       "  'Alignments': 'chr1:27676797-27680421 (+) // 92.55 // p36.11',\n",
       "  'Annotation Date': '30-Mar-16',\n",
       "  'Annotation Description': 'This probe set was annotated using the Matching Probes based pipeline to a Entrez Gene identifier using 33 transcripts. // false // Matching Probes // A',\n",
       "  'Annotation Notes': '---',\n",
       "  'Annotation Transcript Cluster': 'AK074154(9),AK096437(11),AK126120(9),AK293038(10),BC015764(11),BC035725(11),ENST00000318074(11),ENST00000475199(9),ENST00000483926(11),ENST00000490170(11),ENST00000543823(11),ENST00000615284(11),ENST00000616558(10),ENST00000618673(10),NM_001193308(11),NM_032872(11),OTTHUMT00000015778(11),OTTHUMT00000015779(9),OTTHUMT00000015782(11),OTTHUMT00000015783(11),OTTHUMT00000476525(10),OTTHUMT00000476526(10),OTTHUMT00000476527(11),OTTHUMT00000476528(11),uc001bnv.2,uc001bnw.2,uc009vsv.2,XM_005246022(11),XM_006710990(11),XM_006710991(11),XM_011542316(11),XM_011542317(11),XR_946784(11)',\n",
       "  'Archival UniGene Cluster': 'Hs.25895',\n",
       "  'Chromosomal Location': 'chr1p36.11',\n",
       "  'EC': '---',\n",
       "  'Ensembl': 'ENSG00000142765 /// OTTHUMG00000005770',\n",
       "  'Entrez Gene': '84958',\n",
       "  'FlyBase': '---',\n",
       "  'Gene Ontology Biological Process': '0006886 // intracellular protein transport // inferred from electronic annotation /// 0006887 // exocytosis // inferred from direct assay /// 0006906 // vesicle fusion // not recorded /// 0016079 // synaptic vesicle exocytosis // not recorded /// 0017158 // regulation of calcium ion-dependent exocytosis // not recorded /// 0048791 // calcium ion-dependent exocytosis of neurotransmitter // not recorded',\n",
       "  'Gene Ontology Cellular Component': '0005622 // intracellular // inferred from electronic annotation /// 0005623 // cell // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0019897 // extrinsic component of plasma membrane // inferred from sequence or structural similarity /// 0042470 // melanosome // inferred from sequence or structural similarity /// 0070062 // extracellular exosome // inferred from direct assay /// 0070382 // exocytic vesicle // not recorded',\n",
       "  'Gene Ontology Molecular Function': '0005509 // calcium ion binding // not recorded /// 0005515 // protein binding // inferred from physical interaction /// 0005544 // calcium-dependent phospholipid binding // not recorded /// 0017137 // Rab GTPase binding // inferred from electronic annotation /// 0019905 // syntaxin binding // not recorded /// 0030276 // clathrin binding // not recorded /// 0042043 // neurexin family protein binding // inferred from sequence or structural similarity',\n",
       "  'Gene Symbol': 'SYTL1',\n",
       "  'Gene Title': 'synaptotagmin like 1',\n",
       "  'GeneChip Array': 'Human Genome U133 Plus 2.0 Array',\n",
       "  'Genome Version': 'February 2009 (Genome Reference Consortium GRCh37)',\n",
       "  'InterPro': 'IPR000008 // C2 domain // 5.0E-35 /// IPR000008 // C2 domain // 5.0E-35 /// IPR000008 // C2 domain // 5.4E-35 /// IPR000008 // C2 domain // 5.4E-35 /// IPR000008 // C2 domain // 1.3E-28 /// IPR000008 // C2 domain // 1.3E-28 /// IPR000008 // C2 domain // 4.6E-18 /// IPR000008 // C2 domain // 3.3E-16 /// IPR000008 // C2 domain // 1.5E-24 /// IPR000008 // C2 domain // 3.8E-16 /// IPR000008 // C2 domain // 1.5E-24 /// IPR000008 // C2 domain // 4.0E-16',\n",
       "  'MGI Name': '---',\n",
       "  'OMIM': '608042',\n",
       "  'Pathway': '---',\n",
       "  'Probe Set ID': '227134_at',\n",
       "  'QTL': '---',\n",
       "  'RGD Name': '---',\n",
       "  'RefSeq Protein ID': 'NP_001180237 /// NP_116261 /// XP_005246079 /// XP_006711053 /// XP_006711054 /// XP_011540618 /// XP_011540619',\n",
       "  'RefSeq Transcript ID': 'NM_001193308 /// NM_032872 /// XM_005246022 /// XM_006710990 /// XM_006710991 /// XM_011542316 /// XM_011542317 /// XR_946784',\n",
       "  'Representative Public ID': 'AI341537',\n",
       "  'SGD accession number': '---',\n",
       "  'Sequence Source': 'GenBank',\n",
       "  'Sequence Type': 'Consensus sequence',\n",
       "  'Species Scientific Name': 'Homo sapiens',\n",
       "  'SwissProt': 'Q8IYJ3 /// Q8N880 /// A8KAH3 /// Q8TEH0 /// A0A087X1N4',\n",
       "  'Target Description': 'gb:AI341537 /DB_XREF=gi:4078464 /DB_XREF=qq94e02.x1 /CLONE=IMAGE:1939034 /FEA=EST /CNT=50 /TID=Hs.25895.0 /TIER=Stack /STK=42 /UG=Hs.25895 /UG_TITLE=ESTs, Weakly similar to PI-3 kinase (H.sapiens)',\n",
       "  'Trans Membrane': '---',\n",
       "  'Transcript Assignments': 'AK074154 // Homo sapiens mRNA for FLJ00227 protein. // gb // 9 // --- /// AK096437 // Homo sapiens cDNA FLJ39118 fis, clone NTONG2006484, highly similar to Mus musculus slp1 mRNA for synaptotagmin-like protein 1. // gb // 11 // --- /// AK126120 // Homo sapiens cDNA FLJ44132 fis, clone THYMU2008282. // gb // 9 // --- /// AK293038 // Homo sapiens cDNA FLJ75170 complete cds, highly similar to Homo sapiens synaptotagmin-like 1, mRNA. // gb // 10 // --- /// BC015764 // Homo sapiens synaptotagmin-like 1, mRNA (cDNA clone MGC:23187 IMAGE:4851468), complete cds. // gb // 11 // --- /// BC035725 // Homo sapiens synaptotagmin-like 1, mRNA (cDNA clone MGC:45995 IMAGE:5722683), complete cds. // gb // 11 // --- /// ENST00000318074 // havana:known chromosome:GRCh38:1:27342022:27353932:1 gene:ENSG00000142765 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// ENST00000475199 // havana:known chromosome:GRCh38:1:27351472:27353930:1 gene:ENSG00000142765 gene_biotype:protein_coding transcript_biotype:retained_intron // ensembl // 9 // --- /// ENST00000483926 // havana:known chromosome:GRCh38:1:27342057:27353930:1 gene:ENSG00000142765 gene_biotype:protein_coding transcript_biotype:retained_intron // ensembl // 11 // --- /// ENST00000490170 // havana:known chromosome:GRCh38:1:27347544:27353930:1 gene:ENSG00000142765 gene_biotype:protein_coding transcript_biotype:retained_intron // ensembl // 11 // --- /// ENST00000543823 // havana:known chromosome:GRCh38:1:27344873:27353930:1 gene:ENSG00000142765 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// ENST00000615284 // havana:known chromosome:GRCh38:1:27351075:27353937:1 gene:ENSG00000142765 gene_biotype:protein_coding transcript_biotype:retained_intron // ensembl // 11 // --- /// ENST00000616558 // ensembl_havana_transcript:known chromosome:GRCh38:1:27342041:27353893:1 gene:ENSG00000142765 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 10 // --- /// ENST00000618673 // havana:known chromosome:GRCh38:1:27342020:27353901:1 gene:ENSG00000142765 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 10 // --- /// GENSCAN00000042478 // cdna:genscan chromosome:GRCh38:1:27261055:27353852:1 transcript_biotype:protein_coding // ensembl // 10 // --- /// NM_001193308 // Homo sapiens synaptotagmin-like 1 (SYTL1), transcript variant 1, mRNA. // refseq // 11 // --- /// NM_032872 // Homo sapiens synaptotagmin-like 1 (SYTL1), transcript variant 2, mRNA. // refseq // 11 // --- /// OTTHUMT00000015777 // otter: chromosome:VEGA61:1:27345159:27353930:1 gene:OTTHUMG00000005770 gene_biotype:protein_coding transcript_biotype:artifact // vega // 9 // --- /// OTTHUMT00000015778 // otter: chromosome:VEGA61:1:27347544:27353930:1 gene:OTTHUMG00000005770 gene_biotype:protein_coding transcript_biotype:retained_intron // vega // 11 // --- /// OTTHUMT00000015779 // otter: chromosome:VEGA61:1:27351472:27353930:1 gene:OTTHUMG00000005770 gene_biotype:protein_coding transcript_biotype:retained_intron // vega // 9 // --- /// OTTHUMT00000015782 // otter: chromosome:VEGA61:1:27342057:27353930:1 gene:OTTHUMG00000005770 gene_biotype:protein_coding transcript_biotype:retained_intron // vega // 11 // --- /// OTTHUMT00000015783 // otter:known chromosome:VEGA61:1:27342022:27353932:1 gene:OTTHUMG00000005770 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 11 // --- /// OTTHUMT00000476525 // otter:novel chromosome:VEGA61:1:27342020:27353901:1 gene:OTTHUMG00000005770 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 10 // --- /// OTTHUMT00000476526 // otter:known chromosome:VEGA61:1:27342041:27353893:1 gene:OTTHUMG00000005770 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 10 // --- /// OTTHUMT00000476527 // otter:known chromosome:VEGA61:1:27344873:27353930:1 gene:OTTHUMG00000005770 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 11 // --- /// OTTHUMT00000476528 // otter: chromosome:VEGA61:1:27351075:27353937:1 gene:OTTHUMG00000005770 gene_biotype:protein_coding transcript_biotype:retained_intron // vega // 11 // --- /// uc001bnv.2 // --- // ucsc_genes // 11 // --- /// uc001bnw.2 // --- // ucsc_genes // 11 // --- /// uc009vsv.2 // --- // ucsc_genes // 11 // --- /// XM_005246022 // PREDICTED: Homo sapiens synaptotagmin-like 1 (SYTL1), transcript variant X1, mRNA. // refseq // 11 // --- /// XM_006710990 // PREDICTED: Homo sapiens synaptotagmin-like 1 (SYTL1), transcript variant X2, mRNA. // refseq // 11 // --- /// XM_006710991 // PREDICTED: Homo sapiens synaptotagmin-like 1 (SYTL1), transcript variant X4, mRNA. // refseq // 11 // --- /// XM_011542316 // PREDICTED: Homo sapiens synaptotagmin-like 1 (SYTL1), transcript variant X5, mRNA. // refseq // 11 // --- /// XM_011542317 // PREDICTED: Homo sapiens synaptotagmin-like 1 (SYTL1), transcript variant X6, mRNA. // refseq // 11 // --- /// XR_946784 // PREDICTED: Homo sapiens synaptotagmin-like 1 (SYTL1), transcript variant X3, misc_RNA. // refseq // 11 // ---',\n",
       "  'Transcript ID(Array Design)': 'Hs.25895.0',\n",
       "  'UniGene ID': 'Hs.469175',\n",
       "  'Unigene Cluster Type': 'full length',\n",
       "  'WormBase': '---'},\n",
       " '230932_at': {'AGI': '---',\n",
       "  'Alignments': 'chr5:179338447-179339093 (-) // 71.78 // q35.3',\n",
       "  'Annotation Date': '30-Mar-16',\n",
       "  'Annotation Description': 'This probe set was annotated using the Matching Probes based pipeline to a Entrez Gene identifier using 2 transcripts. // false // Matching Probes // A',\n",
       "  'Annotation Notes': '---',\n",
       "  'Annotation Transcript Cluster': 'ENST00000522208(11),OTTHUMT00000374206(11)',\n",
       "  'Archival UniGene Cluster': 'Hs.8038',\n",
       "  'Chromosomal Location': 'chr5q35.3',\n",
       "  'EC': '---',\n",
       "  'Ensembl': 'ENSG00000113269 /// OTTHUMG00000130909',\n",
       "  'Entrez Gene': '55819',\n",
       "  'FlyBase': '---',\n",
       "  'Gene Ontology Biological Process': '0006915 // apoptotic process // inferred from electronic annotation /// 0012501 // programmed cell death // inferred from sequence or structural similarity /// 0016567 // protein ubiquitination // inferred from electronic annotation',\n",
       "  'Gene Ontology Cellular Component': '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from direct assay /// 0005737 // cytoplasm // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation',\n",
       "  'Gene Ontology Molecular Function': '0004842 // ubiquitin-protein transferase activity // inferred from direct assay /// 0005515 // protein binding // inferred from electronic annotation /// 0008270 // zinc ion binding // inferred from electronic annotation /// 0016874 // ligase activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation',\n",
       "  'Gene Symbol': 'RNF130',\n",
       "  'Gene Title': 'ring finger protein 130',\n",
       "  'GeneChip Array': 'Human Genome U133 Plus 2.0 Array',\n",
       "  'Genome Version': 'February 2009 (Genome Reference Consortium GRCh37)',\n",
       "  'InterPro': 'IPR001841 // Zinc finger, RING-type // 5.7E-14 /// IPR003137 // PA domain // 7.3E-11 /// IPR003137 // PA domain // 5.8E-15 /// IPR018957 // Zinc finger, C3HC4 RING-type // 1.2E-10',\n",
       "  'MGI Name': '---',\n",
       "  'OMIM': '---',\n",
       "  'Pathway': '---',\n",
       "  'Probe Set ID': '230932_at',\n",
       "  'QTL': '---',\n",
       "  'RGD Name': '---',\n",
       "  'RefSeq Protein ID': 'NP_001267730 /// NP_060904 /// XP_011532895',\n",
       "  'RefSeq Transcript ID': 'NM_001280801 /// NM_018434 /// XM_011534593',\n",
       "  'Representative Public ID': 'R39126',\n",
       "  'SGD accession number': '---',\n",
       "  'Sequence Source': 'GenBank',\n",
       "  'Sequence Type': 'Consensus sequence',\n",
       "  'Species Scientific Name': 'Homo sapiens',\n",
       "  'SwissProt': 'Q86XS8 /// Q2HIY3 /// Q1WWL6 /// E5RI87',\n",
       "  'Target Description': 'gb:R39126 /DB_XREF=gi:796582 /DB_XREF=yf61d12.s1 /CLONE=IMAGE:26892 /FEA=EST /CNT=13 /TID=Hs.8038.0 /TIER=Stack /STK=10 /UG=Hs.8038 /UG_TITLE=ESTs',\n",
       "  'Trans Membrane': '---',\n",
       "  'Transcript Assignments': 'ENST00000522208 // havana:known chromosome:GRCh38:5:179911651:180071721:-1 gene:ENSG00000113269 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// OTTHUMT00000374206 // otter:putative chromosome:VEGA61:5:179911651:180071721:-1 gene:OTTHUMG00000130909 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 11 // ---',\n",
       "  'Transcript ID(Array Design)': 'Hs.8038.0',\n",
       "  'UniGene ID': '---',\n",
       "  'Unigene Cluster Type': '---',\n",
       "  'WormBase': '---'}}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "probeset_to_genome_map"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualise the top genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Patient clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from math import*\n",
    "#from scipy.dspatial.distance import cosine\n",
    "from scipy.spatial.distance import minkowski\n",
    "from scipy.spatial.distance import cdist\n",
    "from decimal import Decimal\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "from scipy import sparse\n",
    "from time import time\n",
    "from sklearn.cluster import AffinityPropagation\n",
    "import matplotlib.pyplot as plt\n",
    "from itertools import cycle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#TransPosed = Rocket.DATA_all_samples.T # all microarrays, may be multiple per patient versus all probesets, may be multiple per genome\n",
    "Normal = Rocket.DATA_merged_processed.loc[:, (Rocket.DATA_merged_processed.columns !='target') & \n",
    "                                             (Rocket.DATA_merged_processed.columns !='ID')]\n",
    "#AllNormal = Rocket.DATA_merged\n",
    "#probeset_weights = Rocket.get_probeset_weights(method = )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from sklearn.decomposition import PCA, LatentDirichletAllocation\n",
    "from sklearn.cross_decomposition import CCA\n",
    "from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Markov clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: 9827_corr2.CEL, dtype: float64)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient_sim = _helpers.patient_similarity(Normal, sim_type = 'cosine', normalised = True, inflation = 2, minkowski_dim=1)\n",
    "##### apply Markov clustering\n",
    "#######################\n",
    "# non-distributed, non-sparse version, only for small-sized problems (N is order 1000)\n",
    "e = 2\n",
    "r = 2 \n",
    "epsilon = 1e-7\n",
    "convergence = 0.001\n",
    "num_iter = 10\n",
    "Orientation = 'col' # columnwise or rowwise\n",
    "\n",
    "# add loop\n",
    "def add_loop(df_matrix, value=0): \n",
    "    for i in df_matrix.index:\n",
    "        df_matrix.loc[i, i] = value\n",
    "    return df_matrix\n",
    "patient_sim = add_loop(patient_sim, 1)\n",
    "patient_sim = patient_sim - epsilon\n",
    "\n",
    "def normalise(sim, type = 'col'):\n",
    "    if(type == 'col'):\n",
    "        # column normalisation\n",
    "        for variable in sim.keys():\n",
    "            col_vec = sim[variable]\n",
    "            sum_val = sum([p for p in col_vec])\n",
    "            sim[variable] = sim[variable]/sum_val\n",
    "    elif (type == 'row'):\n",
    "        # row normalisation\n",
    "        for variable in sim.keys():\n",
    "            row_vec = sim.loc[variable, :]\n",
    "            sum_val = sum([p for p in row_vec])\n",
    "            sim.loc[variable,:] = sim.loc[variable,:]/sum_val\n",
    "    return sim\n",
    "\n",
    "# step E: expansion, get the nth power of the matrix\n",
    "def expansion(sim):\n",
    "    X = numpy.array(sim)\n",
    "    VarList = sim.keys()\n",
    "    if e == 1:\n",
    "        return sim\n",
    "    elif e > 1:        \n",
    "        return pandas.DataFrame(numpy.linalg.matrix_power(X, e), index = VarList, columns = VarList)\n",
    "     \n",
    "# step I: inflation, per column raise by rth power and column normalise\n",
    "def inflation(sim, type = 'col'):    \n",
    "    if type == 'col':\n",
    "        Axis = 0\n",
    "    elif type == 'row':\n",
    "        Axis = 1\n",
    "    return sim.apply(lambda x: x**r/sum(x**r), axis = Axis)\n",
    "\n",
    "# remove weak connections, values < epsilon\n",
    "def clean(sim):\n",
    "    return sim.applymap(lambda x:0 if x<epsilon else x)\n",
    "    \n",
    "def difference(old, new):\n",
    "    # relative zeroes over entire array\n",
    "    #return (new.apply(lambda x: numpy.ceil(x-epsilon)) - old.apply(lambda x: numpy.ceil(x-epsilon))).sum().sum()/len(old)**2    \n",
    "    return abs(new - old).sum().sum()/len(old)**2    \n",
    "\n",
    "#patient_sim = normalise(patient_sim, type = Orientation)\n",
    "_sim_a = patient_sim\n",
    "for i in range(0,num_iter):\n",
    "    # repeat E and I until convergence, the row-wise elements form the clusters.\n",
    "    _sim_b = clean(inflation(expansion(_sim_a), type = Orientation))\n",
    "    _sim_a = normalise(_sim_a, type = Orientation)\n",
    "    #if ((difference(_sim_a, _sim_b)) < convergence) & (i>0):\n",
    "    #    print(difference(_sim_a, _sim_b))\n",
    "    #    print(\"CONVERGED after \", i, \" iterations\")\n",
    "    #    break;\n",
    "    _sim_a = _sim_b\n",
    "\n",
    "result_mcl = clean(_sim_b)\n",
    "result_mcl.loc[result_mcl.loc['9827_corr2.CEL',:]>epsilon, '9827_corr2.CEL']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Affinity propagation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 5 patient clusters\n"
     ]
    }
   ],
   "source": [
    "from sklearn.cluster import AffinityPropagation\n",
    "patient_sim = _helpers.patient_similarity(Normal, sim_type = 'pearson', normalised = False, inflation=1, minkowski_dim=1)\n",
    "##### apply Affinity Propagation\n",
    "#######################\n",
    "X = numpy.array(patient_sim)\n",
    "af = AffinityPropagation(preference=-10).fit(X)\n",
    "cluster_centers_indices = af.cluster_centers_indices_\n",
    "n_clusters = len(cluster_centers_indices)\n",
    "patient_clusters = patient_sim.keys()[cluster_centers_indices].values\n",
    "patient_cluster_members = af.labels_\n",
    "print(\"There are {} patient clusters\".format(n_clusters))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "AggResults = Rocket.DATA_merged\n",
    "AggResults = _helpers._preprocess(AggResults, Rclass = Rocket)\n",
    "#AggResults = _helpers._group_patients(AggResults, method = 'mean')\n",
    "AggResults['cluster_ap'] = patient_cluster_members\n",
    "\n",
    "#AggResults.groupby(['Treatment risk group in ALL10', 'cluster_ap']).agg({'Microarray file': pandas.Series.nunique})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  after removing the cwd from sys.path.\n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \"\"\"\n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  import sys\n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  if __name__ == '__main__':\n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  # This is added back by InteractiveShellApp.init_path()\n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  if sys.path[0] == '':\n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  del sys.path[0]\n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\pandas\\core\\frame.py:2352: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self[k1] = value[k2]\n"
     ]
    }
   ],
   "source": [
    "AggResults['FU_RFS'] = pandas.to_numeric(AggResults['FU_RFS'])\n",
    "AggResults['FU_EFS'] = pandas.to_numeric(AggResults['FU_EFS'])\n",
    "AggResults['FU_OS'] = pandas.to_numeric(AggResults['FU_OS'])\n",
    "AggResults['WhiteBloodCellcount'] = pandas.to_numeric(AggResults['WhiteBloodCellcount'])\n",
    "AggResults['Age'] = pandas.to_numeric(AggResults['Age'])\n",
    "AggResults['Gender'] = pandas.to_numeric(AggResults['Gender'])\n",
    "AggResults['code_RFS']= pandas.to_numeric(AggResults['code_RFS'])\n",
    "AggResults['code_EFS']= pandas.to_numeric(AggResults['code_EFS'])\n",
    "AggResults['code_OS']= pandas.to_numeric(AggResults['code_OS'])\n",
    "\n",
    "AggResults['mutations_NOTCH_pathway'] = pandas.to_numeric(AggResults['mutations_NOTCH_pathway'])\n",
    "AggResults['mutations_PTEN_AKT_pathway'] = pandas.to_numeric(AggResults['mutations_PTEN_AKT_pathway'])\n",
    "AggResults['mutations_IL7R_pathway'] = pandas.to_numeric(AggResults['mutations_IL7R_pathway'])\n",
    "#AggResults.replace(to_replace=9999, value=0.5, inplace=True)\n",
    "AggResults[['mutations_NOTCH_pathway', \n",
    "            'mutations_PTEN_AKT_pathway', \n",
    "            'mutations_IL7R_pathway']] = AggResults[['mutations_NOTCH_pathway', \n",
    "                                                    'mutations_PTEN_AKT_pathway', \n",
    "                                                    'mutations_IL7R_pathway']].replace([9999],[numpy.nan],\n",
    "                                                                                       inplace=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\ipykernel_launcher.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  after removing the cwd from sys.path.\n"
     ]
    }
   ],
   "source": [
    "AggResults['comb_mutations_NOTCH_IL7R'] = AggResults['mutations_NOTCH_pathway'] + AggResults['mutations_IL7R_pathway']\n",
    "AggResults['comb_mutations_NOTCH_PTEN'] = AggResults['mutations_NOTCH_pathway'] + AggResults['mutations_PTEN_AKT_pathway']\n",
    "AggResults['comb_mutations_IL7R_PTEN'] =  AggResults['mutations_PTEN_AKT_pathway'] + AggResults['mutations_IL7R_pathway']\n",
    "AggResults['comb_mutations_IL7R_PTEN_NOTCH'] =  AggResults['mutations_PTEN_AKT_pathway']\\\n",
    "                                                + AggResults['mutations_IL7R_pathway']\\\n",
    "                                                + AggResults['mutations_NOTCH_pathway']\n",
    "\n",
    "\n",
    "patient_count = AggResults.groupby(['cluster_ap']).agg({'labnr_patient': pandas.Series.nunique})\n",
    "Clustered_by_patients_whitebloodcells = AggResults[AggResults['WhiteBloodCellcount'].apply(lambda x: isnan(x) is False)].groupby(['cluster_ap']).agg(\n",
    "    {'WhiteBloodCellcount': numpy.mean,\n",
    "    'Age': numpy.mean, \n",
    "    'Gender': numpy.mean})\n",
    "\n",
    "# Cancer_gene\n",
    "# Treatment_protocol\n",
    "# Treatment_risk_group_in_ALL_10\n",
    "\n",
    "Clustered_by_patients_CODE = AggResults.groupby(['cluster_ap']).agg(\n",
    "    {'code_RFS': numpy.mean, \n",
    "     'code_EFS': numpy.mean,\n",
    "     'code_OS': numpy.mean})\n",
    "\n",
    "Clustered_by_patients_FU_RFS = AggResults[AggResults['FU_RFS'].apply(lambda x: isnan(x) is False)].groupby(['cluster_ap']).agg(\n",
    "    {'FU_RFS': numpy.median, \n",
    "     'FU_EFS': numpy.median,\n",
    "     'FU_OS': numpy.median})\n",
    "Clustered_by_patients_NotchPath = AggResults[AggResults['mutations_NOTCH_pathway'].apply(lambda x: isnan(x) is False)].groupby(['cluster_ap']).agg(\n",
    "    {'mutations_NOTCH_pathway': numpy.mean})\n",
    "Clustered_by_patients_IL7RPath = AggResults[AggResults['mutations_IL7R_pathway'].apply(lambda x: isnan(x) is False)].groupby(['cluster_ap']).agg(\n",
    "    {'mutations_IL7R_pathway': numpy.mean})\n",
    "Clustered_by_patients_PTENAKTPath = AggResults[AggResults['mutations_PTEN_AKT_pathway'].apply(lambda x: isnan(x) is False)].groupby(['cluster_ap']).agg(\n",
    "    {'mutations_PTEN_AKT_pathway': numpy.mean})\n",
    "Clustered_by_patients_comb_NOTCH_IL7R = AggResults[AggResults['comb_mutations_NOTCH_IL7R'].apply(lambda x: isnan(x) is False)].groupby(['cluster_ap']).agg(\n",
    "    {'comb_mutations_NOTCH_IL7R': numpy.mean})\n",
    "Clustered_by_patients_comb_NOTCH_PTEN = AggResults[AggResults['comb_mutations_NOTCH_PTEN'].apply(lambda x: isnan(x) is False)].groupby(['cluster_ap']).agg(\n",
    "    {'comb_mutations_NOTCH_PTEN': numpy.mean})\n",
    "Clustered_by_patients_comb_IL7R_PTEN = AggResults[AggResults['comb_mutations_IL7R_PTEN'].apply(lambda x: isnan(x) is False)].groupby(['cluster_ap']).agg(\n",
    "    {'comb_mutations_IL7R_PTEN': numpy.mean})\n",
    "Clustered_by_patients_comb_IL7R_PTEN_NOTCH = AggResults[AggResults['comb_mutations_IL7R_PTEN_NOTCH'].apply(lambda x: isnan(x) is False)].groupby(['cluster_ap']).agg(\n",
    "    {'comb_mutations_IL7R_PTEN_NOTCH': numpy.mean})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cluster_agg = pandas.merge(Clustered_by_patients_whitebloodcells, Clustered_by_patients_CODE, how='inner', left_index=True, right_index=True)\n",
    "cluster_agg = pandas.merge(cluster_agg, Clustered_by_patients_comb_IL7R_PTEN, how='inner', left_index=True, right_index=True)\n",
    "cluster_agg = pandas.merge(cluster_agg, Clustered_by_patients_comb_IL7R_PTEN_NOTCH, how = 'inner', left_index=True, right_index=True)\n",
    "cluster_agg = pandas.merge(cluster_agg, Clustered_by_patients_comb_NOTCH_IL7R, how = 'inner', left_index=True, right_index=True)\n",
    "cluster_agg = pandas.merge(cluster_agg, Clustered_by_patients_comb_NOTCH_PTEN, how = 'inner', left_index=True, right_index=True)\n",
    "cluster_agg = pandas.merge(cluster_agg, Clustered_by_patients_FU_RFS, how='inner', left_index=True, right_index=True)\n",
    "cluster_agg = pandas.merge(cluster_agg, Clustered_by_patients_IL7RPath, how='inner', left_index=True, right_index=True)\n",
    "cluster_agg = pandas.merge(cluster_agg, Clustered_by_patients_NotchPath, how='inner', left_index=True, right_index=True)\n",
    "cluster_agg = pandas.merge(cluster_agg, Clustered_by_patients_PTENAKTPath, how='inner', left_index=True, right_index=True)\n",
    "cluster_agg = pandas.merge(cluster_agg, patient_count, how='inner', left_index=True, right_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "print(\"Cluster centers:\",patient_sim.keys()[cluster_centers_indices].values)\n",
    "print(patient_cluster_members)\n",
    "\n",
    "plt.close('all')\n",
    "plt.figure(figsize=(14,9))\n",
    "plt.clf()\n",
    "\n",
    "colors = cycle('bgrcmykbgrcmykbgrcmykbgrcmyk')\n",
    "for k, col in zip(range(n_clusters), colors):\n",
    "    class_members = patient_cluster_members == k\n",
    "    cluster_center = X[cluster_centers_indices[k]]\n",
    "    plt.plot(X[class_members, 0], X[class_members, 1], col + '.', \n",
    "             label = patient_sim.keys()[cluster_centers_indices[k]])\n",
    "    plt.plot(cluster_center[0], cluster_center[1], 'o', markerfacecolor=col,\n",
    "             markeredgecolor='k', markersize=14)\n",
    "    for x in X[class_members]:\n",
    "        plt.plot([cluster_center[0], x[0]], [cluster_center[1], x[1]], col)\n",
    "\n",
    "plt.legend()\n",
    "        \n",
    "plt.title('Estimated number of clusters from Affinity Propagation: %d' % n_clusters)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "##### CREATE graph from similarity matrix\n",
    "##################\n",
    "# nodes\n",
    "VarList = TransPosed.keys()\n",
    "nodes = []\n",
    "node_index = 0\n",
    "for patient_name in VarList:\n",
    "    nodes.append((node_index, {'name': patient_name}))\n",
    "    node_index = node_index + 1\n",
    "\n",
    "edges = []\n",
    "# edges\n",
    "patient_sim = patient_similarity(Normal, sim_type = 'pearson', normalised = True, inflation=2)\n",
    "node_index_x = 0\n",
    "node_index_y = 0\n",
    "for patient_name_x in VarList:\n",
    "    for patient_name_y in VarList:        \n",
    "        edges.append((node_index_x, node_index_y, patient_sim.iloc[node_index_x, node_index_y]))\n",
    "        node_index_y = node_index_y + 1\n",
    "    node_index_x = node_index_x + 1\n",
    "    node_index_y = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "G = nx.Graph()\n",
    "G.add_nodes_from(nodes)\n",
    "G.add_weighted_edges_from(edges, weight = 'weight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\networkx\\drawing\\nx_pylab.py:126: MatplotlibDeprecationWarning: pyplot.hold is deprecated.\n",
      "    Future behavior will be consistent with the long-time default:\n",
      "    plot commands add elements without first clearing the\n",
      "    Axes and/or Figure.\n",
      "  b = plt.ishold()\n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\networkx\\drawing\\nx_pylab.py:138: MatplotlibDeprecationWarning: pyplot.hold is deprecated.\n",
      "    Future behavior will be consistent with the long-time default:\n",
      "    plot commands add elements without first clearing the\n",
      "    Axes and/or Figure.\n",
      "  plt.hold(b)\n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\matplotlib\\__init__.py:917: UserWarning: axes.hold is deprecated. Please remove it from your matplotlibrc and/or style files.\n",
      "  warnings.warn(self.msg_depr_set % key)\n",
      "c:\\users\\bramva~1\\envs\\worken~1\\lib\\site-packages\\matplotlib\\rcsetup.py:152: UserWarning: axes.hold is deprecated, will be removed in 3.0\n",
      "  warnings.warn(\"axes.hold is deprecated, will be removed in 3.0\")\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAdgAAAFCCAYAAABSJMy8AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XtwVOXh//HP2U2y2RCIJCGiQRHlIo0gVFJFLCJCNQIi\nIFpBBhD8AWpAgcHS4qVVy4yDo1YdryNUR7GUSlEErYpGBBwC4ZKEBIlyUS5JIDESkqwJe35/bPGr\nFUIu5+zZy/s102mFc57nM2rz4Xn27HMM0zRNAQAAS7mcDgAAQCSiYAEAsAEFCwCADShYAABsQMEC\nAGADChYAABtQsAAA2ICCBQDABhQsAAA2oGABALABBQsAgA0oWAAAbEDBAgBgAwoWAAAbULAAANiA\nggUAwAYULAAANqBgAQCwAQULAIANKFgAAGxAwQIAYAMKFgAAG1CwAADYgIIFAMAGFCwAADaIcTpA\nSCgrk5YskXbskKqqpKQkqXdvafJkqUMHp9MBAMKQYZqm6XQIx+TmSgsXSmvWBP66ru7/fs/rlUxT\nysqS5s+XMjOdyQgACEvRW7DPPy/NnSvV1gaK9HQMI1C2ixZJM2YELx8AIKxF5xbxyXKtqTnztaYZ\nuG7u3MBfU7IAgCaIvhVsbq40aFDTyvV/JSRIOTlSv36WxwIARJboe4p44cLAtnBL1NYG7gcA4Ayi\nawVbViZ17vzzh5maKz5e2r+fp4sBAI2KrhXskiWtH8MwrBkHABDRoqtgd+xo3epVCmwT5+dbkwcA\nELGiq2CrqqwZp7LSmnEAABErugo2Kcmacdq3t2YcAEDEiq6C7d078JBSa3i9Uq9e1uQBAEQsniJu\nJjM+XgZPEQMAziC6VrBpaYGzhQ2jRbefkPR2XZ2MtDRrcwEAIk50FawUOLjf623RrXWSTh4zYRiG\nrrzySstiAQAiS/QVbGZm4OD+hIRm3WYmJGiOpC0/+bWNGzfKMAx9+umnViYEAESA6PoM9qda+DYd\no5HtZZ/Pp7i4OBvCAgDCTfStYE+aMSNwcP+oUYEni/9329jrDfz6qFGB6/77Fh3TNHX77befckiP\nx9NoAQMAokf0rmB/qrw8cPxhfn7gEIn27QNfxZk06bRPC1dXV6tt27anHfL666/XmpMvcgcARB0K\ntpXOtGJds2aNrr/++iClAQCEiujdIraIaZqaOnXqaX8/KytLhmHo2LFjQUwFAHAaK1iLVFRUKCUl\npdFrXC6XGhoa+JwWAKIAK1iLJCcny+/3N3qN3++Xy+ViyxgAogAFayHDMM64ZSxJH3zwgQzD0LJl\ny4KUDAAQbGwR2+TgwYNKT09v0rXl5eVKTU21OREAIJgoWBv5/X653e4mXevxeHT8+PEmXw8ACG1s\nEdvI5XI1ejDFT/l8PsXExGjgwIFBSAYAsBsr2CD58ssv1aNHjyZf/9JLL+nOO++0MREAwE4UbBDV\n19fL4/GoOX/L9+zZowsuuMC+UAAAW7BFHESxsbHy+/0aO3Zsk+/p0qWL4uPj5fP5bEwGALAaBeuA\nZcuWKS8vr8nX+3w+xcfH6/LLL2/W6hcA4BwK1iF9+/ZVTU1Ns+7ZtGmTXC6XFi1aZFMqAIBV+Aw2\nBIwYMUKrVq1q9n0FBQXKyMiwIREAoLUo2BCRk5OjQYMGNfu+2NhYVVRUKDEx0fpQAIAWo2BDSFVV\nlc4666wW3durVy9t27ZNLhe7/gAQCvhpHEKSkpLk9/t17bXXNvve/Px8ud1uzZ8/34ZkAIDmYgUb\not59913deOONLb4/JyeHU6EAwEEUbAgrLS3VOeec0+Kv5rjdbh0+fJgXCQCAA9giDmFnn3226uvr\n1b9//xbdf+LECXXo0EHdunXTDz/8YHE6AEBjKNgQ53a7tWHDBv39739v8RglJSXyeDyaNm0aB1UA\nQJCwRRxG9u3bZ8m5xCtXrmzV57sAgDOjYMOMz+fT5Zdfru3bt7dqHJfLpa+++ooXCQCATdgiDjMe\nj0fbtm3T008/3apx/H6/unTpok6dOun48eMWpQMAnMQKNowVFhbqkksusWSsW265RUuXLuWgCgCw\nCD9Nw1hGRoaOHTum7t27t3qsZcuWye12a/HixRYkAwBQsGEuMTFRxcXFevDBBy0Z74477pDL5VJB\nQYEl4wFAtGKLOIJ88cUXLf7O7Kl06NBBRUVFSklJsWxMAIgWrGAjyBVXXKGjR4/q/PPPt2S88vJy\npaamKisri4MqAKCZKNgIk5ycrD179mjmzJmWjfn+++/L4/Fo0aJFHFQBAE3EFnEE+89//qPrrrvO\n8nHXrVunq666yvJxASCSULAR7uDBg+rTp4/Ky8stHTcpKUkFBQXq1KmTpeMCQKRgizjCnXvuuTpw\n4IAmTJhg6bhVVVU677zzdOWVV3JQBQCcAgUbBWJjY/Xaa69p2bJllo+9ceNGJSYm6g9/+IP8fr/l\n4wNAuGKLOMqUlJSoX79+qqqqsmX8VatWadiwYbaMDQDhhBVslOnatasOHTqk4cOH2zL+8OHD1aZN\nG+3atcuW8QEgXFCwUcjr9erdd9/VCy+8YMv4NTU1uvjii9WrVy9VVFTYMgcAhDq2iKPc9u3b1b9/\nf9XW1to2x5133qlnn31WcXFxts0BAKGGFWyUu/TSS3Xo0CENHDjQtjlefvlleTweLVmyhIMqAEQN\nChZKSkrSp59+qoULF9o6z+TJk+X1erV582Zb5wGAUMAWMX5m/fr1Gjx4sO1nD1900UX67LPPdO65\n59o6DwA4hRUsfmbAgAH65ptvdOmll9o6z1dffaX09HSNHTuWgyoARCQKFr+QlpamLVu2aN68ebbP\ntXz5ciUmJurJJ5/koAoAEYUtYjRq9erVGjlypBoaGmyfKyYmRqtXr9bQoUNtnwsA7EbB4oz27dun\noUOHavfu3UGZr2PHjsrJyVH37t2DMh8A2IEtYpxR586dlZ+frylTpgRlvsOHD6tHjx4aOnQoB1UA\nCFsULJrE4/HolVde0RtvvKGYmJigzPnRRx8pJSVFf/rTn2x/qhkArMYWMZqtqKhIQ4YM0cGDB4M2\np8vl0tKlSzV27FgZhhG0eQGgpVjBotl69uypXbt2afTo0UGb0+/369Zbb1VKSory8vKCNi8AtBQF\nixZJTEzU8uXL9dxzz8ntdgdt3srKSl122WXKzMwM6goaAJqLLWK0Wm5urrKysnT06NGgzz19+nQt\nWrRIbdq0CfrcANAYChaWqKio0NixY7V27VpH5n/xxRc1depUuVxsygAIDfw0giWSk5P14Ycf6tFH\nHw3aU8Y/NW3aNLVr106ffPJJ0OcGgFNhBQvLrV27VqNGjdL333/vyPw9e/bUypUr1a1bN0fmBwCJ\nFSxsMHjwYO3cuVP9+vVzZP6ioiJ1795dv//971VZWelIBgCgYGGL9PR0bdiwQbNnz1ZsbKwjGf7x\nj38oOTlZCxcuVH19vSMZAEQvtohhuxUrVmjChAmOvpbO4/Horbfe0siRIzmoAkBQULAIipKSEo0Y\nMULFxcWO5jjvvPO0cuVK9e3b19EcACIfW8QIiq5duyovL0+TJ09WXFycYzm++eYb/frXv1ZWVhYH\nVQCwFQWLoPF6vXr11Vf1wgsvKCEhwdEs77//vtLT0zVv3jxHt64BRC62iOGI7du3a8SIEfrmm2+c\njiK3262XXnpJkyZN4qAKAJahYOGYqqoqTZw4UR988IHq6uqcjqOUlBQtX75cgwYNcjoKgAjAH9fh\nmKSkJK1YsUJ//etfHd8ylqSjR4/qmmuuUf/+/VVSUuJ0HABhjhUsQsL69es1evRolZWVOR3lR1On\nTtXjjz+u9u3bOx0FQBiiYBEyysrKdMstt2jTpk2qra11Oo4kyTAMLVq0SNnZ2Y4dmAEgPLFFjJCR\nlpamjz/+WHPmzFFiYqLTcSRJpmlqzpw5SklJ0cqVK8WfRwE0FStYhKQ1a9Zo3LhxqqqqCqlS69mz\np95880316dPH6SgAQhwrWISkrKwsbdu2TX369AmJB6BOKioqUt++fTV27FgOqgDQKAoWIatz587a\nuHGjJk2apKSkJKfj/Mzy5cuVnp6uBQsWqKamxuk4AEIQW8QIC0uXLtX06dN17NixkNoylgIvEnjx\nxRc1YcIEDqoA8CMKFmGjqKhII0eO1MGDB0PyeMPzzz9fr7/+ugYOHOh0FAAhgD9uI2z07NlTeXl5\nuvHGG5WSkuJ0nF/Yv3+/rr76ag0ZMoSDKgBQsAgviYmJeuONN/TII48oKSkpJN/t+vHHH6tbt266\n5557VFlZ6XQcAA5hixhhKzc3V6NHj9Z3332n6upqp+Ocktvt1hNPPKG77rqLgyqAKEPBIqwdPXpU\nEyZM0JYtW0LqmMX/lZqaqldffVXDhw8PyVU3AOuxRYywlpKSolWrVmnWrFkhfWbwkSNHdOONNyoz\nM1Pbtm1zOg6AIGAFi4ixdu1a3XbbbaqpqQnZLeOTxo0bp0WLFumcc85xOgoAm1CwiCgHDhzQrbfe\nqq+//lqHDh1yOk6jDMPQgw8+qHnz5oXUaVUArMEWMSJKenq6PvnkE40fP16pqalOx2mUaZr685//\nrLPPPluvvfaa/H6/05EAWIgVLCLWihUrNHXqVPl8vpA8mKKDpImSeks6S5K/bVv1GjdOFz7yiNSh\ng7PhALQaBYuIVlJSojFjxqiiokLffvut03EkSf0kzZeUJcmU9NPN4RpJMS6Xfrj2WiU+9piUmelE\nRAAWoGAR8Wpra5Wdna3Vq1c7/rnsNElPSIqX5G7kuhOSTsTEqOFPfwp8Prtpk1RYKB0/LtXVSfHx\nUmKi9KtfSb/5jTR5MqteIMRQsIgaixcv1uzZs1VfX+/IlvHJcm3TjHvM//6nSQ9LJCVJF14omaZk\nGIHC7dBB6t2bAgYcQMEiqmzfvl1jxoxRQ0OD9u3bF7R5+0n6VM0rV8u43VJMjHTDDdL8+Ww7A0HC\nU8SIKpdeeqm2bNmiyy67TBdccEHQ5p2vwLawI06ckHw+acUKacAA6fnnnUoCRBVWsIhKpmnqySef\n1GOPPaa6ujpbX5reQdI+SV7bZmiB3/5W+uwzp1MAEY0VLKKSYRiaPXu2Vq5cqfbt29u6mp2owOeo\nIWXdusDntM8843QSIGJRsIhqV111lfLy8nTRRRepe/futszRWz//Kk5ImTlTOuccKTfX6SRAxKFg\nEfXS0tL0wQcf6JZbbtHZZ58tr9fazdyzLB3NBocPB77qc//9TicBIgqfwQI/sWbNGk2cOFGJiYna\ns2ePJWO+JmmCJSMFwe9/Ly1d6nQKICKwggV+IisrS7m5uUpNTVVGRoYlY1YoBD+DPZ233pL+8Aen\nUwARgRUscAo+n09z5szRO++8o/LyctXV1bV4rGoFPoMNm9esG0bg5Kh+/ZxOAoQ1ChZoxNKlS5Wd\nna127do1e8v45OESYVWuJ116qcSL4YFWoWCBMygqKtLo0aOVkJCgvLy8Jt0zTdJzCnwGE3blelJZ\nGccrAq3AZ7DAGfTs2VO5ubnq0aOHunbtKo/H0+j10yQ9rcBh/mFbrpI0caLTCYCwxgoWaCLTNPXC\nCy/ogQceUNu2bbV3795fXOPomcN24McD0GKsYIEmMgxDM2bM0Jo1a+T3+3X55Zf/4pr5CrEjEVur\nqMjpBEDYomCBZsrMzFReXp5SUlJ0ySWXKC4uTlLgzOEsRdj/qe64w+kEQNiKqJ8FQLCkpKTo3Xff\n1W233ab27durc+fOishPLDdvdjoBELb4DBZopbVr12r8+PFacuKErisvdzqO9fgRAbQIBQtY4MCB\nA9rbq5cGVFY6HcV6/IgAWoQtYsAC6enp6n/99U7HABBCKFjAIq4+fQLHDAKA2CIGrFNWJnXsGFFb\nqn5JZkOD3G6301GAsMMKFrBKWpqUmup0CkuZkmJjY1VSUuJ0FCDsULCAlS67zOkEljElHVfgBKtu\n3brpiSeecDoSEFYoWMBK11zjdAJL/b+f/O+5c+ee8vQqAKfGZ7CAlcrKpLPPdjqFJUyd+k/gHo9H\nBw8eVHJycrAjAWGFFSxgpbQ0KSPD6RSW2H+aX/f5fEpJSdHbb78d1DxAuKFgAatFwGeVpqRnznDN\nmDFjNHbs2GDEAcISW8SAHfr0kbZvdzpFi9VKOl/SkSZcm5ycrEOHDv340gMAAaxgATu8/LIUE+N0\nihbxS1qtppWrJFVUVMjj8WjLli02pgLCDwUL2CEzU/rb36QwPKChVtLCFtzXr18/3X///VbHAcIW\nW8SAnZ5/Xrr77rA53ckv6S5JL7ZijK5du2rXrl1yufjzO6IbBQvY7T//ka67zukUZ2RKWixpigVj\nuVwu7d+/X+np6RaMBoQn/ogJ2O13v5MGDXI6RaNMSWtlTblKkt/vV6dOnfTyyy9bNCIQfljBAsGQ\nmysNGCDV1zud5BdMBb7zeoFN4w8cOFA5OTk2jQ6ELlawQDBkZkpPPy3Fxjqd5Bd8ksbYOP5nn32m\nNm3aqLq62sZZgNBDwQLBMmNGoGQ9HqeT/Oi4pHsl2f0Fm5qaGrVt21YffvihzTMBoYOCBYJpxgzp\n88+l0aMDRevQ13j8CpTrHLXuieHm+t3vfqfJkycHcUbAOXwGCzilvFxaskTKz5fy8qTiYunECVun\nbPjvf95T4LuuTh0NkZaWpoMHD/Iid0Q0ChYIJUVF0ty5UkGBVFEh+f2B79D6/YEHpPz+Jg1z8v/U\nPkkVkiolFUjKlfR3Nf2UJrvt3LlTPXv2dDoGYAsKFgg35eXSs89Kq1ZJpaWB0jUM/ZCUpF3ffadd\n332n2m7dNHvHjpAp0sY88sgjWrBggdMxAMtRsECE2bhxo7KzsxUXF6fvv/9ehYWFTkc6o4yMDBUU\nFDgdA7AUDzkBEaZ///7atGmTpkyZoiNHjujmm2+W1+t1OlajCgsLFRsbq6NHjzodBbAMBQtEIJfL\npSlTpqi4uFjp6elKTEzUTTfd5HSsRjU0NCg1NVVLly51OgpgCbaIgShQWFiomTNnqrS0VKZpaufO\nnU5HalRWVpZWr17tdAygVShYIEqYpqm3335bc+bM0cUXX6x169appqbG6VinlZiYqPLycsXHxzsd\nBWgRtoiBKGEYhsaMGaOdO3fqiiuukNfr1bBhw5yOdVrV1dXyer364osvnI4CtAgFC0SZhIQEPfzw\nw9q8ebPi4+PVpUsXZWRkOB3rtPr376+ZM2c6HQNoNraIgSj30UcfadasWUpJSVFeXp6OHz/udKRT\n6tSpk/bt28eL3BE2+DcViHJDhgzRtm3bNGbMGHm9Xg0dOtTpSKf07bffyu12a+/evU5HAZqEggWg\n2NhYzZo1S4WFhTr//PPVsWNH/epXv3I61il16dJFTz31lNMxgDNiixjAL+Tm5io7O1s+n08lJSUh\n+S7XzMxMbdq0yekYwGlRsABOye/36/XXX9f8+fPVtWtXrVu3zulIvxAbG6sjR46oXbt2TkcBfoEt\nYgCn5HK5NHHiRBUVFek3v/mNUlJSdPHFFzsd62fq6+uVlJSk9957z+kowC+wggXQJMXFxZo1a5b2\n7NmjQ4cOhdy28ZgxY7R8+XKnYwA/omABNJlpmnrnnXd03333KSUlRZs3b3Y60s8kJSXpyJEjiomJ\ncToKwBYxgKYzDEMjR47Uzp07NXLkSCUnJ6t79+5Ox/pRVVWVYmNjlZ+f73QUgIIF0Hzx8fFasGCB\ntm3bpr59+6pTp05q27at07F+1Lt3b/3xj390OgaiHFvEAFotJydH2dnZMk0zpF6cfsEFF+jrr7+W\nYRhOR0EUYgULoNWuvvpq5eXlafr06UpNTVW3bt2cjiRJ2rt3r1wulw4fPux0FEQhChaAJWJiYnT3\n3XerqKhIgwcPVlpaWshsG59zzjlavHix0zEQZdgiBmCLrVu3Kjs7W6WlpSopKXE6jiRpwIAB+vzz\nz52OgShBwQKwjWmaevPNNzVv3jx5PB7t2bPH6UiKi4vTd999J6/X63QURDi2iAHYxjAMjR8/XsXF\nxbrllluUnJzs+LbxDz/8oISEBH366aeO5kDkYwULIGh2796te++9V9u3b9eBAwecjqMJEybotdde\nczoGIhQFCyDo3nvvPc2aNUs+n0/ffvuto1mSk5NVXl7Oi9xhOf6NAhB0w4YNU2Fhoe655x4lJycr\nMTHRsSwVFRVyu90h8yAWIgcFC8ARHo9H999/v3bs2KGRI0cqNTXV0TzdunXTY4895mgGRBa2iAGE\nhM8//1zZ2dk6dOiQSktLHcvRo0cPFRcXOzY/IgcFCyBknDhxQq+88ooeeOAB1dTU6Pjx445lqays\n1FlnneXY/Ah/bBEDCBlut1vTpk1TcXGxJk2a5GjBtW/fXsuWLXNsfoQ/VrAAQtaOHTs0c+ZMFRQU\n6OjRo45kGDJkiD788ENH5kZ4o2ABhDTTNLVs2TLNnTtXFRUVqqmpCXoGj8ejY8eOKTY2NuhzI3yx\nRQwgpBmGoVtvvVXFxcWaPXu2kpKSgp7B5/MpLi5OmzZtCvrcCF+sYAGEla+//lqzZ8/WJ598ou+/\n/z7o80+fPl3PP/980OdF+KFgAYSlDz74QDNnztT+/ftVV1cX1LlTU1NVWlrK6U9oFP92AAhL1113\nnfLz8/Xoo48G/WnjI0eOyO12h8R5yghdFCyAsBUXF6c5c+Zo586dmjhxotq0aRPU+Tt16qS//e1v\nQZ0T4YMtYgAR44svvtA999yjgoIC+Xy+oM2bkZGhgoKCoM2H8EDBAogofr9fixcv1v333x/U784a\nhqHq6molJCQEbU6ENraIAUQUl8ulKVOmqKSkRLNmzVJ8fHxQ5jVNU23atNHq1auDMh9CHytYABGt\nsLBQ2dnZ+vzzz1VfXx+UOYcNG6ZVq1YFZS6ELgoWQMQzTVMrVqxQdna2Dh48GJQ5PR6Pampq+CpP\nFOOfPICIZxiGRo8erd27d+vhhx+Wx+OxfU6fzye3283DT1GMggUQNRISEvTQQw9p165dGj16tNxu\nt+1z9urVS3PnzrV9HoQetogBRK2PP/5YM2bM0O7du22fKy0tzdEXySP4WMECiFrXXnutCgsL9dRT\nT8nr9do6V1lZmQzD0JEjR2ydB6GDggUQ1WJjYzVr1izt3btXd9xxhwzDsHW+Dh066OWXX7Z1DoQG\ntogB4Cdyc3M1bdo0bd261dZ5evfure3bt9s6B5xFwQLA//D7/Xr99dd1zz33qLq62rZ5DMNQXV2d\n4uLibJsDzmGLGAD+h8vl0sSJE3XgwAHNmTPHtnlM05TH41FOTo5tc8A5rGAB4AyKi4s1ffp0W4tw\n9OjR+te//mXb+Ag+ChYAmsA0Tb377ru64447bHuJQHx8vI4fP87pTxGCf4oA0ASGYejGG2/Ut99+\nq0ceecSWOerq6uR2u7Vnzx5bxkdwUbAA0Azx8fFasGCB9u/fr5tuusmWOS688EItWLDAlrERPGwR\nA0Ar5OTkaPz48Tpw4IDlY3fs2FGHDh2yfFwEBytYAGiFq6++Wnv37tWzzz5r+SEVhw8flmEYOnbs\nmKXjIjgoWABopZiYGN19990qKyvT5MmTLR+/Xbt2evPNNy0fF/ZiixgALLZ161bddttt2rVrl6Xj\n9u3bV3l5eZaOCftQsABgA9M0tXTpUk2cOFENDQ2WjWsYhurr64Pyqj20DlvEAGADwzA0btw4VVRU\n6L777rNsXNM0FRMTo9zcXMvGhD1YwQJAEOzevVvjxo3T5s2bLRvz5ptv1j//+U/LxoO1KFgACKL3\n3ntPY8aMkc/ns2S8+Ph41dbWWjIWrMUWMQAE0bBhw1RVVWXZaVB1dXUyDIPvy4YgChYAgszj8WjB\nggU6cOCABg8ebMmY5557rh566CFLxoI12CIGAIetX79eWVlZlhwokZaWptLSUgtSobVYwQKAwwYM\nGKDKyko999xzrR6rrKxMhmGopqbGgmRoDQoWAEKA2+3WXXfdpaNHj2rUqFGtHq9NmzZavny5BcnQ\nUmwRA0AI2rFjh6655hpVVFS0apw+ffpo69atFqVCc1CwABCiTNPUW2+9pXHjxrV6rBMnTvAi9yDj\n7zYAhCjDMHTbbbepurq61S8RcLvdys/PtygZmoIVLACEia+//loDBgzQ4cOHWzwGpz8FDwULAGFm\n9erVGjZsWIvvj4uLs+wkKZweW8QAEGZuuOEG+Xw+3XvvvS26/4cffpBhGK1+gAqNo2ABIAzFxcXp\nySef1KFDh3ThhRe2aIyUlBTLjmzEL7FFDAARYMOGDRowYECL7k1OTtbRo0ctTgRWsAAQAa688kqd\nOHFCf/nLX5p9b0VFxY8vcod1KFgAiBAul0sPPPCAKisr1bNnz2bfHxcXp3feeceGZNGJLWIAiFCF\nhYXq1auXmvtjPiMjQwUFBTalih6sYAEgQmVkZOjEiRN65plnmnVfYWGhDMOwKVX0YAULAFGgtrZW\nV111lfLy8pp135dffqlu3brZlCqysYIFgCjg9Xq1ZcsW7dmzp1mr0+7du1vydp9oxAoWAKLQG2+8\nodtvv71Z91AXzUPBAkCUqq+v16BBg7Rhw4Ym33Ps2DElJibamCpysEUMAFEqNjZW69evV2lpqdxu\nd5Puadu2rR5++GF7g0UIVrAAAEnSqlWrNGLEiCZd6/V6VVNTY3Oi8EbBAgB+5Pf7NXToUK1du7ZJ\n1zc0NDR59Rtt2CIGAPzI5XLp448/VlVVVZOuj4mJ0cqVK21OFZ5YwQIATmvdunUaOHDgGa/r1KmT\nvvnmmyAkCh8ULACgUaZpavjw4Vq9enWTrkUABQsAaJK6ujp5vd4zXrdr1y517949CIlCG5/BAgCa\nJD4+XqZpnvG4xR49eui3v/1tkFKFLlawAIAWGTVqlP797383ek00VwwFCwBosYaGBsXGxjZ6TXV1\ntdq0aROkRKGDLWIAQIvFxMTINE3t3r37tNckJiZq+vTpQUwVGljBAgAsc9NNNzX6vdhoqhwKFgBg\nKdM05XKdfoM0WmqHLWIAgKUMw5Bpmtq/f/9pf/+VV145/QBlZdLjj0u33y6NGBH478cfl8rLbUps\nD1awAABnE4a/AAACi0lEQVRbZWVl6f333z/l7/2sgnJzpYULpTVrAn9dV/d/v+f1SqYpZWVJ8+dL\nmZk2JrYGBQsACArDME7566ZpSs8/L82dK9XWBor09IMEynbRImnGDJuSWoOCBQAETWlpqTp27Piz\nX5sm6bn4eLl/umI9k4SEkC9ZChYAEHT9+vXTli1b1E/Sp5Ja9C3ZhAQpJ0fq18/SbFahYAEAjnnb\nMDRSUoveKGsY0qhR0r/+ZXEqa1CwAABnlJVJnTv//GGm5oqPl/bvlzp0sC6XRfiaDgDAGUuWtH4M\nw7BmHBtQsAAAZ+zY0brVqxR46jg/35o8FqNgAQDOqKqyZpzKSmvGsRgFCwBwRlKSNeO0b2/NOBaj\nYAEAzujdO/CQUmt4vVKvXtbksRhPEQMAnMFTxAAA2CAtLXC28GmOUDwjw5BuuCEky1ViBQsAcFJu\nrjRokFRT0/x7Q/wkJ1awAADnZGYGzhROSGjefSfPIg7RcpWkGKcDAACi3MkD+3mbDgAANti8OfA+\n2NWrA0VaW/t/v3fyfbA33BB4H2wIr1xPomABAKGlvDxw/GF+fuAQifbtA1/FmTQpZB9oOhUKFgAA\nG/CQEwAANqBgAQCwAQULAIANKFgAAGxAwQIAYAMKFgAAG1CwAADYgIIFAMAGFCwAADagYAEAsAEF\nCwCADShYAABsQMECAGADChYAABtQsAAA2ICCBQDABhQsAAA2oGABALABBQsAgA0oWAAAbEDBAgBg\nAwoWAAAbULAAANiAggUAwAYULAAANqBgAQCwAQULAIANKFgAAGxAwQIAYAMKFgAAG1CwAADYgIIF\nAMAGFCwAADagYAEAsAEFCwCADShYAABsQMECAGADChYAABv8f4noJqN/OpfrAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1f31beac128>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "##### apply Spring-force\n",
    "#######################\n",
    "pos = nx.spring_layout(G, k = None, dim = 3, scale = 1.0)\n",
    "nx.draw_spring(G, k = 30, dim = 2, scale = 1.0, iterations =1500)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "##### APPLY community detector\n",
    "# maximize betweenness and modularity\n",
    "##################################\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "##### LOAD IN DATA\n",
    "###################\n",
    "# https://stackoverflow.com/questions/14529838/apply-multiple-functions-to-multiple-groupby-columns\n",
    "# "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
