{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "import multiprocessing\n",
    "import gc\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "import math\n",
    "import scipy as sc\n",
    "import sklearn as sk\n",
    "from random import randint, sample\n",
    "\n",
    "from numba import jit\n",
    "\n",
    "from scipy.stats import ks_2samp as ks2\n",
    "from scipy.stats import mannwhitneyu as mwu\n",
    "from scipy.stats import wasserstein_distance as w1_dist\n",
    "from scipy.stats import energy_distance as w2_dist\n",
    "from scipy.stats import anderson_ksamp as anderson # test if samples are from the same distribution\n",
    "\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "#%load_ext autoreload\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sc\n",
    "from sklearn import decomposition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA\n",
    "from sklearn.feature_selection import f_classif\n",
    "from sklearn.feature_selection import SelectFdr\n",
    "from sklearn.feature_selection import SelectFpr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Shared functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _dim_reduction(df= None, y= None, n_comp=50, method='PCA', pc_solver='arpack'):\n",
    "    # method: PCA\n",
    "    #print(\"There are {} features to reduce...woohooo\".format(len(gene_columns)))\n",
    "    ind = df.index\n",
    "    red_data_dict = {}\n",
    "    print(\"Running reducer on all samples..\")\n",
    "    X = df.loc[:, :]\n",
    "    if method=='PCA':\n",
    "        reducer = decomposition.PCA(copy= True, n_components= n_comp, svd_solver= pc_solver)\n",
    "        X_red = reducer.fit_transform(X)\n",
    "    elif method=='LDA':\n",
    "        reducer = LDA(n_components= n_comp, solver= 'svd')\n",
    "        X_red = reducer.fit_transform(X, y)\n",
    "    \n",
    "    red_cols = ['red_'+str(i) for i in range(0, n_comp)]\n",
    "    df_red = pd.DataFrame(data= X_red, index= ind, columns= red_cols)\n",
    "    \n",
    "    red_data_dict['ALL'] = {}\n",
    "    red_data_dict['ALL']['data'] = df_red\n",
    "    red_data_dict['ALL']['reducer'] = reducer\n",
    "    \n",
    "    return red_data_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _distance_significance(x,y, method='MWU'):\n",
    "    # difference between distributions\n",
    "    #\n",
    "    # consider 1st (EMD) and 2nd (Cramer Von Mises) Wasserstein distance --> equal to area between CDF's for 1D distro's\n",
    "    # consider Kolmogorov-Smirnov\n",
    "    # consider MW-U\n",
    "    # consider modified z-scores\n",
    "    # consider quantiles\n",
    "    # consider Kruskal Wallis\n",
    "    # diff per feature, per sample -> expensive\n",
    "    # scipy.stats import wasserstein_distance\n",
    "    # scipy.stats import energy_distance\n",
    "    # Kullback-Leibler divergence, Jensen-Shannon divergence, etc. only works for coupled data\n",
    "    # Spearman-Rank, Cosine-distance, etc. only works for coupled data or data of the same size.\n",
    "    \n",
    "    return True\n",
    "\n",
    "# significances of differences between pheno's.\n",
    "def distribution_difference_phenotypes(df, batches=[], phenotypes=[]):\n",
    "    tot_res = []\n",
    "    for _phenotype in phenotypes:    \n",
    "        res = []\n",
    "        for _batch_nr in batches:\n",
    "            _all = df.loc[df.batch_number!=_batch_nr][_phenotype]\n",
    "            _batch = df.loc[df.batch_number==_batch_nr][_phenotype]\n",
    "            _ks = ks2(_all.values, _batch.values)\n",
    "            _mwu = mwu(_all.values, _batch.values, alternative='two-sided')\n",
    "            try:\n",
    "                _w1 = w1_dist(_all[~np.isnan(_all)], _batch[~np.isnan(_batch)])\n",
    "                _w2 = w2_dist(_all[~np.isnan(_all)], _batch[~np.isnan(_batch)])\n",
    "            except ValueError as e:\n",
    "                print('ValError, defaulting to NaN')\n",
    "                _w1 = _w2 = np.nan\n",
    "\n",
    "            res.append({'batch_number': _batch_nr,\n",
    "                        'ks_stat': _ks[0], 'ks_p': _ks[1],\n",
    "                        'mwu_stat': _mwu[0], 'mwu_p': _mwu[1],\n",
    "                        'w1_dist': _w1, 'w2_dist': _w2})\n",
    "        res_df = pd.DataFrame(res)\n",
    "        tot_res.append({_phenotype: res_df})\n",
    "    return tot_res\n",
    "\n",
    "def _cohort_bias_metrics_colwise_reduced(cohort_dict, one_vs_all = True, equi_size = False):\n",
    "    '''\n",
    "     cohort_dict: {cohort_name: {'data': PC's, 'reducer': PCA transformer}}\n",
    "     one_vs_all: if True, compares each cohort to all other cohorts in bulk, \n",
    "                 else compares each cohort to each other cohort separately\n",
    "    '''\n",
    "    # Assumes reduction to 3 dimensions..\n",
    "    # Because the cohorts have different sample sizes we cannot simply do an inter-cohort correlation\n",
    "    # first we collect them in N columns, with N the number of PC's, then we cut those\n",
    "    # vectors in M pieces where M is the average number of samples per cohort, then we perform the\n",
    "    # correlation those artificial cohorts.\n",
    "    # TO DO: add statistical significance tests..Kruskal-Wallis H Test, MW-U, FDR, KS\n",
    "    metrics = {}\n",
    "    cohorts = list(cohort_dict.keys())\n",
    "    num_cohorts_ = len(cohorts) \n",
    "    if one_vs_all==False:\n",
    "        # avg samples per cohort: M        \n",
    "        \n",
    "        if equi_size:\n",
    "            pc1_matrix = cohort_dict[cohorts[0]]['data'][:,0]\n",
    "            pc2_matrix = cohort_dict[cohorts[0]]['data'][:,1]\n",
    "            pc3_matrix = cohort_dict[cohorts[0]]['data'][:,2]            \n",
    "            for _cohort in cohorts[1:-1]:\n",
    "                pc1_matrix = np.hstack((pc1_matrix, cohort_dict[_cohort]['data'][:,0]))\n",
    "                pc2_matrix = np.hstack((pc2_matrix, cohort_dict[_cohort]['data'][:,1]))\n",
    "                pc3_matrix = np.hstack((pc3_matrix, cohort_dict[_cohort]['data'][:,2]))\n",
    "            \n",
    "            len_ = pc1_matrix.shape[0]\n",
    "            avg_samples = int(np.floor(len_/num_cohorts_))\n",
    "            tot_num = avg_samples*len(cohorts)\n",
    "            # cut in M pieces\n",
    "            masked_ = np.random.randint(len_, size=(len_ - tot_num))\n",
    "            pc1_matrix_ = np.delete(pc1_matrix, masked_) #np.ma.array(pc1_matrix, mask=False).mask[masked_].compressed()\n",
    "            pc2_matrix_ = np.delete(pc2_matrix, masked_) #np.ma.array(pc2_matrix, mask=False).mask[masked_].compressed()\n",
    "            pc3_matrix_ = np.delete(pc3_matrix, masked_) #np.ma.array(pc3_matrix, mask=False).mask[masked_].compressed()\n",
    "            # reshape\n",
    "            pc1_matrix_ = pc1_matrix_.reshape((avg_samples, num_cohorts_))\n",
    "            pc2_matrix_ = pc2_matrix_.reshape((avg_samples, num_cohorts_))\n",
    "            pc3_matrix_ = pc3_matrix_.reshape((avg_samples, num_cohorts_)) \n",
    "            \n",
    "            pc_matrices = (pc1_matrix_, pc2_matrix_, pc3_matrix_)\n",
    "                        \n",
    "            # rank correlation, only works for equisized sets\n",
    "            metrics['spearman']={}\n",
    "            metrics['spearman'][0] = spearmanr(pc_matrices[0][:, :], pc_matrices[0][:, :])\n",
    "            metrics['spearman'][1] = spearmanr(pc_matrices[1][:, :], pc_matrices[1][:, :])\n",
    "            metrics['spearman'][2] = spearmanr(pc_matrices[2][:, :], pc_matrices[2][:, :])\n",
    "            # Wasserstein-Distance 1st and 2nd, \n",
    "            metrics['wasserstein']={'1st': [np.zeros((num_cohorts_, num_cohorts_)), \n",
    "                                            np.zeros((num_cohorts_, num_cohorts_)),\n",
    "                                            np.zeros((num_cohorts_, num_cohorts_))], \n",
    "                                    \n",
    "                                    '2nd': [np.zeros((num_cohorts_, num_cohorts_)),\n",
    "                                            np.zeros((num_cohorts_, num_cohorts_)),\n",
    "                                            np.zeros((num_cohorts_, num_cohorts_))]}   \n",
    "            \n",
    "            metrics['significance']={'MW-U': [np.zeros((num_cohorts_, num_cohorts_)), \n",
    "                                              np.zeros((num_cohorts_, num_cohorts_)),\n",
    "                                              np.zeros((num_cohorts_, num_cohorts_))]}   \n",
    "            for idx,pc_mat in enumerate(pc_matrices):\n",
    "                for cohort_ind in range(0,num_cohorts_):    \n",
    "                    for cohort_ind_sec in range(0,cohort_ind):\n",
    "                        metrics['wasserstein']['1st'][idx][cohort_ind, cohort_ind_sec] = w1_dist(\n",
    "                                                                          u_values=pc_mat[:, cohort_ind],\n",
    "                                                                          v_values=pc_mat[:, cohort_ind_sec])\n",
    "                        \n",
    "                        metrics['wasserstein']['2nd'][idx][cohort_ind, cohort_ind_sec] = w2_dist(\n",
    "                                                                          u_values=pc_mat[:, cohort_ind],\n",
    "                                                                          v_values=pc_mat[:, cohort_ind_sec]) \n",
    "\n",
    "                        metrics['significance']['MW-U'][idx][cohort_ind, cohort_ind_sec] = mannwhitneyu(\n",
    "                                                                          pc_mat[:, cohort_ind],\n",
    "                                                                          pc_mat[:, cohort_ind_sec])[1]\n",
    "        else: # non-equisized\n",
    "            # wasserstein 1/2\n",
    "            # log 10 of p-values of FDR of each cohort versus other cohorts (only for non-equisized cohorts)\n",
    "            metrics['wasserstein']={'1st': [np.zeros((num_cohorts_, num_cohorts_)), \n",
    "                                            np.zeros((num_cohorts_, num_cohorts_)),\n",
    "                                            np.zeros((num_cohorts_, num_cohorts_))], \n",
    "                                    \n",
    "                                    '2nd': [np.zeros((num_cohorts_, num_cohorts_)),\n",
    "                                            np.zeros((num_cohorts_, num_cohorts_)),\n",
    "                                            np.zeros((num_cohorts_, num_cohorts_))]}      \n",
    "\n",
    "            metrics['significance']={'MW-U': [np.zeros((num_cohorts_, num_cohorts_)), \n",
    "                                              np.zeros((num_cohorts_, num_cohorts_)),\n",
    "                                              np.zeros((num_cohorts_, num_cohorts_))] }\n",
    "            for idx in range(0,3):\n",
    "                for cohort_ind in range(0,num_cohorts_):    \n",
    "                    for cohort_ind_sec in range(0,cohort_ind):\n",
    "                        #print(\"cohort:\",cohorts[cohort_ind], \"idx:\",idx)\n",
    "                        metrics['wasserstein']['1st'][idx][cohort_ind, cohort_ind_sec] =\\\n",
    "                                    w1_dist(u_values=cohort_dict[cohorts[cohort_ind]]['data'][:, idx],\n",
    "                                            v_values=cohort_dict[cohorts[cohort_ind_sec]]['data'][:, idx])  \n",
    "                        \n",
    "                        metrics['wasserstein']['2nd'][idx][cohort_ind, cohort_ind_sec] =\\\n",
    "                                    w2_dist(u_values=cohort_dict[cohorts[cohort_ind]]['data'][:, idx],\n",
    "                                            v_values=cohort_dict[cohorts[cohort_ind_sec]]['data'][:, idx])\n",
    "                        \n",
    "                        metrics['significance']['MW-U'][idx][cohort_ind, cohort_ind_sec] =\\\n",
    "                                    mannwhitneyu(cohort_dict[cohorts[cohort_ind]]['data'][:, idx],\n",
    "                                                 cohort_dict[cohorts[cohort_ind_sec]]['data'][:, idx])[1]                        \n",
    "                        \n",
    "                        # FDR\n",
    "                       \n",
    "    else: # one-vs-all\n",
    "        # for each cohort compare to the distributions of the rest\n",
    "        # collect \n",
    "        pc1_matrix = cohort_dict[cohorts[0]]['data'][:,0]\n",
    "        pc2_matrix = cohort_dict[cohorts[0]]['data'][:,1]\n",
    "        pc3_matrix = cohort_dict[cohorts[0]]['data'][:,2]            \n",
    "        for _cohort in cohorts[1:-1]:\n",
    "            pc1_matrix = np.hstack((pc1_matrix, cohort_dict[_cohort]['data'][:,0]))\n",
    "            pc2_matrix = np.hstack((pc2_matrix, cohort_dict[_cohort]['data'][:,1]))\n",
    "            pc3_matrix = np.hstack((pc3_matrix, cohort_dict[_cohort]['data'][:,2]))\n",
    "            \n",
    "        pc_matrices = (pc1_matrix, pc2_matrix, pc3_matrix)\n",
    "\n",
    "        metrics['wasserstein']={'1st': [np.zeros((num_cohorts_)), \n",
    "                                        np.zeros((num_cohorts_)),\n",
    "                                        np.zeros((num_cohorts_))], \n",
    "                                    \n",
    "                                '2nd': [np.zeros((num_cohorts_)),\n",
    "                                        np.zeros((num_cohorts_)),\n",
    "                                        np.zeros((num_cohorts_))]}     \n",
    "        metrics['significance']={'MW-U': [np.zeros((num_cohorts_)), \n",
    "                                        np.zeros((num_cohorts_)),\n",
    "                                        np.zeros((num_cohorts_))] }\n",
    "    \n",
    "        for idx in range(0,3):\n",
    "            for cohort_ind in range(0, num_cohorts_):\n",
    "                metrics['wasserstein']['1st'][idx][cohort_ind]=\\\n",
    "                                w1_dist(u_values=cohort_dict[cohorts[cohort_ind]]['data'][:, idx],\n",
    "                                        v_values=pc_matrices[idx])\n",
    "                    \n",
    "                metrics['wasserstein']['2nd'][idx][cohort_ind] =\\\n",
    "                                w2_dist(u_values=cohort_dict[cohorts[cohort_ind]]['data'][:, idx],\n",
    "                                        v_values=pc_matrices[idx])\n",
    "                \n",
    "                metrics['significance']['MW-U'][idx][cohort_ind]=mannwhitneyu(\n",
    "                                cohort_dict[cohorts[cohort_ind]]['data'][:, idx],\n",
    "                                pc_matrices[idx])[1]\n",
    "    return metrics\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#@jit\n",
    "def median_shift(x, batch_ids):\n",
    "    '''\n",
    "     x : dataset \n",
    "     batch_ids : dict\n",
    "    ''' \n",
    "    \n",
    "    delta = np.zeros([x.shape[0], len(batch_ids)])\n",
    "    for idx, _batch in tqdm(enumerate(batch_ids.keys())):\n",
    "        median_batch = np.median(x[:, batch_ids[_batch]], axis=1)\n",
    "        #median_all = np.median(x, axis=1)\n",
    "        median_all = np.median(np.delete(x, batch_ids[_batch], axis=1), axis=1)\n",
    "        delta[:,idx] = median_batch - median_all\n",
    "        \n",
    "    return delta \n",
    "\n",
    "@jit\n",
    "def mean_shift(x, batch_ids):\n",
    "    '''\n",
    "     x : dataset \n",
    "     batch_ids : dict\n",
    "    ''' \n",
    "    \n",
    "    delta = np.zeros([x.shape[0], len(batch_ids)])\n",
    "    for idx, _batch in tqdm(enumerate(batch_ids.keys())):\n",
    "        mean_batch = np.mean(x[:, batch_ids[_batch]], axis=1)\n",
    "        #mean_all = np.mean(x, axis=1)\n",
    "        mean_all = np.mean(np.delete(x, batch_ids[_batch], axis=1), axis=1)\n",
    "        delta[:,idx] = mean_batch - mean_all\n",
    "        \n",
    "    return delta "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "class fs_mannwhitney():\n",
    "    pvalues_ = None\n",
    "    scores_ = None\n",
    "    def __init__(self, pvalue = 0.01, mode='auto'):\n",
    "        # mode : 'auto', 'exact', 'asymp'\n",
    "        self.pvalue = pvalue\n",
    "        self.mode = mode\n",
    "\n",
    "    def apply_test(self, pos, neg, column):\n",
    "        zscore, p_value = mwu(pos[:,column], neg[:,column], alternative=\"two-sided\") # mode=self.mode\n",
    "        return zscore, p_value\n",
    "    \n",
    "    def fit(self, x, y):\n",
    "        zero_idx = np.where(y == 0)[0]\n",
    "        one_idx = np.where(y == 1)[0]\n",
    "        pos_samples = x[one_idx]\n",
    "        neg_samples = x[zero_idx]                \n",
    "        results_ = np.array(list(map(lambda c: \n",
    "            self.apply_test(pos_samples, neg_samples, c), range(0,x.shape[1]))))\n",
    "        self.scores_ = results_[:, 0]\n",
    "        self.pvalues_ = results_[:, 1]\n",
    "        return self\n",
    "\n",
    "    def transform(self, x):\n",
    "        not_signif = self.p_values<self.pvalue\n",
    "        to_delete = [idx for idx, item in enumerate(not_signif) if item == False]\n",
    "        return np.delete(x, to_delete, axis = 1), to_delete\n",
    "\n",
    "class fs_ks():\n",
    "    pvalues_ = None\n",
    "    scores_ = None\n",
    "    def __init__(self, pvalue = 0.01):\n",
    "        self.pvalue = pvalue\n",
    "\n",
    "    def apply_test(self, pos, neg, column):\n",
    "        zscore, p_value = ks2(pos[:,column], neg[:,column])\n",
    "        return zscore, p_value\n",
    "    \n",
    "    def fit(self, x, y):\n",
    "        zero_idx = np.where(y == 0)[0]\n",
    "        one_idx = np.where(y == 1)[0]\n",
    "        pos_samples = x[one_idx]\n",
    "        neg_samples = x[zero_idx]                \n",
    "        results_ = np.array(list(map(lambda c: \n",
    "            self.apply_test(pos_samples, neg_samples, c), range(0,x.shape[1]))))\n",
    "        self.scores_ = results_[:, 0]\n",
    "        self.pvalues_ = results_[:, 1]\n",
    "        return self\n",
    "\n",
    "    def transform(self, x):\n",
    "        not_signif = self.p_values<self.pvalue\n",
    "        to_delete = [idx for idx, item in enumerate(not_signif) if item == False]\n",
    "        return np.delete(x, to_delete, axis = 1), to_delete\n",
    "    \n",
    "class fs_ad():\n",
    "    pvalues_ = None\n",
    "    crit_vals = None\n",
    "    scores_ = None\n",
    "    \n",
    "    def __init__(self, pvalue = 0.01):\n",
    "        self.pvalue = pvalue\n",
    "\n",
    "    def apply_test(self, samples):\n",
    "        zscore, crit_val, p_value = anderson(samples)\n",
    "        return zscore, crit_val, p_value\n",
    "    \n",
    "    def fit(self, x):             \n",
    "        results_ = np.array(list(map(lambda c: \n",
    "            self.apply_test(x[:,c].reshape([-1,1])), range(0, x.shape[1]))))\n",
    "        self.scores_ = results_[:, 0] \n",
    "        self.crit_vals = results_[: ,1]\n",
    "        self.pvalues_ = results_[:, 2]\n",
    "        return self\n",
    "\n",
    "    def transform(self, x):\n",
    "        not_signif = self.p_values<self.pvalue\n",
    "        to_delete = [idx for idx, item in enumerate(not_signif) if item == False]\n",
    "        return np.delete(x, to_delete, axis = 1), to_delete\n",
    "    \n",
    "class fs_ws1():\n",
    "    scores_ = None\n",
    "    def __init__(self, pvalue = 0.01):\n",
    "        self.pvalue = pvalue\n",
    "\n",
    "    def apply_test(self, pos, neg, column):\n",
    "        zscore = w1_dist(pos[:,column], neg[:,column])\n",
    "        return zscore\n",
    "    \n",
    "    def fit(self, x, y):\n",
    "        zero_idx = np.where(y == 0)[0]\n",
    "        one_idx = np.where(y == 1)[0]\n",
    "        pos_samples = x[one_idx]\n",
    "        neg_samples = x[zero_idx]                \n",
    "        results_ = np.array(list(map(lambda c: \n",
    "            self.apply_test(pos_samples, neg_samples, c), range(0,x.shape[1]))))\n",
    "        self.scores_ = results_\n",
    "        return self\n",
    "    \n",
    "class fs_ws2():\n",
    "    scores_ = None\n",
    "    def __init__(self, pvalue = 0.01):\n",
    "        self.pvalue = pvalue\n",
    "\n",
    "    def apply_test(self, pos, neg, column):\n",
    "        zscore = w2_dist(pos[:,column], neg[:,column])\n",
    "        return zscore\n",
    "    \n",
    "    def fit(self, x, y):\n",
    "        zero_idx = np.where(y == 0)[0]\n",
    "        one_idx = np.where(y == 1)[0]\n",
    "        pos_samples = x[one_idx]\n",
    "        neg_samples = x[zero_idx]                \n",
    "        results_ = np.array(list(map(lambda c: \n",
    "            self.apply_test(pos_samples, neg_samples, c), range(0,x.shape[1]))))\n",
    "        self.scores_ = results_\n",
    "        return self\n",
    "    \n",
    "    \n",
    "def get_filtered_genomes(x, y, Rclass = None):\n",
    "    try:\n",
    "        alpha = Rclass.PIPELINE_PARAMETERS['feature_selection']['pvalue']\n",
    "        filter_type = Rclass.PIPELINE_PARAMETERS['feature_selection']['method']\n",
    "        F_function =  Rclass.PIPELINE_PARAMETERS['feature_selection']['score_function']\n",
    "    except Exception as e:\n",
    "        print(\"Exception with {} handling the function get_filtered_genomes\".format(e))\n",
    "        alpha = 0.05\n",
    "        filter_type = 'FDR'\n",
    "        F_function = 'ANOVA'    \n",
    "    \n",
    "    #  Mann-Whitney, between classes\n",
    "    # scipy.stats.mannwhitneyu, https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mannwhitneyu.html\n",
    "    if filter_type == 'mannwhitney':\n",
    "        Selector = fs_mannwhitney(pvalue = alpha).fit(x,y)\n",
    "        x_out = Selector.transform(x)[0]\n",
    "    elif filter_type == 'FDR':\n",
    "    # Use FDR with a number of different statistical measures:\n",
    "    # f_classif, chi2, \n",
    "        FDR_function = f_classif if F_function == 'ANOVA' else eval(F_function)\n",
    "        FDR = SelectFdr(alpha = alpha, score_func = FDR_function) #f_classif, chi2\n",
    "        Selector = FDR.fit(x, y)\n",
    "        x_out = FDR.transform(x)\n",
    "    elif filter_type == 'FPR':\n",
    "        FPR_function = f_classif if F_function == 'ANOVA' else eval(F_function)\n",
    "        FPR = SelectFpr(alpha = alpha, score_func = FPR_function) #f_classif, chi2\n",
    "        Selector = FPR.fit(x, y)\n",
    "        x_out = FPR.transform(x)        \n",
    "\n",
    "    return x_out, Selector\n",
    "\n",
    "#@jit('int32[:](float64[:,:], int32[:]))\n",
    "def get_differential_probes(x, y, method='FDR', alpha=0.05, **kwargs):\n",
    "    '''\n",
    "     x : array (samples x probes)\n",
    "     y : array (target values)\n",
    "     method : FDR/MWU/KS \n",
    "     n_top : number of top discriminating probes \n",
    "     kwargs: arguments for stat method\n",
    "     \n",
    "     return: index array of top n discriminators\n",
    "    '''\n",
    "        \n",
    "    idx = x.columns\n",
    "    if method=='FDR':\n",
    "        F_function = 'ANOVA'\n",
    "        FDR_function = f_classif if F_function == 'ANOVA' else eval(F_function)\n",
    "        FDR = SelectFdr(FDR_function, alpha = alpha) #f_classif, chi2\n",
    "        selector = FDR.fit(x, y)\n",
    "    elif method=='FPR':\n",
    "        F_function = 'ANOVA'\n",
    "        FPR_function = f_classif if F_function == 'ANOVA' else eval(F_function)\n",
    "        FPR = SelectFdr(FPR_function, alpha = alpha) #f_classif, chi2\n",
    "        selector = FPR.fit(x, y)      \n",
    "    elif method=='MWU':\n",
    "        selector = fs_mannwhitney(pvalue = alpha).fit(x.values,y)\n",
    "    elif method=='KS':\n",
    "        selector = fs_ks(pvalue = alpha).fit(x.values,y)\n",
    "    elif method=='AD':\n",
    "        selector = fs_ad().fit(x.values) \n",
    "    elif method=='WASS1':\n",
    "        selector = fs_ws1().fit(x.values, y)\n",
    "    elif method=='WASS2':\n",
    "        selector = fs_ws2().fit(x.values, y)\n",
    "    if method in ['WASS1', 'WASS2']:\n",
    "        return pd.DataFrame(data=selector.scores_.T, \n",
    "                        columns=['score_'+method],\n",
    "                        index=idx)\n",
    "    else:\n",
    "        return pd.DataFrame(data=np.vstack((selector.pvalues_, selector.scores_)).T, \n",
    "                        columns=['pvals_'+method, 'score_'+method],\n",
    "                        index=idx)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data\n",
    "Due to resources limitations we analysis one methylation set at a time. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "MethylationForm = 'beta' # 'beta', 'M'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"/media/koekiemonster/DATA-FAST/genetic_expression/hackathon_2/Lung\"\n",
    "# \"/home/bramiozo/DATA/LungCancerResearch/\" \n",
    "source_dir = \"/media/koekiemonster/DATA-FAST/genetic_expression/hackathon_2/Lung/\"\n",
    "\n",
    "# _prepped/methylation_Mvalues.csv Lung_Methylation.txt\n",
    "Meth_Raw = pd.read_csv(source_dir+\"Lung_Methylation.txt\", sep=\"\\t\", engine=\"python\")\n",
    "\n",
    "Meth_Raw.set_index('probeID', inplace=True)\n",
    "GeneMapping = Meth_Raw[['Chr', 'Start', 'Stop', 'Strand', 'Gene', 'Relation_CpG_Island']]\n",
    "Meth_Raw.drop(['Chr', 'Start', 'Stop', 'Strand', 'Gene', 'Relation_CpG_Island'], inplace=True, axis=1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "if MethylationForm=='M':\n",
    "    offset = 1e-3\n",
    "    def Mtransform(x):\n",
    "        try:\n",
    "            return np.log2((x+offset)/(1-x+offset))\n",
    "        except ValueError:\n",
    "            return np.nan\n",
    "\n",
    "    pat_cols = [_col for _col in methylation.columns if \"TCGA\" in _col]    \n",
    "    Meth_Raw = Meth_Raw.copy().loc[:, pat_cols].progress_apply(Mtransform, axis=0)\n",
    "    gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "methylation_meta = pd.read_csv(source_dir+\"HumanMethylation450_meta.csv\", sep=\",\", engine=\"python\", skiprows=7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gender (small) Pack Years (small), Smoking Status (small), Diagnosis (small), Age At Diagnosis (Days) (small)\n",
    "# batch_number (large), histological_type (large), cigarettes_per_day.exposures (large), tissue_or_organ_of_origin.diagnoses (large)\n",
    "# race.demographic\n",
    "pheno_small = pd.read_csv(\"../../_docs/Lung_Phenotype_Metadata.txt\", sep=\"\\t\")\n",
    "pheno_small = pheno_small.copy().loc[pheno_small[\"Sample Type\"]=='Primary Tumor']\n",
    "pheno_small.set_index('SampleID', inplace=True)\n",
    "pheno_large = pd.read_csv(\"../../_docs/Lung_Table_Phenotypes.txt\", sep=\"\\t\")\n",
    "pheno_large.set_index('submitter_id.samples', inplace=True)\n",
    "pheno_large['SampleID'] = pheno_large.index\n",
    "pheno_large = pheno_large.copy().loc[pheno_large[\"sample_type.samples\"]=='Primary Tumor']\n",
    "\n",
    "pheno = pheno_small[['Gender', 'Pack Years', 'Smoking Status', 'Diagnosis', 'Age At Diagnosis (Days)']]\\\n",
    "                    .merge(pheno_large[['batch_number', 'histological_type', 'cigarettes_per_day.exposures',\n",
    "                                       'tissue_or_organ_of_origin.diagnoses', 'race.demographic']], how='inner',\n",
    "                                       left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# split by diagnosis\n",
    "LA_samples = pheno.loc[pheno.Diagnosis=='Lung Adenocarcinoma'].index.values\n",
    "LSCC_samples = pheno.loc[pheno.Diagnosis=='Lung Squamous Cell Carcinoma'].index.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "93"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# to check for cohort bias we only use the reference probe initially\n",
    "gc.collect()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:1494: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "https://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike\n",
      "  return self._getitem_tuple(key)\n"
     ]
    }
   ],
   "source": [
    "Meth_data = {'LA': Meth_Raw.loc[:, LA_samples],\n",
    "             'LSCC': Meth_Raw.loc[:, LSCC_samples]}\n",
    "del Meth_Raw\n",
    "\n",
    "gc.collect()\n",
    "\n",
    "Meth_data_T = {'LA': Meth_data['LA'].T, \n",
    "               'LSCC': Meth_data['LSCC'].T}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Seperability of cohorts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _clean(df, complete=True):\n",
    "    drop_rows = df[df.isna().sum(axis=1)>450000].index.tolist()\n",
    "    drop_cols = df.columns[df.isna().sum(axis=0)>400].tolist()\n",
    "    print(\"We drop {} samples and {} probes\".format(len(drop_rows), len(drop_cols)))\n",
    "    df_out = df.copy().drop(drop_rows, axis=0).drop(drop_cols, axis=1)\n",
    "    if complete:\n",
    "        drop_cols = df_out.copy().columns[df_out.copy().isna().sum(axis=0)>0].tolist()\n",
    "        df_out = df_out.copy().drop(drop_cols, axis=1)\n",
    "    initial_points = df.shape[0]*df.shape[1]\n",
    "    removed = (initial_points - df_out.shape[0]*df_out.shape[1])\n",
    "    print(\"Removed {} points from an initial {} points\".format(removed, initial_points))\n",
    "    return df_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We drop 65 samples and 89512 probes\n",
      "Removed 87750581 points from an initial 269980812 points\n",
      "We drop 134 samples and 89512 probes\n",
      "Removed 103868270 points from an initial 246187539 points\n"
     ]
    }
   ],
   "source": [
    "_cleaned_LA = _clean(Meth_data_T['LA'])\n",
    "_cleaned_LSCC = _clean(Meth_data_T['LSCC'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of NA's: 174262840\n",
      "Total number of datapoints: 516168351\n",
      "NA percentage: 0.3376085334608204\n"
     ]
    }
   ],
   "source": [
    "tot_na = Meth_data['LA'].isna().sum().sum() + Meth_data['LSCC'].isna().sum().sum()\n",
    "tot_points = Meth_data['LA'].size + Meth_data['LSCC'].size\n",
    "print(\"Total number of NA's:\", tot_na)\n",
    "print(\"Total number of datapoints:\", tot_points)\n",
    "print(\"NA percentage:\", tot_na/tot_points)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We drop 199 samples and 89841 probes\n",
      "Removed 200133567 points from an initial 516168351 points\n"
     ]
    }
   ],
   "source": [
    "_cleaned_ALL = _clean(pd.concat([Meth_data_T['LA'], Meth_data_T['LSCC']]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, '# affected samples')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0wAAAJcCAYAAADKPCSYAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nOzde7htdV0v/vdHQHYkyjUCNgaiRWqGSoalpliGl8TTT8xLiZekPHoyTVMrk06dyjpqWh2N8oKXwGsPHC2NVI6ccxIFb6BoIuFxEzeRO4kIn98fc2xbLPd37bVxzbUWm9freeYzx/iOy/zM7xws5nuPMb6zujsAAAB8pzusdQEAAADrlcAEAAAwIDABAAAMCEwAAAADAhMAAMCAwAQAADAgMAFwu1VVB1ZVV9WOc9j3cVX1tmn6rlV1bVXtsEL7fn1VvWyafmhVbVqJ/U77e3BVfXGl9gdwWycwAcxBVX28qn6wqu5WVZ/cyrpdVWdX1R0WtP1BVb15ma/15mkfD1jQdveq6gXzp1XVQ7f9nbASuvv/dfeduvumpdarqqdV1f9exv5+tbt/fyVqm46duy/Y9+nd/UMrsW+A7YHABLDCqmqnJD+Q5EtJ7p9kycA02S/JE7+Ll/16kj/4Lrafq3mcwbm9WqmzVAAsj8AEsPLuneTz3d1JDsvyAtOfJPm9UbCoqndV1cVVdVVVfbSq7rVolROS3KeqfmprL1RVD6iqM6vq6qq6pKpeNVjvoVW1qap+q6q+VlUXVNVTFizfuar+e1X9v2k/r6+q71m07Yur6uIkb9rC/u9eVf9rek9fq6p3LFj2mqr66lTjWVX14AXLjpv6421Vdc10du4Hq+qlVXXptN0jFqx/WlX90XTW7+qqOrmq9hi857tU1Ruq6qKqunA607esgFJVB03v55qqOjXJXguW3eLSv+lM0vnTuv9aVU+pqh9O8vokD5wu37tyWvfNVfW6qvr7qrouycOmtj9Y9Pqjz+m0qvrlBfPfPotVVR+dmj8zveYvLL7Er6p+eNrHlVX1uap67IJlb66qv6yq90/v5YyqOng5/QVwWyEwAayQqnr69CX3/2T2pffKJL+R5BXTl82Dltj8vUmuTvK0wfJ/SHKPJN+XWQB7+6Ll1yf5wyT/bUsbd/dDu/u0afY1SV7T3XdOcnCSdy5R1/dn9sV//yTHJDm+qjZfrvXHSX4wyaFJ7j6t87uLtt0js7Ntx25h37+f5B+T7J5kY5I/X7DsE9N+90jyt0neVVUbFiz/uSRvnbb9VJIPZvb/tP2T/Nckf7XotZ6a5BlJ9k3yrSSvHbzfN0/L757kvkkekeSXB+su9rdJzsqsv34/s/76DlX1vdPrP7K7d03yE0k+3d3nJvnVJP88Xb6324LNnpzZZ7trki1dsrfU5zTU3Q+ZJn90es13LFw+nS39n5l9Tt+X5L8kefuifT8xye9l9lmcl8ExCHBbJTABrJDuftP0JfesJIcnuU+Sc5Lcubt36+5/XWrzJC9L8rKquuMW9v3G7r6mu29IclySH62quyxa7a+S3LWqHrmVUm9Mcveq2qu7r+3uj21l/Zd19w3d/b+SvD/JE6qqMgtBz+/ur3f3NZkFtoWXFd6c5OXTtv8+qOMHkuzX3d/o7m8Hge5+W3df3t3f6u5XJtk5ycIv6ad39we7+1tJ3pVk7yR/3N03JjkpyYFVtTBwvLW7z+nu6zLr5ycsPnNUVfskeVSSX+/u67r70iSvzjIulayquyb5sQV99dHMgsbIzUnuXVXf090XdffntvISJ3f3/+num7v7G4N1vuNz2lrdy3B4kjtl1rff7O4PJ3lfkictWOfvuvvj02fx9syCLsB2Q2ACWAFVtcd0FumqzM4YnJbki5l9yb+iqn59a/vo7r9PsinJryza9w5V9cdV9eWqujrJBdOivRZtf0NmZza2NhjAMzM7M/SFqvpEVT1miXWvmELGZl/J7H6rvZPskuSs6X1fmeQDU/tmly3x5T5JfjNJJfn4dKnXMzYvqKoXVtW50+V6Vya5y6L3e8mC6X9P8rUFAypsDmd3WrDOVxe9h50W7S+Zhbedkly04D39VWZnVrZmv2y5r77DtM4vZHY26aLpcrZDtrL/r25l+ehz+m7tl+Sr3X3zon3vv2D+4gXT1+eW/Q5wmycwAayA6SzLbpmFnb+Zpj+Q5Oems0t/tsxd/XaS38osjGz25CRHJfnpzILDgVN7bWH7NyXZLcnPL1Hrl7r7SZkFgVckefd0mdiW7L5o2V2T/FuSr2UWTO41vb/duvsu3b3wy3JnCd19cXc/q7v3y6zf/sd0X9ODMwtTT0iy+9SXVw3e73IdsOg93Di9h4W+muSGJHsteE937u7F94ttyUXZcl9t0XR27Gcyu0TwC0n+evOi0SZbef3R55Qk1+WWx9P3b2VfC/1bkgNqwQiO074v3IZ9ANymCUwAK2vhqHj3zezyvGWb7jM6J7e8/2XXzL7IX57ZF98/XGL7byV5eZIXj9apql+sqr2nswZXTs03j9bPbDCKO05B5jFJ3jVt+9dJXl1V3zftd/+q+tmtvMWFdRxdVRun2SsyCwU3Z/Z+v5XksiQ7VtXvJrnzcvc78ItVdc+q2iWze5zevXiI7+6+KLN7dV5ZVXeuqjtU1cE1DaSxYOCGAxfvvLu/kuTM/EdfPSiz+6y29L73qaqjpoBzQ5Jr8x/9f0mSjVu6LHMZvuNzmto/neTnq2qXmg0f/sxF212S5G6DfZ6R2Vmj36yqnWo2NP3PZXbZI8DtgsAEsLLun+STVbVnkpu6+4pbsY/fyWywg83ektllUBcm+XySrd1zdGJmZzxGjkzyuaq6NrMBIJ44uMcomV1udUVmZxrenuRXu/sL07IXZ3aT/8emSwX/Kbe8z2hrfizJGVMdpyR5Xnefn9kADh9I8i+Zve9vZOuXpG3NWzMb0OHiJBuS/NpgvacmuWNm/XxFkndndhYomZ2l2vw5bMmTk/x4ZkO8vzyzz21L7pDkBZn16deT/FSSZ0/LPpzkc0kurqrFZ8CWstTn9Ook38wsGJ2Q7xww5LgkJ0yXId7ivqfu/mZmAemRmZ2R+x9Jnrpg3wDbvZqNegsAtzSdTXhbd2/c2rrrWVWdltn7+Jvvcj+/k9l9WYtH4ANgO+aHBAFgGbp73f4wMADz45I8AACAAZfkAQAADDjDBAAAMHCbvodpr7326gMPPHCtywAAANaps84662vdvffW19yy23RgOvDAA3PmmWeudRkAAMA6VVVf+W62d0keAADAgMAEAAAwIDABAAAM3KbvYQIAALbsxhtvzKZNm/KNb3xjrUtZFRs2bMjGjRuz0047reh+BSYAANgObdq0KbvuumsOPPDAVNValzNX3Z3LL788mzZtykEHHbSi+3ZJHgAAbIe+8Y1vZM8999zuw1KSVFX23HPPuZxNE5gAAGA7dXsIS5vN670KTAAAAAPuYQIAgNuD445b9f3d6U53yrXXXrvFZYceemgOOeSQnHTSSStb1wpzhgkAAFhV5557bm666aacfvrpue6669a6nCUJTAAAwKo68cQT80u/9Et5xCMekZNPPnmty1nSXANTVV1QVWdX1aer6sypbY+qOrWqvjQ97z61V1W9tqrOq6rPVtX95lkbAACwNt7xjnfkiU98Yp70pCflxBNPXOtylrQaZ5ge1t2Hdvdh0/xLknyou++R5EPTfJI8Msk9psexSV63CrUBAACr6Mwzz8xee+2Vu971rnn4wx+eT33qU/n617++1mUNrcUleUclOWGaPiHJ4xa0v6VnPpZkt6radw3qAwAA5uTEE0/MF77whRx44IE5+OCDc/XVV+c973nPWpc1NO/A1En+sarOqqpjp7Z9uvuiafriJPtM0/sn+eqCbTdNbbdQVcdW1ZlVdeZll102r7oBAIAVdvPNN+ed73xnzj777FxwwQW54IILcvLJJ6/ry/LmPaz4g7r7wqr6viSnVtUXFi7s7q6q3pYddvfxSY5PksMOO2ybtgUAgNutlR5WfBmuv/76bNy48dvzz3rWs7L//vtnv/32+3bbQx7ykHz+85/PRRddlH33XX8XmM01MHX3hdPzpVX1d0kekOSSqtq3uy+aLrm7dFr9wiQHLNh849QGAADcBt18883f0fbyl7/8FvM77LBDLr744tUqaZvN7ZK8qvreqtp183SSRyQ5J8kpSY6ZVjsmyeZxBE9J8tRptLzDk1y14NI9AACAVTfPM0z7JPm7qtr8On/b3R+oqk8keWdVPTPJV5I8YVr/75M8Ksl5Sa5P8vQ51gYAALBVcwtM3X1+kh/dQvvlSR6+hfZO8px51QMAALCt1mJYcQAAgNsEgQkAAGBAYAIAABiY9+8wAQAA68Bxpx23svt76Nb3d6c73SnXXnvtLdq++MUv5ld+5Vdy5ZVX5oYbbsiDH/zgHH/88UmSj3/843nhC1+YSy65JLvsskvuf//757WvfW122WWX/MM//ENe9rKX5frrr8/OO++cI444Iq985StX9D1ticC0klbix8DW4AfFAABgtfzar/1anv/85+eoo45Kkpx99tlJkksuuSRHH310TjrppDzwgQ9Mkrz73e/ONddck/PPPz/Pfe5z8/73vz+HHHJIbrrppm+HrHkTmAAAgFVz0UUXZePGjd+e/5Ef+ZEkyV/+5V/mmGOO+XZYSpLHP/7xSZIXvehF+e3f/u0ccsghSWY/dvvsZz97Vep1DxMAALBqnv/85+eII47IIx/5yLz61a/OlVdemSQ555xzcv/733+L2yy1bN4EJgAAYNU8/elPz7nnnpujjz46p512Wg4//PDccMMNa13WkMAEAACsqv322y/PeMYzcvLJJ2fHHXfMOeeck3vd614566yztrj+UsvmTWACAABWzQc+8IHceOONSZKLL744l19+efbff/8897nPzQknnJAzzjjj2+u+973vzSWXXJIXvehF+cM//MP8y7/8S5Lk5ptvzutf//pVqdegDwAAcDuwnGHAV9r1119/iwEeXvCCF2TTpk153vOelw0bNiRJ/vRP/zTf//3fnyQ56aST8sIXvjCXXnpp7nCHO+QhD3lIjjzyyOyzzz75sz/7szzpSU/K9ddfn6rKYx7zmFV5DwITAAAwFzfffPMW21/1qldtsf2BD3xgTj/99C0ue8xjHrNqIWkhl+QBAAAMCEwAAAADAhMAAGynunutS1g183qvAhMAAGyHNmzYkMsvv/x2EZq6O5dffvm3B5JYSQZ9AACA7dDGjRuzadOmXHbZZWtdyqrYsGHDLUbkWykCEwAAbId22mmnHHTQQWtdxm2eS/IAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgIG5B6aq2qGqPlVV75vmD6qqM6rqvKp6R1XdcWrfeZo/b1p+4LxrAwAAWMpqnGF6XpJzF8y/Ismru/vuSa5I8syp/ZlJrpjaXz2tBwAAsGbmGpiqamOSRyf5m2m+khyR5N3TKickedw0fdQ0n2n5w6f1AQAA1sS8zzD9WZLfTHLzNL9nkiu7+1vT/KYk+0/T+yf5apJMy6+a1r+Fqjq2qs6sqjMvu+yyedYOAADczs0tMFXVY5Jc2t1nreR+u/v47j6suw/be++9V3LXAAAAt7DjHPf9k0keW1WPSrIhyZ2TvCbJblW143QWaWOSC6f1L0xyQJJNVbVjkrskuXyO9QEAACxpbmeYuvul3b2xuw9M8sQkH+7upyT5SJLHT6sdk+TkafqUaT7T8g93d8+rPgAAgK1Zi99henGSF1TVeZndo/SGqf0NSfac2l+Q5CVrUBsAAMC3zfOSvG/r7tOSnDZNn5/kAVtY5xtJjl6NegAAAJZjLc4wAQAA3CYITAAAAAMCEwAAwIDABAAAMCAwAQAADAhMAAAAAwITAADAgMAEAAAwIDABAAAMCEwAAAADAhMAAMCAwAQAADAgMAEAAAwITAAAAAMCEwAAwIDABAAAMCAwAQAADAhMAAAAAwITAADAgMAEAAAwIDABAAAMCEwAAAADAhMAAMDAjmtdwPbkuJy2AvsAAADWC2eYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABuYWmKpqQ1V9vKo+U1Wfq6rfm9oPqqozquq8qnpHVd1xat95mj9vWn7gvGoDAABYjnmeYbohyRHd/aNJDk1yZFUdnuQVSV7d3XdPckWSZ07rPzPJFVP7q6f1AAAA1sxWA1NV/UlV3bmqdqqqD1XVZVX1i1vbrmeunWZ3mh6d5Igk757aT0jyuGn6qGk+0/KHV1Vtw3sBAABYUcs5w/SI7r46yWOSXJDk7kletJydV9UOVfXpJJcmOTXJl5Nc2d3fmlbZlGT/aXr/JF9Nkmn5VUn23MI+j62qM6vqzMsuu2w5ZQAAANwqywlMO07Pj07yru6+ark77+6buvvQJBuTPCDJIdte4nfs8/juPqy7D9t7772/290BAAAMLScwva+qvpDk/kk+VFV7J/nGtrxId1+Z5CNJHphkt6raHMI2Jrlwmr4wyQFJMi2/S5LLt+V1AAAAVtJWA1N3vyTJTyQ5rLtvTHJ9ZvcbLamq9q6q3abp70nyM0nOzSw4PX5a7ZgkJ0/Tp0zzmZZ/uLt7+W8FAABgZS1n0IddkvznJK+bmvZLctgy9r1vko9U1WeTfCLJqd39viQvTvKCqjovs3uU3jCt/4Yke07tL0jykm15IwAAACttx62vkjclOSuzs0zJ7NK5dyV531Ibdfdnk9x3C+3nZ3Y/0+L2byQ5ehn1AAAArIrl3MN0cHf/SZIbk6S7r09iuG8AAGC7t5zA9M3pHqROkqo6OLMfpQUAANiuLeeSvJcn+UCSA6rq7Ul+MsnT5lkUAADAerDVwNTdp1bVJ5McntmleM/r7q/NvTIAAIA1NgxMVXW/RU0XTc93raq7dvcn51cWAADA2lvqDNMrl1jWSY5Y4VoAAADWlWFg6u6HrWYhAAAA681W72Gqqg2Z/XDtgzI7s3R6ktdPv5sEAACw3VrOKHlvSXJNkj+f5p+c5K3xI7MAAMB2bjmB6d7dfc8F8x+pqs/PqyAAAID1Yjk/XPvJqjp880xV/XiSM+dXEgAAwPqwnDNM90/yf6vq/03zd03yxao6O0l3933mVh0AAMAaWk5gOnLuVQAAAKxDWw1M3f2Vqto9yQEL1/fDtQAAwPZuOcOK/36SpyX5cmbDiid+uBYAALgdWM4leU9IcnB3f3PexQAAAKwnyxkl75wku827EAAAgPVmOWeY/ijJp6rqnCQ3bG7s7sfOrSoAAIB1YDmB6YQkr0hydpKb51sOAADA+rGcwHR9d7927pUAAACsM8sJTKdX1R8lOSW3vCTPsOIAAMB2bTmB6b7T8+EL2gwrDgAAbPeW88O1D1uNQgAAANab5ZxhSlU9Osm9kmzY3Nbd/3VeRQEAAKwHW/0dpqp6fZJfSPJfklSSo5P8wJzrAgAAWHPL+eHan+jupya5ort/L8kDk/zgfMsCAABYe8sJTP8+PV9fVfsluTHJvvMrCQAAYH1Yzj1M76uq3ZL8aZJPZjZC3t/MtSoAAIB1YDmj5P3+NPmeqnpfkg3dfdV8ywIAAFh7yxn04eiq2nWafVGSN1XVfZfaBgAAYHuwnHuYXtbd11TVg5L8dJI3JHn9fMsCAABYe8sJTDdNz49Ocnx3vz/JHedXEgAAwPqwnMB0YVX9VWa/xfT3VbXzMrcDAAC4TVtO8HlCkg8m+dnuvjLJHpndywQAALBdW84oedcnee+C+YuSXDTPogAAANYDl9YBAAAMCEwAAAADAhMAAMDA8B6mqromSY+Wd/ed51IRAADAOjEMTN29a5JU1e9nNsjDW5NUkqck2XdVqgMAAFhDy7kk77Hd/T+6+5ruvrq7X5fkqHkXBgAAsNaWE5iuq6qnVNUOVXWHqnpKkuvmXRgAAMBaW05genJmP157yfQ4emoDAADYri3nh2sviEvwAACA26GtnmGqqh+sqg9V1TnT/H2q6nfmXxoAAMDaWs4leX+d5KVJbkyS7v5skifOsygAAID1YDmBaZfu/viitm/NoxgAAID1ZDmB6WtVdXCmH7Gtqsdn9rtMAAAA27WtDvqQ5DlJjk9ySFVdmORfM/vxWgAAgO3acgJTd/dPV9X3JrlDd19TVQfNuzAAAIC1tpxL8t6TJN19XXdfM7W9e34lAQAArA/DM0xVdUiSeyW5S1X9/IJFd06yYd6FAQAArLWlLsn7oSSPSbJbkp9b0H5NkmfNsygAAID1YBiYuvvkJCdX1QO7+59XsSYAAIB1YTn3MP1qVe22eaaqdq+qN86xJgAAgHVhOYHpPt195eaZ7r4iyX3nVxIAAMD6sJzAdIeq2n3zTFXtkeUNRw4AAHCbtpzg88ok/1xV75rmj07y3+ZXEgAAwPqw1cDU3W+pqjOTHDE1/Xx3f36+ZQEAAKy95VySlyR7JLmuu/8iyWVVddAcawIAAFgXthqYqurlSV6c5KVT005J3jbPogAAANaD5Zxh+k9JHpvkuiTp7n9Lsus8iwIAAFgPlhOYvtndnaSTpKq+d74lAQAArA/LCUzvrKq/SrJbVT0ryT8l+ev5lgUAALD2hqPkVdXO3X1Dd//3qvqZJFcn+aEkv9vdp65ahQAAAGtkqWHF/znJ/arqrd39S0mEJAAA4HZlqcB0x6p6cpKfqKqfX7ywu987v7IAAADW3lKB6VeTPCXJbkl+btGyTiIwAQAA27WlAtO+3f3sqvpUdx+/ahUBAACsE0uNkrf5h2p/dTUKAQAAWG+WOsN0eVX9Y5KDquqUxQu7+7HzKwsAAGDtLRWYHp3kfknemuSVq1MOAADA+jEMTN39zSQfq6qf6O7LVrEmAACAdWGpM0zfVlX/Pck9k2zY3NbdR8yrKAAAgPVgqUEfNnt7knOTHJTk95JckOQTc6wJAABgXVhOYNqzu9+Q5Mbu/l/d/Ywkzi4BAADbveVcknfj9HxRVT06yb8l2WN+JQEAAKwPywlMf1BVd0nyG0n+PMmdkzx/rlUBAACsA1sNTN39vmnyqiQPm285AAAA68dy7mECAAC4XRKYAAAABgQmAACAga0Gpqr6nQXTO8+3HAAAgPVjGJiq6sVV9cAkj1/Q/M/zLwkAAGB9WGqUvC8kOTrJ3arq9Gl+z6r6oe7+4qpUBwAAsIaWuiTvyiS/leS8JA9N8pqp/SVV9X/nXBcAAMCaW+oM088m+d0kByd5VZLPJrmuu5++GoUBAACsteEZpu7+re5+eJILkrw1yQ5J9q6q/11V/3OV6gMAAFgzS51h2uyD3X1mkjOr6tnd/aCq2mvehQEAAKy1rQ4r3t2/uWD2aVPb1+ZVEAAAwHqxTT9c292fmVchAAAA6802BSYAAIDbE4EJAABgQGACAAAYEJgAAAAGBCYAAICBuQWmqjqgqj5SVZ+vqs9V1fOm9j2q6tSq+tL0vPvUXlX12qo6r6o+W1X3m1dtAAAAyzHPM0zfSvIb3X3PJIcneU5V3TPJS5J8qLvvkeRD03ySPDLJPabHsUleN8faAAAAtmpugam7L+ruT07T1yQ5N8n+SY5KcsK02glJHjdNH5XkLT3zsSS7VdW+86oPAABga1blHqaqOjDJfZOckWSf7r5oWnRxkn2m6f2TfHXBZpumtsX7OraqzqyqMy+77LK51QwAADD3wFRVd0ryniS/3t1XL1zW3Z2kt2V/3X18dx/W3YftvffeK1gpAADALc01MFXVTpmFpbd393un5ks2X2o3PV86tV+Y5IAFm2+c2gAAANbEPEfJqyRvSHJud79qwaJTkhwzTR+T5OQF7U+dRss7PMlVCy7dAwAAWHU7znHfP5nkl5KcXVWfntp+K8kfJ3lnVT0zyVeSPGFa9vdJHpXkvCTXJ3n6HGsDAADYqrkFpu7+30lqsPjhW1i/kzxnXvUAAABsq1UZJQ8AAOC2SGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJmj9cTIAABHfSURBVAAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAICBHde6AAAAYH047rTjVmY/D12Z/awHzjABAAAMCEwAAAADAhMAAMCAwAQAADAgMAEAAAwITAAAAAMCEwAAwIDABAAAMCAwAQAADAhMAAAAAwITAADAgMAEAAAwIDABAAAMCEwAAAADAhMAAMCAwAQAADAgMAEAAAwITAAAAAMCEwAAwIDABAAAMCAwAQAADAhMAAAAAwITAADAgMAEAAAwIDABAAAMCEwAAAADAhMAAMCAwAQAADAgMAEAAAwITAAAAAMCEwAAwIDABAAAMDC3wFRVb6yqS6vqnAVte1TVqVX1pel596m9quq1VXVeVX22qu43r7oAAACWa55nmN6c5MhFbS9J8qHuvkeSD03zSfLIJPeYHscmed0c6wIAAFiWuQWm7v5okq8vaj4qyQnT9AlJHreg/S0987Eku1XVvvOqDQAAYDlW+x6mfbr7omn64iT7TNP7J/nqgvU2TW3foaqOraozq+rMyy67bH6VAgAAt3trNuhDd3eSvhXbHd/dh3X3YXvvvfccKgMAAJhZ7cB0yeZL7abnS6f2C5McsGC9jVMbAADAmlntwHRKkmOm6WOSnLyg/anTaHmHJ7lqwaV7AAAAa2LHee24qk5M8tAke1XVpiQvT/LHSd5ZVc9M8pUkT5hW//skj0pyXpLrkzx9XnUBAAAs19wCU3c/abDo4VtYt5M8Z161AAAA3BprNugDAADAeicwAQAADAhMAAAAAwITAADAgMAEAAAwIDABAAAMCEwAAAADAhMAAMCAwAQAADAgMAEAAAwITAAAAAMCEwAAwIDABAAAMCAwAQAADAhMAAAAAwITAADAgMAEAAAwIDABAAAMCEwAAAADAhMAAMCAwAQAADAgMAEAAAwITAAAAAMCEwAAwIDABAAAMCAwAQAADAhMAAAAAwITAADAgMAEAAAwIDABAAAMCEwAAAADO651AQAAwDpx2mkrs5+Hrsxu1gNnmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABgQmAACAAYEJAABgQGACAAAYEJgAAAAGBCYAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCQAAYEBgAgAAGBCYAAAABtZVYKqqI6vqi1V1XlW9ZK3rAQAAbt/WTWCqqh2S/GWSRya5Z5InVdU917YqAADg9mzHtS5ggQckOa+7z0+SqjopyVFJPr+mVa22445bX/sBAIDbseruta4hSVJVj09yZHf/8jT/S0l+vLufu2i9Y5McO83+UJIvrmqhq2OvJF9b6yJuB/Tz6tDP86ePV4d+Xh36ef708erQz6tjOf38A9299619gfV0hmlZuvv4JMevdR3zVFVndvdha13H9k4/rw79PH/6eHXo59Whn+dPH68O/bw6VqOf1809TEkuTHLAgvmNUxsAAMCaWE+B6RNJ7lFVB1XVHZM8Mckpa1wTAABwO7ZuLsnr7m9V1XOTfDDJDkne2N2fW+Oy1sp2fcnhOqKfV4d+nj99vDr08+rQz/Onj1eHfl4dc+/ndTPoAwAAwHqzni7JAwAAWFcEJgAAgAGBaYVV1Rur6tKqOmdB2x5VdWpVfWl63n1qr6p6bVWdV1Wfrar7LdjmmGn9L1XVMQva719VZ0/bvLaqaqnX2B4N+vi4qrqwqj49PR61YNlLp/76YlX97IL2I6e286rqJQvaD6qqM6b2d0yDkKSqdp7mz5uWH7g673htVNUBVfWRqvp8VX2uqp43tTueV8gSfex4XkFVtaGqPl5Vn5n6+fem9m3um5Xq/+3REv385qr61wXH86FTu78Zt1JV7VBVn6qq903zjuU52EI/O5ZXWFVdMPXDp6vqzKlt/X3P6G6PFXwkeUiS+yU5Z0HbnyR5yTT9kiSvmKYfleQfklSSw5OcMbXvkeT86Xn3aXr3adnHp3Vr2vaRS73G9vgY9PFxSV64hXXvmeQzSXZOclCSL2c2qMgO0/TdktxxWuee0zbvTPLEafr1SZ49Tf/nJK+fpp+Y5B1r3Rdz7ud9k9xvmt41yb9M/el4nn8fO55Xtp8ryZ2m6Z2SnDEdd9vUNyvZ/9vjY4l+fnOSx29hfX8zbn1fvyDJ3yZ531LHmWN5xfvZsbzyfXxBkr0Wta277xnOMK2w7v5okq8vaj4qyQnT9AlJHreg/S0987Eku1XVvkl+Nsmp3f317r4iyalJjpyW3bm7P9azT/gti/a1pdfY7gz6eOSoJCd19w3d/a9JzkvygOlxXnef393fTHJSkqOmf3k4Ism7p+0Xf16b+/jdSR6++V8qtkfdfVF3f3KavibJuUn2j+N5xSzRxyOO51thOiavnWZ3mh6dbe+blez/7c4S/Tzib8atUFUbkzw6yd9M87fmv3PH8lYs7uetcCyvrHX3PUNgWh37dPdF0/TFSfaZpvdP8tUF622a2pZq37SF9qVe4/bkudMp2jcuOLW6rX28Z5Iru/tbi9pvsa9p+VXT+tu96TKO+2b2L8aO5zlY1MeJ43lFTZfWfDrJpZn9z/TL2fa+Wcn+3y4t7ufu3nw8/7fpeH51Ve08tfmbcev8WZLfTHLzNH9r/jt3LG/d4n7ezLG8sjrJP1bVWVV17NS27r5nCEyrbEq4cx3LfTVeYx16XZKDkxya5KIkr1zbcrYfVXWnJO9J8uvdffXCZY7nlbGFPnY8r7Duvqm7D02yMbN/RT9kjUvaLi3u56q6d5KXZtbfP5bZJTMvnnMN2+3fjKp6TJJLu/usta5le7ZEPzuWV96Duvt+SR6Z5DlV9ZCFC9fL9wyBaXVcMp0WzPR86dR+YZIDFqy3cWpbqn3jFtqXeo3bhe6+ZPof9c1J/jqzL0TJtvfx5Zmd4t1xUfst9jUtv8u0/narqnbK7Iv827v7vVOz43kFbamPHc/z091XJvlIkgdm2/tmJft/u7agn4+cLj3t7r4hyZty649nfzOSn0zy2Kq6ILPL5Y5I8po4llfad/RzVb3NsbzyuvvC6fnSJH+XWZ+uu+8ZAtPqOCXJ5hE7jkly8oL2p06jfhye5Krp9OAHkzyiqnafLsV5RJIPTsuurqrDp+uJn7poX1t6jduFzQf95D8l2TyC3ilJnlizkYIOSnKPzG4A/ESSe9Rs1J87ZnYz7CnTvzJ8JMnjp+0Xf16b+/jxST48rb9dmo6xNyQ5t7tftWCR43mFjPrY8byyqmrvqtptmv6eJD+T2f1i29o3K9n/251BP39hwZeSyuw+gYXHs78Z26C7X9rdG7v7wMyOsw9391PiWF5Rg37+Rcfyyqqq762qXTdPZ9Y/52Q9fs/odTBCxvb0SHJiZpfQ3JjZtZLPzOza3w8l+VKSf0qyx7RuJfnLzK6lPzvJYQv284zMbsI8L8nTF7QfNh1MX07yF0lqat/ia2yPj0Efv3Xqw89O/xHsu2D9357664uZRkeZ2h+V2ahkX07y2wva75bZ/zjOS/KuJDtP7Rum+fOm5Xdb676Ycz8/KLNT1J9N8unp8SjH86r0seN5Zfv5Pkk+NfXnOUl+99b2zUr1//b4WKKfPzwdz+ckeVv+YyQ9fzO+u/5+aP5j9DbH8ur0s2N5Zfv2bpmNxPiZJJ/bfLyN+mAt+3nzRgAAACzikjwAAIABgQkAAGBAYAIAABgQmAAAAAYEJgAAgAGBCYChqvqjqnpYVT2uql66gvt9WlX9xUrtb5mv+diqesk0/biquueCZadV1WFb2f7Aquqq+i8L2v6iqp62le2Oq6rrq+r7FrRde6vfCACrSmACYCk/nuRjSX4qyUe3ZcOq2nEuFd1K3X1Kd//xNPu4JPdcav2BS5M8b/pRz23xtSS/cSteD4A1JjAB8B2q6k+r6rNJfizJPyf55SSvq6rfnZafVlWvqapPV9U5VfWAqf24qnprVf2fJG+tqg1V9aaqOruqPlVVD1vwMgdM+/lSVb18wWv/YlV9fNr3X1XVDtPjzdNrnV1Vz19U7w5V9a/TL8DvVlU3VdVDpmUfrap7bD6rVVU/keSxSf50eo2Dp90cPb3uv1TVgwddc1lmP3Z4zOIFVfWsqvpEVX2mqt5TVbssWPzGJL9QVXss2uZ7q+r90zbnVNUvLP3JALDaBCYAvkN3vyjJM5O8ObPQ9Nnuvk93/9cFq+3S3Ycm+c+ZBYLN7pnkp7v7SUmeM9td/0iSJyU5oao2TOs9IMn/l+Q+mYWVw6rqh5P8QpKfnPZ9U5KnJDk0yf7dfe9pX29aVO9NSb44vfaDknwyyYOrauckB3T3lxas+3+TnJLkRd19aHd/eVq0Y3c/IMmvJ3l5xl6R5IVVtcOi9vd29491948mOXfqv82unfroeYu2OTLJv3X3j3b3vZN8YInXBWANCEwAjNwvyWeSHJJZAFjsxCTp7o8muXNV7Ta1n9Ld/z5NPyjJ26b1vpDkK0l+cFp2andfPq373mndhye5f5JPVNWnp/m7JTk/yd2q6s+r6sgkV2+hntOTPGR6/NG0vx9L8ollvt/3Ts9nJTlwtFJ3n5/kjCRPXrTo3lV1elWdnVnIu9ei5a9NckxV7bqg7ewkP1NVr6iqB3f3VcusFYBVIjABcAtVdegUVv7/du4XNKswiuP49wwHGgYiWkS0qUU0i4JtSTAMxCYWFcQ0sExQcBbBaFFQYYhlmDQsyFAsImwwWFg1GBSmiG5D3TE8z4XXd/fOIYhDvp9yuc97/77txzn3jAOjwFNguLavbes5NPtObfa/bPBWbecH8LBWfo5k5oHMvJaZi8BhYBq4ANxrud4L4DilcvUM2A6coASpjVip2x/A776/uglcqc/beABcqhWw68DW3hMy8yPwiFJ1a9YWKMF0DrjRtDxKkjYPA5Mk6ReZOVvb4RYoLW7PgeEaYJZ6Dj0NEBHHgE8d1ZGXlGoLEbEf2EtpnYNSWdlRQ9gp4BXl+6CRZqJc/X1fROwEBjJzEhijhIx+r4GjwGpmLgOzwHnah1V8BoZa1jekVsvmgZM9y0PAu4gYbN65xe36TFsAImI38DUzJ4BbtL+XJOkf2lQTjCRJm0NE7AIWM3M1Ig5m5nzLYcsRMQMMAuc6LnWHMixiDvgOnM3MlYiAEnAmgT3ARGa+qfceA6YiYgD4RqnILAH36xrAmhHn9bpvKVP9oIS1M5TqTb/HwN2IuAyMrPdfrGMcmOnZv0pp1Xtft2sCWWZ+iIgnQDO04hBl+MQq5V0v/uGzSJL+ksjs74iQJGl9ETENjDYhR5Kk/5UteZIkSZLUwQqTJEmSJHWwwiRJkiRJHQxMkiRJktTBwCRJkiRJHQxMkiRJktTBwCRJkiRJHX4C+wz+AVfVdXkAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1008x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(14,10))\n",
    "ax.hist(x=Meth_data['LA'].isna().sum(axis=0).values, bins=40, color=['red'], stacked=True, label='LA', alpha=0.5)\n",
    "ax.hist(x=Meth_data['LSCC'].isna().sum(axis=0).values, bins=40, color=['green'], stacked=True, label='LSCC', \n",
    "            alpha=0.5)\n",
    "\n",
    "plt.legend()\n",
    "plt.title(\"# NaN's per sample, distribution\")\n",
    "plt.xlabel('#probes with NaNs')\n",
    "plt.ylabel('# affected samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, '# affected probes')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA1YAAAJcCAYAAAD6sDW5AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nOzde9hmZV03/O9PQNDcsJsIGHggJXnRlIQQ2/gaFA5mYT5ioK9ORtJGXtPSUMsHTOnNt5SyzKIk0QoktAeeQolMyjYoY5LsJCfExyE24wCCEsjm9/xxrbGL2/u+54Y11ww3fj7HcR33Wr91nuc614WHx3yPtdZ5VXcHAACAB+8RW3sCAAAAy51gBQAAMJJgBQAAMJJgBQAAMJJgBQAAMJJgBQAAMJJgBQBLUFUnV9WfbKFzvbeq3jpsf39VXb0Zx/5wVa0etn+iqv5hM479kqr66801HsByIlgBLENV9cmq+o6q+vaq+pdNtO2quqyqHjFVe2tVvXeJ53rvMMYhU7UnVlVP7V9UVc9+4FfCpnT3x7v7SZtqt9Tg191HdvcZY+dVVfsM/7vYdmrsP+3uI8aODbAcCVYAy0xVbZfkvyX5XJKDkiwarAZ7JDlmxGlvTvLWEf1navof9w+yf00Hz4ejb4ZrBNia/B8swPLzlCRXdncnOThLC1b/f5I3LxRAqurPq+qGqvpyVf19VT15TpMzkjy1qv7vTZ2oqg6pqjVVdVtV3VhV71ig3bOral1VvbGqvlRV11bVS6aOb19Vv1lV/3sY5/er6lFz+p5YVTck+eN5xv+JqvrHqvrd4bo+W1WHTx2/qKpOqap/THJHkm+vqj2q6ryqurmq1lbVK+YMu0NVfaCqbq+qf6mqp02Nt0dVfbCq1lfV56vqVZv6rqb6ftcw3u1V9YEkO8z9nqb2T6yq64a2V1fV4VW1Kskbk/x4VX2lqv51kWu8qKp+6v6nX/A7uraqfnBqf/qu2N8Pf28dzvnMuY8WVtX3VNUlw9iXVNX3zPn+3zL8N7q9qv66qnZd6ncG8FAjWAEsE1X18qq6Nck/JnnmsP2LSd5WVbdW1b6LdP9QktuS/MQCxz+cZL8k35pJUPvTOcfvSPJrSU6Zr3N3P7u7Lxp2fzvJb3f345I8IcnZi8zr25LsmmTPJKuTnFZVGx97+/Uk35HkwCRPHNr8jzl9d87k7t3xC4z/jCT/PpzjpCQfqqqdp46/dOj72CRfSHJWknWZ3OF7YZJfq6rDptofleTPh/P+WZL/WVXbDXeC/leSfx3meXiSV1fVcxa59iRJVT0yyf9M8v5h3D9P8t8XaPukJCck+e7ufmyS5yS5trs/ksl/nw9092O6+2lT3eZe4wP9jhbyrOHvjsM5/3nOXHdO8ldJ3plklyTvSPJXVbXLVLMXJ3l5Jv+7e2SS1y7hvAAPSYIVwDLR3X/c3Tsm+VSSQ5M8NcnlSR7X3Tt29+cX657kTUneNPxDfu7Yp3f37d19V5KTkzytqh4/p9kfJNm7qo7cxFTvTvLEqtq1u7/S3Rdvov2buvuu7v67TP4h/qKqqkzCwGu6++buvj2T4DD9OON9SU4a+v7nAmPflOS3uvvu7v5AkquT/PDU8fd29xXdfU8mQe17k5zY3Xd296VJ/ijJy6baf6q7z+nuuzMJCjtk8t/iu5Os6O5f7e6vdfc1Sf4wS3v88tAk203N85wklyzQ9t4k2yc5oKq26+5ru/vfNzH+169xmPdcm/qOHqwfTvK57n7/cO4zk3w2yY9Mtfnj7v634b/f2ZmEaIBlSbACWAaqaufhrtSXk3xPkosy+Qfwk5LcUlWv3tQY3X1+JndjfnrO2NtU1a9X1b9X1W1Jrh0O7Tqn/11J3jJ8FnNcJneaPjs8/vW8Rdre0t1fndr/QiZ3i1YkeXSSTw3XfWuSjwz1jdZ3952bmMt1wyOTc8ff6ItT23sk2RjiptvvOV/77r4v/3V3678l2WPjXIf5vjHJbpuY38bzzjfPb9Dda5O8OpPwe1NVnVVVe8zXdr45L2BT39GDtUe+8Trmfp83TG3fkeQxm+G8AFuFYAWwDAx3bXbMJBT90bD9kSQ/Mtyt+q0lDvXLmfyD/9FTtRdn8ojbDyZ5fJJ9hnrN0/+Pk+yY5AWLzPVz3X1sJo93vS3JOVX1LQs032nOsb2T/EeSLyX5zyRPHq5vx+5+fHdP/8N7OgwsZM/h7tfc8ecb4z+S7FxVj53T/rqp/b02bgyP/60c+n0xyeen5rpjdz+2u5+7hDlev8A859Xdf9bd35dJmOtMvuO513K/Lps4/2Lf0Vdz//+tfNsDGPc/hjlOm/t9AjxsCFYAy8v0KoDflcljgUs2vAd1eSbvM2302CR3JdmQyT+if22R/vdk8h7OiQu1qar/p6pWDHd0bh3K9y0yrTdX1SOr6vuTPC/Jnw99/zDJqVX1rcO4ey7lnaU5vjXJq4b3oI5O8n8lOX+Ba/tikn9K8v9V1Q5V9dRM7r5NL2F+UFW9oCaLgLw6k+/t4iSfTHL7sLDEo4a7gE+pqu8e5v7smlqefo5/TnLP1DxfkOSQ+RpW1ZOq6rCq2j7JnZmEz43f7Y1J9qkHvvLfYt/RpUmOGY4dnMl7ZxutH8797QuMe36S76iqF1fVtlX140kOSPKXD3B+AMuCYAWwvByU5F+GBQDu7e5bHsQYv5LJIgkbvS+TR7SuS3JlJkFhMWdmcpdlIauSXFFVX8lkIYtjFnkH6oYkt2Ryd+NPk/xMd392OHZikrVJLh4eUfybTB59fCA+kcmiHF/KZOGNF3b3hkXaH5vJHbv/SPIXmbzD9TdTx89N8uPDnF+a5AXDu0n3ZhIKD0zy+eF8f5TJHcBkcqfrn+Y7YXd/LZM7gD+RybL2P57JYiPz2T6TRT2+lMl3961J3jAc+/Ph74baxG+bzbHYd/SmTBYguSXJmzNZsGPjvO8Y2v/j8PjjoXOua0Mm38kvZhLafynJ87r7Sw9gbgDLRt3/sWoA2DJq8oPCf9LdK2c0/k8k+anhsbmtqqr+KJM7cRds7bkAMBujflARANi07v6pTbcCYDnzKCAAAMBIHgUEAAAYyR0rAACAkbxjNdh11117n3322drTAAAAHqI+9alPfam7V8x3TLAa7LPPPlmzZs3WngYAAPAQVVVfWOiYRwEBAABGEqwAAABGEqwAAABG8o4VAACwqLvvvjvr1q3LnXfeubWnskXssMMOWblyZbbbbrsl9xGsAACARa1bty6Pfexjs88++6SqtvZ0Zqq7s2HDhqxbty777rvvkvt5FBAAAFjUnXfemV122eVhH6qSpKqyyy67POC7c4IVAACwSd8MoWqjB3OtghUAAMBI3rECAAAemJNP3uLjPeYxj8lXvvKVeY8deOCB2X///XPWWWdt3nk9AO5YAQAAy9ZVV12Ve++9Nx//+Mfz1a9+davNQ7ACAACWrTPPPDMvfelLc8QRR+Tcc8/davMQrAAAgGXrAx/4QI455pgce+yxOfPMM7faPAQrAABgWVqzZk123XXX7L333jn88MPz6U9/OjfffPNWmcvMg1VVbVNVn66qvxz2962qT1TV2qr6QFU9cqhvP+yvHY7vMzXGG4b61VX1nKn6qqG2tqpeP1Wf9xwAAMDDx5lnnpnPfvaz2WefffKEJzwht912Wz74wQ9ulblsiTtWP5/kqqn9tyU5tbufmOSWJMcN9eOS3DLUTx3apaoOSHJMkicnWZXk94awtk2SdyU5MskBSY4d2i52DgAA4GHgvvvuy9lnn53LLrss1157ba699tqce+65W+1xwJkut15VK5P8cJJTkvxCTX5p67AkLx6anJHk5CTvTnLUsJ0k5yT53aH9UUnO6u67kny+qtYmOWRot7a7rxnOdVaSo6rqqkXOAQAAjLW5l1tfgjvuuCMrV678+v4rXvGK7Lnnntljjz2+XnvWs56VK6+8Mtdff3123333LTq/Wf+O1W8l+aUkjx32d0lya3ffM+yvS7LnsL1nki8mSXffU1VfHtrvmeTiqTGn+3xxTv0ZmzjH/VTV8UmOT5K99977QVweAACwJdx3333fUDvppJPut7/NNtvkhhtu2FJTup+ZPQpYVc9LclN3f2pW5xiru0/r7oO7++AVK1Zs7ekAAADL1CzvWH1vkh+tqucm2SHJ45L8dpIdq2rb4Y7SyiTXDe2vS7JXknVVtW2SxyfZMFXfaLrPfPUNi5wDAABgs5vZHavufkN3r+zufTJZfOJvu/slST6W5IVDs9VJNv6K13nDfobjf9vdPdSPGVYN3DfJfkk+meSSJPsNKwA+cjjHeUOfhc4BAACw2W2N37E6MZOFLNZm8j7Ue4b6e5LsMtR/Icnrk6S7r0hydpIrk3wkySu7+97hbtQJSS7IZNXBs4e2i50DAABgs5v14hVJku6+KMlFw/Y1+a9V/abb3Jnk6AX6n5LJyoJz6+cnOX+e+rznAAAAmIWtcccKAADgYWWL3LECAAAePk6+6OTNO96zNz3eYx7zmHzlK1+5X+3qq6/OT//0T+fWW2/NXXfdle///u/PaaedliT55Cc/mde+9rW58cYb8+hHPzoHHXRQ3vnOd+bRj350PvzhD+dNb3pT7rjjjmy//fY57LDD8va3v33UNQhWsAkP9v84lvJ/EAAAPHivetWr8prXvCZHHXVUkuSyyy5Lktx44405+uijc9ZZZ+WZz3xmkuScc87J7bffnmuuuSYnnHBC/uqv/ir7779/7r333q+HsTEEq4eizfVL1lvhF7EBAGBLuf7667Ny5cqv73/nd35nkuRd73pXVq9e/fVQlSQvfOFk0fDXve51+eVf/uXsv//+SSY/KvyzP/uzo+fiHSsAAGBZes1rXpPDDjssRx55ZE499dTceuutSZLLL788Bx100Lx9Fjs2hmAFAAAsSy9/+ctz1VVX5eijj85FF12UQw89NHfddddWmYtgBQAALFt77LFHfvInfzLnnntutt1221x++eV58pOfnE996lPztl/s2BiCFQAAsCx95CMfyd13350kueGGG7Jhw4bsueeeOeGEE3LGGWfkE5/4xNfbfuhDH8qNN96Y173udfm1X/u1/Nu//VuS5L777svv//7vj56LxSsAAIAHZGusfnzHHXfcb6GKX/iFX8i6devy8z//89lhhx2SJL/xG7+Rb/u2b0uSnHXWWXnta1+bm266KY94xCPyrGc9K6tWrcpuu+2W3/qt38qxxx6bO+64I1WV5z3veaPnJ1gBAAAPeffdd9+89Xe84x3z1p/5zGfm4x//+LzHnve8522WMDXNo4AAAAAjCVYAAAAjCVYAAMAmdffWnsIW82CuVbACAAAWtcMOO2TDhg3fFOGqu7Nhw4avL4ixVBavAAAAFrVy5cqsW7cu69ev39pT2SJ22GGH+61AuBSCFQAAsKjtttsu++6779aexkOaRwEBAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGEqwAAABGmlmwqqodquqTVfWvVXVFVb15qL+3qj5fVZcOnwOHelXVO6tqbVV9pqqePjXW6qr63PBZPVU/qKouG/q8s6pqqO9cVRcO7S+sqp1mdZ0AAACzvGN1V5LDuvtpSQ5MsqqqDh2Ova67Dxw+lw61I5PsN3yOT/LuZBKSkpyU5BlJDkly0lRQeneSV0z1WzXUX5/ko929X5KPDvsAAAAzMbNg1RNfGXa3Gz69SJejkrxv6Hdxkh2ravckz0lyYXff3N23JLkwk5C2e5LHdffF3d1J3pfk+VNjnTFsnzFVBwAA2Oxm+o5VVW1TVZcmuSmTcPSJ4dApw+N+p1bV9kNtzyRfnOq+bqgtVl83Tz1Jduvu64ftG5LstsD8jq+qNVW1Zv369Q/uIgEAgG96Mw1W3X1vdx+YZGWSQ6rqKUnekGT/JN+dZOckJ854Dp0F7pR192ndfXB3H7xixYpZTgMAAHgY2yKrAnb3rUk+lmRVd18/PO53V5I/zuS9qSS5LsleU91WDrXF6ivnqSfJjcOjghn+3rR5rwgAAOC/zHJVwBVVteOw/agkP5Tks1OBpzJ59+nyoct5SV42rA54aJIvD4/zXZDkiKraaVi04ogkFwzHbquqQ4exXpbk3KmxNq4euHqqDgAAsNltO8Oxd09yRlVtk0mAO7u7/7Kq/raqViSpJJcm+Zmh/flJnptkbZI7krw8Sbr75qp6S5JLhna/2t03D9s/l+S9SR6V5MPDJ0l+PcnZVXVcki8kedHMrhIAAPimN7Ng1d2fSfJd89QPW6B9J3nlAsdOT3L6PPU1SZ4yT31DksMf4JQBAAAelC3yjhUAAMDDmWAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAwkmAFAAAw0syCVVXtUFWfrKp/raorqurNQ33fqvpEVa2tqg9U1SOH+vbD/trh+D5TY71hqF9dVc+Zqq8aamur6vVT9XnPAQAAMAuzvGN1V5LDuvtpSQ5MsqqqDk3ytiSndvcTk9yS5Lih/XFJbhnqpw7tUlUHJDkmyZOTrErye1W1TVVtk+RdSY5MckCSY4e2WeQcAAAAm93MglVPfGXY3W74dJLDkpwz1M9I8vxh+6hhP8Pxw6uqhvpZ3X1Xd38+ydokhwyftd19TXd/LclZSY4a+ix0DgAAgM1upu9YDXeWLk1yU5ILk/x7klu7+56hybokew7beyb5YpIMx7+cZJfp+pw+C9V3WeQcc+d3fFWtqao169evH3OpAADAN7GZBqvuvre7D0yyMpM7TPvP8nwPVHef1t0Hd/fBK1as2NrTAQAAlqktsipgd9+a5GNJnplkx6radji0Msl1w/Z1SfZKkuH445NsmK7P6bNQfcMi5wAAANjsZrkq4Iqq2nHYflSSH0pyVSYB64VDs9VJzh22zxv2Mxz/2+7uoX7MsGrgvkn2S/LJJJck2W9YAfCRmSxwcd7QZ6FzAAAAbHbbbrrJg7Z7kjOG1fsekeTs7v7LqroyyVlV9dYkn07ynqH9e5K8v6rWJrk5k6CU7r6iqs5OcmWSe5K8srvvTZKqOiHJBUm2SXJ6d18xjHXiAucAAADY7GYWrLr7M0m+a576NZm8bzW3fmeSoxcY65Qkp8xTPz/J+Us9BwAAwCxskXesAAAAHs4EKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJFmFqyqaq+q+lhVXVlVV1TVzw/1k6vquqq6dPg8d6rPG6pqbVVdXVXPmaqvGmprq+r1U/V9q+oTQ/0DVfXIob79sL92OL7PrK4TAABglnes7knyi919QJJDk7yyqg4Yjp3a3QcOn/OTZDh2TJInJ1mV5Peqapuq2ibJu5IcmeSAJMdOjfO2YawnJrklyXFD/bgktwz1U4d2AAAAMzGzYNXd13f3vwzbtye5Ksmei3Q5KslZ3X1Xd38+ydokhwyftd19TXd/LclZSY6qqkpyWJJzhv5nJHn+1FhnDNvnJDl8aA8AALDZbZF3rIZH8b4rySeG0glV9ZmqOr2qdhpqeyb54lS3dUNtofouSW7t7nvm1O831nD8y0P7ufM6vqrWVNWa9evXj7pGAADgm9fMg1VVPSbJB5O8urtvS/LuJE9IcmCS65O8fdZzWEh3n9bdB3f3wStWrNha0wAAAJa5mQarqtouk1D1p939oSTp7hu7+97uvi/JH2byqF+SXJdkr6nuK4faQvUNSXasqm3n1O831nD88UN7AACAzW6WqwJWkvckuaq73zFV332q2Y8luXzYPi/JMcOKfvsm2S/JJ5NckmS/YQXAR2aywMV53d1JPpbkhUP/1UnOnRpr9bD9wiR/O7QHAADY7LbddJMH7XuTvDTJZVV16VB7Yyar+h2YpJNcm+Snk6S7r6iqs5NcmcmKgq/s7nuTpKpOSHJBkm2SnN7dVwzjnZjkrKp6a5JPZxLkMvx9f1WtTXJzJmEMAABgJmYWrLr7H5LMtxLf+Yv0OSXJKfPUz5+vX3dfk/96lHC6fmeSox/IfAEAAB6sLbIqIAAAwMOZYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADDSJoNVVT2hqrYftp9dVa+qqh1nPzUAAIDlYSl3rD6Y5N6qemKS05LsleTPZjorAACAZWQpweq+7r4nyY8l+Z3ufl2S3Wc7LQAAgOVjKcHq7qo6NsnqJH851Lab3ZQAAACWl6UEq5cneWaSU7r781W1b5L3z3ZaAAAAy8e2m2rQ3VdW1YlJ9h72P5/kbbOeGAAAwHKxlFUBfyTJpUk+MuwfWFXnzXpiAAAAy8VSHgU8OckhSW5Nku6+NMm3z3BOAAAAy8qSFq/o7i/Pqd03i8kAAAAsR5t8xyrJFVX14iTbVNV+SV6V5J9mOy0AAIDlYyl3rP7fJE9OcleSM5PcluTVs5wUAADAcrKUVQHvSPLLVfW2yW7fPvtpAQAALB9LWRXwu6vqsiSfSXJZVf1rVR00+6kBAAAsD0t5x+o9SX6uuz+eJFX1fUn+OMlTZzkxAACA5WIp71jduzFUJUl3/0OSe2Y3JQAAgOVlwTtWVfX0YfPvquoPMlm4opP8eJKLZj81AACA5WGxRwHfPmf/pKntnsFcAAAAlqUFg1V3/8CWnAgAAMBytZRVAR9fVe+oqjXD5+1V9fgtMTkAAIDlYCmLV5ye5PYkLxo+t2WyKiAAAABZ2nLrT+ju/z61/+aqunRWEwIAAFhulnLH6j+H365KklTV9yb5z9lNCQAAYHlZyh2rn0nyvqn3qm5Jsnp2UwIAAFheFg1WVfWIJE/q7qdV1eOSpLtv2yIzAwAAWCYWfRSwu+9L8kvD9m1CFQAAwDdayjtWf1NVr62qvapq542fmc8MAABgmVjKO1Y/Pvx95VStk3z75p8OAADA8rPJYNXd+26JiQAAACxXmwxWVbVDkp9L8n2Z3Kn6eJLf7+47Zzw3AACAZWEpjwK+L8ntSX5n2H9xkvcnOXpWkwIAAFhOlhKsntLdB0ztf6yqrpzVhAAAAJabpawK+C9VdejGnap6RpI1s5sSAADA8rKUO1YHJfmnqvrfw/7eSa6uqsuSdHc/dWazAwAAWAaWEqxWzXwWAAAAy9hSllv/wpaYCAAAwHK1lHesAAAAWIRgBQAAMJJgBQAAMNKC71hV1e1JeqHj3f24mcwIAABgmVkwWHX3Y5Okqt6S5Pok709SSV6SZPctMjsAAIBlYCmPAv5od/9ed9/e3bd197uTHDXriQEAACwXSwlWX62ql1TVNlX1iKp6SZKvznpiAAAAy8VSgtWLk7woyY3D5+ihBgAAQJb2A8HXxqN/AAAAC9rkHauq+o6q+mhVXT7sP7WqfmX2UwMAAFgelvIo4B8meUOSu5Okuz+T5JhZTgoAAGA5WUqwenR3f3JO7Z5ZTAYAAGA5Wkqw+lJVPSHDjwVX1Qsz+V0rAAAAsoTFK5K8MslpSfavquuSfD6THwkGAAAgS7tj1d39g0lWJNm/u79vKf2qaq+q+lhVXVlVV1TVzw/1navqwqr63PB3p6FeVfXOqlpbVZ+pqqdPjbV6aP+5qlo9VT+oqi4b+ryzqmqxcwAAAMzCUoLVB5Oku7/a3bcPtXOW0O+eJL/Y3QckOTTJK6vqgCSvT/LR7t4vyUeH/SQ5Msl+w+f4JO9OJiEpyUlJnpHkkCQnTQWldyd5xVS/VUN9oXMAAABsdgs+ClhV+yd5cpLHV9ULpg49LskOmxq4u6/P8C5Wd99eVVcl2TOT38R69tDsjCQXJTlxqL+vuzvJxVW1Y1XtPrS9sLtvHuZ1YZJVVXVRksd198VD/X1Jnp/kw4ucAwAAYLNb7B2rJyV5XpIdk/zIVP32TO4SLVlV7ZPku5J8IsluQ+hKkhuS7DZs75nki1Pd1g21xerr5qlnkXPMndfxmdwdy9577/1ALgkAAODrFgxW3X1uknOr6pnd/c8P9gRV9ZhMHid8dXffNrwGtfEcXVX9YMdeisXO0d2nZbIwRw4++OCZzk+ZEo0AABhsSURBVAMAAHj4Wso7Vj9TVTtu3Kmqnarq9KUMXlXbZRKq/rS7PzSUbxwe8cvw96ahfl2Svaa6rxxqi9VXzlNf7BwAAACb3VKC1VO7+9aNO919SyaP9S1qWKHvPUmu6u53TB06L8nGlf1WJzl3qv6yYXXAQ5N8eXic74IkRwyBbqckRyS5YDh2W1UdOpzrZXPGmu8cAAAAm91SfsfqEVW10xCoNq7St5R+35vkpUkuq6pLh9obk/x6krOr6rgkX0jyouHY+Umem2RtkjuSvDxJuvvmqnpLkkuGdr+6cSGLJD+X5L1JHpXJohUfHuoLnQMAAGCzW0pAenuSf66qPx/2j05yyqY6dfc/JKkFDh8+T/vO5MeI5xvr9CTf8Phhd69J8pR56hvmOwcAAMAsbDJYdff7qmpNksOG0gu6+8rZTgsAAGD5WMo7Vkmyc5KvdvfvJllfVfvOcE4AAADLyiaDVVWdlMmP675hKG2X5E9mOSkAAIDlZCl3rH4syY8m+WqSdPd/JHnsLCcFAACwnCwlWH1tWFiik6SqvmW2UwIAAFhelhKszq6qP0iyY1W9IsnfJPnD2U4LAABg+VhwVcCq2r677+ru36yqH0pyW5InJfkf3X3hFpshAADAQ9xiy63/c5KnV9X7u/ulSYQpAACAeSwWrB5ZVS9O8j1V9YK5B7v7Q7ObFgAAwPKxWLD6mSQvSbJjkh+Zc6yTCFYAAABZPFjt3t0/W1Wf7u7TttiMAAAAlpnFVgXc+IPAP7MlJgIAALBcLXbHakNV/XWSfavqvLkHu/tHZzctAACA5WOxYPXDSZ6e5P1J3r5lpgMAALD8LBisuvtrSS6uqu/p7vVbcE4AAADLymJ3rL6uqn4zyQFJdthY6+7DZjUpAACA5WSxxSs2+tMkVyXZN8mbk1yb5JIZzgkAAGBZWUqw2qW735Pk7u7+u+7+ySTuVgEAAAyW8ijg3cPf66vqh5P8R5KdZzclAACA5WUpweqtVfX4JL+Y5HeSPC7Ja2Y6KwAAgGVkk8Gqu/9y2Pxykh+Y7XQAAACWn6W8YwUAAMAiBCsAAICRBCsAAICRNhmsqupXpra3n+10AAAAlp8Fg1VVnVhVz0zywqnyP89+SgAAAMvLYqsCfjbJ0Um+vao+PuzvUlVP6u6rt8jsAAAAloHFHgW8Nckbk6xN8uwkvz3UX19V/zTjeQEAACwbi92xek6S/5HkCUnekeQzSb7a3S/fEhMDAABYLha8Y9Xdb+zuw5Ncm+T9SbZJsqKq/qGq/tcWmh8AAMBD3mJ3rDa6oLvXJFlTVT/b3d9XVbvOemIAAADLxSaXW+/uX5ra/Ymh9qVZTQgAAGC5eUA/ENzd/zqriQAAACxXDyhYAQAA8I0EKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJEEKwAAgJFmFqyq6vSquqmqLp+qnVxV11XVpcPnuVPH3lBVa6vq6qp6zlR91VBbW1Wvn6rvW1WfGOofqKpHDvXth/21w/F9ZnWNAAAAyWzvWL03yap56qd294HD5/wkqaoDkhyT5MlDn9+rqm2qapsk70pyZJIDkhw7tE2Stw1jPTHJLUmOG+rHJbllqJ86tAMAAJiZmQWr7v77JDcvsflRSc7q7ru6+/NJ1iY5ZPis7e5ruvtrSc5KclRVVZLDkpwz9D8jyfOnxjpj2D4nyeFDewAAgJnYGu9YnVBVnxkeFdxpqO2Z5ItTbdYNtYXquyS5tbvvmVO/31jD8S8P7b9BVR1fVWuqas369evHXxkAAPBNaUsHq3cneUKSA5Ncn+TtW/j899Pdp3X3wd198IoVK7bmVAAAgGVsiwar7r6xu+/t7vuS/GEmj/olyXVJ9ppqunKoLVTfkGTHqtp2Tv1+Yw3HHz+0BwAAmIktGqyqavep3R9LsnHFwPOSHDOs6Ldvkv2SfDLJJUn2G1YAfGQmC1yc192d5GNJXjj0X53k3KmxVg/bL0zyt0N7AACAmdh2000enKo6M8mzk+xaVeuSnJTk2VV1YJJOcm2Sn06S7r6iqs5OcmWSe5K8srvvHcY5IckFSbZJcnp3XzGc4sQkZ1XVW5N8Osl7hvp7kry/qtZmsnjGMbO6RgAAgGSGwaq7j52n/J55ahvbn5LklHnq5yc5f576NfmvRwmn63cmOfoBTRYAAGCErbEqIAAAwMOKYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADCSYAUAADDSzIJVVZ1eVTdV1eVTtZ2r6sKq+tzwd6ehXlX1zqpaW1WfqaqnT/VZPbT/XFWtnqofVFWXDX3eWVW12DkAAABmZZZ3rN6bZNWc2uuTfLS790vy0WE/SY5Mst/wOT7Ju5NJSEpyUpJnJDkkyUlTQendSV4x1W/VJs4BAAAwEzMLVt3990lunlM+KskZw/YZSZ4/VX9fT1ycZMeq2j3Jc5Jc2N03d/ctSS5Msmo49rjuvri7O8n75ow13zkAAABmYku/Y7Vbd18/bN+QZLdhe88kX5xqt26oLVZfN099sXN8g6o6vqrWVNWa9evXP4jLAQAA2IqLVwx3mnprnqO7T+vug7v74BUrVsxyKgAAwMPYlg5WNw6P8WX4e9NQvy7JXlPtVg61xeor56kvdg4AAICZ2NLB6rwkG1f2W53k3Kn6y4bVAQ9N8uXhcb4LkhxRVTsNi1YckeSC4dhtVXXosBrgy+aMNd85AAAAZmLbWQ1cVWcmeXaSXatqXSar+/16krOr6rgkX0jyoqH5+Umem2RtkjuSvDxJuvvmqnpLkkuGdr/a3RsXxPi5TFYefFSSDw+fLHIOAACAmZhZsOruYxc4dPg8bTvJKxcY5/Qkp89TX5PkKfPUN8x3DgAAgFnZaotXAAAAPFwIVgAAACMJVgAAACMJVgAAACMJVgAAACMJVgAAACMJVgAAACMJVgAAACPN7AeCAQDgm83JF5384Po9+8H1e0g7+eSH1jgz5o4VAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASIIVAADASFslWFXVtVV1WVVdWlVrhtrOVXVhVX1u+LvTUK+qemdVra2qz1TV06fGWT20/1xVrZ6qHzSMv3boW1v+KgEAgG8WW/OO1Q9094HdffCw//okH+3u/ZJ8dNhPkiOT7Dd8jk/y7mQSxJKclOQZSQ5JctLGMDa0ecVUv1WzvxwAAOCb1UPpUcCjkpwxbJ+R5PlT9ff1xMVJdqyq3ZM8J8mF3X1zd9+S5MIkq4Zjj+vui7u7k7xvaiwAAIDNbmsFq07y11X1qao6fqjt1t3XD9s3JNlt2N4zyRen+q4baovV181T/wZVdXxVramqNevXrx9zPQAAwDexbbfSeb+vu6+rqm9NcmFVfXb6YHd3VfWsJ9HdpyU5LUkOPvjgmZ8PAAB4eNoqd6y6+7rh701J/iKTd6RuHB7jy/D3pqH5dUn2muq+cqgtVl85Tx0AAGAmtniwqqpvqarHbtxOckSSy5Ocl2Tjyn6rk5w7bJ+X5GXD6oCHJvny8MjgBUmOqKqdhkUrjkhywXDstqo6dFgN8GVTYwEAAGx2W+NRwN2S/MWwAvq2Sf6suz9SVZckObuqjkvyhSQvGtqfn+S5SdYmuSPJy5Oku2+uqrckuWRo96vdffOw/XNJ3pvkUUk+PHwAAABmYosHq+6+JsnT5qlvSHL4PPVO8soFxjo9yenz1NckecroyQIAACzBQ2m5dQAAgGVJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABhJsAIAABjpYRusqmpVVV1dVWur6vVbez4A/J/27j3YrvEO4/j3EUo1SJEaI3EpGZqqxqURE/dSihatqhl1aRlj0LhF0c5o2o5bDQnD6JhWQ5vSjEuktIMhGRF3EgmRlmraSpESIp2SEXn6x34Py3YqJ1k5Z5+c/Xxm9pz1vutda//Ont9k5bffd60TERHRd63Z6gC6g6R+wDXA/sBLwOOSJtue09rIetiYMb3rPG1mzNQxK37M3it+TERERES0Xp8srIDhwAu2XwSQdDNwKNBehdWq0psKq94USzdYmWIMUpBFREREtJpstzqGVU7SEcCBtk8s7WOAXW2f1jTuJOCk0twW+HOPBtoeNgZea3UQsVpLDkUdyZ+oKzkUdSWH+pYtbA/sbEdfnbHqEtvXAde1Oo6+TNITtndpdRyx+koORR3Jn6grORR1JYfaR199eMV8YHClPaj0RURERERErHJ9tbB6HBgiaStJnwCOAia3OKaIiIiIiOij+uRSQNtLJZ0G3A30A663/WyLw2pXWWoZdSWHoo7kT9SVHIq6kkNtok8+vCIiIiIiIqIn9dWlgBERERERET0mhVVERERERERNKayiFknXS1og6ZlK34aS7pX0fPn56dIvSVdJekHSLEk7tS7y6A0kDZY0RdIcSc9KOr30J4eiSyStI+kxSU+XHPpJ6d9K0qMlV35fHmSEpLVL+4Wyf8tWxh+9g6R+kmZIurO0kz/RZZLmSZotaaakJ0pfrmNtKIVV1DUeOLCp7zzgPttDgPtKG+CrwJDyOgm4todijN5rKXC27aHACOBUSUNJDkXXLQH2tf1FYBhwoKQRwKXAWNvbAG8AJ5TxJwBvlP6xZVzE6cBzlXbyJ1bUPraHVf5eVa5jbSiFVdRi+wFgYVP3ocANZfsG4LBK/41ueAQYIGnTnok0eiPbL9t+qmwvpvEfm81IDkUXlVz4T2muVV4G9gVuKf3NOdSRW7cAX5akHgo3eiFJg4CDgV+Wtkj+RH25jrWhFFbRHTax/XLZfgXYpGxvBvyzMu6l0hdBWVKzI/AoyaFYAWUZ10xgAXAv8FfgTdtLy5BqnryfQ2X/ImCjno04eplxwA+AZaW9EcmfWDEG7pH0pKSTSl+uY22oT/4dq+g9bFtSnukfH0tSf+BW4Azbb1W/AE4OxfLYfg8YJmkAcDuwXYtDitWEpEOABbaflLR3q+OJ1dbutudL+gxwr6S51Z25jrWPzFhFd3i1Y1q7/FxQ+ucDgyvjBpW+aGOS1qJRVE2wfVvpTg7FCrP9JjAF2I3G8pqOLw+refJ+DpX9GwCv93Co0XuMBL4uaR5wM40lgFeS/IkVYHt++bmAxpc7w8l1rC2lsIruMBk4rmwfB9xR6T+2PBFnBLCoMk0ebajcm/Ar4DnbV1R2JYeiSyQNLDNVSPoksD+Ne/WmAEeUYc051JFbRwD32843yW3K9vm2B9neEjiKRj4cTfInukjSpySt17ENfAV4hlzH2pLy70HUIekmYG9gY+BV4MfAJGAisDnwd+BI2wvLf6KvpvEUwf8C37X9RCvijt5B0u7ANGA2H9zf8EMa91klh2K5JO1A48bwfjS+LJxo+6eSPktjBmJDYAbwHdtLJK0D/IbG/XwLgaNsv9ia6KM3KUsBR9s+JPkTXVVy5fbSXBP4ne0LJW1ErmNtJ4VVRERERERETVkKGBERERERUVMKq4iIiIiIiJpSWEVERERERNSUwioiIiIiIqKmFFYRERERERE1pbCKiIhuIeliSftIOkzS+S2K4XhJV/fA+/xR0oDyOqXSv7ekO7tw/HhJ8yWtXdoblz9au7zjLOnySnu0pDEr91tEREQdKawiIqK77Ao8AuwFPNDiWLqV7YNsvwkMAE5Z3vj/4z3geyt4zBLgG5I2Xsn3jIiIVSSFVURErFKSLpM0C/gS8DBwInCtpAvK/lGS5kiaJenm0jdc0sOSZkh6SNK2pf94SZMk3StpnqTTJJ1Vxj0iacMybqqkKyXNlPSMpOGdxDVQ0q2SHi+vkaV/r3LczHLe9ZqOO0fSqLI9VtL9ZXtfSRPK9rxS3FwCbF3OdVk5RX9Jt0iaK2lC+QOhnRkHnClpzab37y/pPklPSZot6dDK7qXAdcCZnfy+3yqfxdOS+nRhGxHRG6y5/CERERFdZ/scSROBY4GzgKm2R1aGnAdsZXuJpAGlby6wh+2lkvYDLgK+WfZtD+wIrAO8AJxre0dJY8t7jCvj1rU9TNKewPXluKorgbG2H5S0OXA38DlgNHCq7emS+gPvNB03DTgbuArYBVhb0lrAHnx0Ju48YHvbw6CxFLDE/nngX8B0YCTwYCcf3T9K/zHAHyr97wCH236rFG+PSJps22X/NcAsST9vOt8FwAG251c+54iI6CYprCIiojvsBDwNbAc817RvFjBB0iRgUunbALhB0hDAwFqV8VNsLwYWS1rEB0XHbGCHyribAGw/IGn9ToqJ/YChlQmj9UshNR24osw+3Wb7pabjngR2lrQ+jaV3T9EosPYARi3/o+CxjnNKmglsSeeFFcDFwB3AXZU+AReVgnEZsBmwCfBK+X3fknRjieXtynHTgfGlyL2tC3FGREQNKawiImKVkTQMGA8MAl4D1m10ayawm+23gYOBPYGvAT+S9AXgZzQKqMMlbQlMrZx2SWV7WaW9jA9fx8yHNbfXAEbYbp6RukTSXcBBwHRJB9ie+/5J7Hcl/Q04HniIRmG4D7ANHy0aO1ON/z0+5tpr+/nyWR1Z6T4aGAjsXGKZR2P2rmocjYLv15VznSxpVxqf95OSdrb9ehfijYiIlZB7rCIiYpWxPbMsg/sLMBS4n8ZytGG235a0BjDY9hTgXBozVf3Lz/nlNMev5Nt/G0DS7sAi24ua9t8DfL+jUYpAJG1te7btS4HHacyyNZtGY8ngA2X7ZGBGZTleh8XAetRzYXmvDhsAC0pRtQ+wRfMBthcCE4ETOvrK7/Wo7QuAfwODa8YVEREfI4VVRESsUpIGAm/YXgZsZ3tOZXc/4LeSZgMzgKvK0/R+DlwsaQYrv5rinXL8L6gUGBWjgF3KQzPm0CiOAM4oD3mYBbwL/KmTY6cBmwIP236Vxn1P05oHlRmh6eV8lzXv7wrbz9KYfeowocQ9m8Y9ZXM7PRAuB6pPB7ysPOziGRozbU+vTDwREdE1+uiXbREREasXSVOB0bafaHUsERHRnjJjFRERERERUVNmrCIiIiIiImrKjFVERERERERNKawiIiIiIiJqSmEVERERERFRUwqriIiIiIiImlJYRURERERE1PQ//qKg6gG7BrQAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1008x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(14,10))\n",
    "ax.hist(x=Meth_data['LA'].isna().sum(axis=1).values, bins=40, color=['red'], stacked=True, label='LA', alpha=0.5)\n",
    "ax.hist(x=Meth_data['LSCC'].isna().sum(axis=1).values, bins=40, color=['green'], stacked=True, label='LSCC', alpha=0.5)\n",
    "plt.legend()\n",
    "plt.title(\"# NaN's per probe, distribution\")\n",
    "plt.xlabel('#samples with NaNs')\n",
    "plt.ylabel('# affected probes')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** I.e. About 130 samples exclusively have NaN's ** \n",
    "\n",
    "** I.e. practically all of the samples of the Adenocarcinoma and Squamous Cell carcinoma sets have about 80.000 probes with NaN's ** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "del Meth_data_T\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "_dim_reduction() got an unexpected keyword argument 'min_samples'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-22-eaafc1450d67>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mgc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcollect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mdata_reduced\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mdata_reduced\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'LA'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_dim_reduction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_cleaned_LA\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmin_samples\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_comp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m50\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'PCA'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpc_solver\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'arpack'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mdata_reduced\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'LSCC'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_dim_reduction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_cleaned_LSCC\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmin_samples\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_comp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m50\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'PCA'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpc_solver\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'arpack'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: _dim_reduction() got an unexpected keyword argument 'min_samples'"
     ]
    }
   ],
   "source": [
    "gc.collect()\n",
    "data_reduced = {}\n",
    "data_reduced['LA'] = _dim_reduction(_cleaned_LA, min_samples=10, n_comp=50, method='PCA', pc_solver='arpack')\n",
    "data_reduced['LSCC'] = _dim_reduction(_cleaned_LSCC, min_samples=10, n_comp=50, method='PCA', pc_solver='arpack')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_reduced['ALL'] = _dim_reduction(_cleaned_ALL, min_samples=10, n_comp=50, method='PCA', pc_solver='arpack')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(14,10))\n",
    "for _key in data_reduced.keys():\n",
    "    plt.plot(data_reduced[_key]['ALL']['reducer'].explained_variance_ratio_, label=_key)\n",
    "    plt.semilogy()\n",
    "    plt.ylabel('Variance explained ratio')\n",
    "    plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(14,10))\n",
    "plt.hist(data_reduced['LA']['ALL']['reducer'].components_[0], bins=50, histtype='step', label='LA');\n",
    "plt.hist(data_reduced['LSCC']['ALL']['reducer'].components_[0], bins=50, histtype='step', label='LSCC');\n",
    "plt.hist(data_reduced['ALL']['ALL']['reducer'].components_[0], bins=50, histtype='step', label='ALL');\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_filter= (data_reduced['ALL']['ALL']['reducer'].components_[0]>0.001) | \\\n",
    "         (data_reduced['ALL']['ALL']['reducer'].components_[0]<-0.004)\n",
    "pca_0_idx = np.argwhere(_filter)\n",
    "Genes = _cleaned_ALL.columns[pca_0_idx]\n",
    "print(\"Important genes are for the first principal component: \\n {}\".format(Genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# top 100 probes\n",
    "N = 10\n",
    "top_N_pc0 = np.argsort(np.abs(data_reduced['ALL']['ALL']['reducer'].components_[0]))[-N:]\n",
    "top_N_pc1 = np.argsort(np.abs(data_reduced['ALL']['ALL']['reducer'].components_[1]))[-N:]\n",
    "top_N_pc2 = np.argsort(np.abs(data_reduced['ALL']['ALL']['reducer'].components_[2]))[-N:]\n",
    "Top_N_Probes_PC0 = _cleaned_ALL.columns[top_N_pc0]\n",
    "Top_N_Probes_PC1 = _cleaned_ALL.columns[top_N_pc1]\n",
    "Top_N_Probes_PC2 = _cleaned_ALL.columns[top_N_pc2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp = _cleaned_ALL.merge(pheno[~pheno.index.duplicated()]['Diagnosis'], how='left', left_index=True, right_index=True)\n",
    "y = tmp.Diagnosis.apply(lambda x: 0 if x=='Lung Adenocarcinoma' else 1).values\n",
    "x = tmp.drop('Diagnosis', axis=1)\n",
    "del tmp\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running reducer on all samples..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.6/dist-packages/sklearn/discriminant_analysis.py:388: UserWarning: Variables are collinear.\n",
      "  warnings.warn(\"Variables are collinear.\")\n"
     ]
    }
   ],
   "source": [
    "ALL_LDA = _dim_reduction(x, y, n_comp=1, method='LDA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "N = 500 \n",
    "top_N_lda = np.argsort(ALL_LDA['ALL']['reducer'].coef_)[0][-N:]\n",
    "Top_N_Probes_LDA = _cleaned_ALL.columns[top_N_lda]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "LDA_scores = pd.DataFrame(data=ALL_LDA['ALL']['reducer'].coef_.reshape([-1,1]), \n",
    "                          index=_cleaned_ALL.columns, \n",
    "                          columns=['LDA_score'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'Top_N_Probes_PC0' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-27-61eda8631052>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mGeneCount\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mGeneMapping\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Gene'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mascending\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'tot_count'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m GeneOccurrencePC0 = pd.DataFrame(GeneMapping.loc[Top_N_Probes_PC0,:].groupby('Gene').size().sort_values(ascending=False),\n\u001b[0m\u001b[1;32m      3\u001b[0m                               columns=['occurrence_pc0'])\n\u001b[1;32m      4\u001b[0m GeneOccurrencePC1 = pd.DataFrame(GeneMapping.loc[Top_N_Probes_PC1,:].groupby('Gene').size().sort_values(ascending=False),\n\u001b[1;32m      5\u001b[0m                               columns=['occurrence_pc1'])\n",
      "\u001b[0;31mNameError\u001b[0m: name 'Top_N_Probes_PC0' is not defined"
     ]
    }
   ],
   "source": [
    "GeneCount = pd.DataFrame(GeneMapping.groupby('Gene').size().sort_values(ascending=False), columns=['tot_count'])\n",
    "GeneOccurrencePC0 = pd.DataFrame(GeneMapping.loc[Top_N_Probes_PC0,:].groupby('Gene').size().sort_values(ascending=False),\n",
    "                              columns=['occurrence_pc0'])\n",
    "GeneOccurrencePC1 = pd.DataFrame(GeneMapping.loc[Top_N_Probes_PC1,:].groupby('Gene').size().sort_values(ascending=False),\n",
    "                              columns=['occurrence_pc1'])\n",
    "GeneOccurrencePC2 = pd.DataFrame(GeneMapping.loc[Top_N_Probes_PC2,:].groupby('Gene').size().sort_values(ascending=False),\n",
    "                              columns=['occurrence_pc2'])\n",
    "GeneOccurrenceLDA = pd.DataFrame(GeneMapping.loc[Top_N_Probes_LDA,:].groupby('Gene').size().sort_values(ascending=False),\n",
    "                              columns=['occurrence_lda'])\n",
    "GeneOccurrence = pd.concat([GeneOccurrencePC0, GeneOccurrencePC1, GeneOccurrencePC2, GeneOccurrenceLDA], \n",
    "                           axis=1)\n",
    "GeneOccurrence['tot_count'] = GeneOccurrence.apply(lambda x: np.sum(x), axis=1).sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clustering of cohorts\n",
    "data_reduced['LA']['ALL']['data'] = data_reduced['LA']['ALL']['data'][~data_reduced['LA']['ALL']['data'].index.duplicated(keep='first')]\n",
    "data_reduced['LSCC']['ALL']['data'] = data_reduced['LSCC']['ALL']['data'][~data_reduced['LSCC']['ALL']['data'].index.duplicated(keep='first')]\n",
    "\n",
    "merged_LA_reduced = data_reduced['LA']['ALL']['data'].merge(pheno, how='left', left_index=True, right_index=True)\n",
    "merged_LSCC_reduced = data_reduced['LSCC']['ALL']['data'].merge(pheno, how='left', left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "red_cols = [_col for _col in merged_LA_reduced.columns if 'red_' in _col]\n",
    "red_cols += ['batch_number'] \n",
    "plot_LA = merged_LA_reduced[red_cols].groupby('batch_number').mean()\n",
    "\n",
    "red_cols = [_col for _col in merged_LA_reduced.columns if 'red_' in _col]\n",
    "red_cols += ['batch_number'] \n",
    "plot_LSCC = merged_LSCC_reduced[red_cols].groupby('batch_number').mean()\n",
    "\n",
    "fig, ax = plt.subplots(ncols=2, nrows=2, figsize=(14,14))\n",
    "sns.scatterplot(data=plot_LA, x='red_0', y='red_1', s=160, label='LA', ax=ax[0,0])\n",
    "sns.scatterplot(data=plot_LSCC, x='red_0', y='red_1', s=160, label='LSCC', ax=ax[0,0])\n",
    "ax[0,0].set_xlabel('PC 0')\n",
    "ax[0,0].set_ylabel('PC 1')\n",
    "ax[0,0].legend()\n",
    "ax[0,0].axvline(0, color='black')\n",
    "ax[0,0].axhline(0, color='black')\n",
    "\n",
    "sns.scatterplot(data=plot_LA, x='red_1', y='red_2', s=160, ax=ax[0,1])\n",
    "sns.scatterplot(data=plot_LSCC, x='red_1', y='red_2', s=160, ax=ax[0,1])\n",
    "ax[0,1].set_xlabel('PC 1')\n",
    "ax[0,1].set_ylabel('PC 2')\n",
    "ax[0,1].axvline(0, color='black')\n",
    "ax[0,1].axhline(0, color='black')\n",
    "\n",
    "sns.scatterplot(data=plot_LA, x='red_2', y='red_3', s=160, ax=ax[1,0])\n",
    "sns.scatterplot(data=plot_LSCC, x='red_2', y='red_3', s=160, ax=ax[1,0])\n",
    "ax[1,0].set_xlabel('PC 2')\n",
    "ax[1,0].set_ylabel('PC 3')\n",
    "ax[1,0].axvline(0, color='black')\n",
    "ax[1,0].axhline(0, color='black')\n",
    "\n",
    "sns.scatterplot(data=plot_LA, x='red_3', y='red_4', s=160, ax=ax[1,1])\n",
    "sns.scatterplot(data=plot_LSCC, x='red_3', y='red_4', s=160, ax=ax[1,1])\n",
    "ax[1,1].set_xlabel('PC 3')\n",
    "ax[1,1].set_ylabel('PC 4')\n",
    "ax[1,1].axvline(0, color='black')\n",
    "ax[1,1].axhline(0, color='black')\n",
    "\n",
    "\n",
    "plt.suptitle('Mean methylation PC values per cohort')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The cohort means for the first principle components are distributed around the origin, so assuming an equal sample contribution per cohort the overall \n",
    "effect is likely diffuse.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Batch seperability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_reduced['LA']['ALL']['data'] = merged_LA_reduced\n",
    "data_reduced['LSCC']['ALL']['data'] = merged_LSCC_reduced"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "num_test_batches = 15\n",
    "test_batches = {}\n",
    "\n",
    "num_of_batches=data_reduced['LA']['ALL']['data']['batch_number'].nunique()\n",
    "batches = data_reduced['LA']['ALL']['data']['batch_number'].unique()\n",
    "batch_indices = sample(range(0, num_of_batches), num_test_batches)\n",
    "test_batches['LA'] = batches[batch_indices].tolist()\n",
    "\n",
    "num_of_batches=data_reduced['LSCC']['ALL']['data']['batch_number'].nunique()\n",
    "batches = data_reduced['LSCC']['ALL']['data']['batch_number'].unique()\n",
    "batch_indices = sample(range(0, num_of_batches), num_test_batches)\n",
    "test_batches['LSCC'] = batches[batch_indices].tolist()\n",
    "\n",
    "axc = 0 # 0: distribution of the mean probe values over the samples, 1: mean sample values over the probes\n",
    "xr = -20, 20\n",
    "fig, ax = plt.subplots(ncols=2, figsize=(14,10))\n",
    "for rdx, _dkey in enumerate(data_reduced.keys()):   \n",
    "    print(\"Processing ds: {}\".format(_dkey))\n",
    "    for cdx, _batch in enumerate(test_batches[_dkey]):\n",
    "        #print('Batch: {}'.format(_batch))\n",
    "        Red_columns = [_col for _col in data_reduced[_dkey]['ALL']['data'].columns if 'red_' in _col]\n",
    "        data_reduced[_dkey]['ALL']['data'].query('batch_number==\"'+str(_batch)+'\"').loc[:,Red_columns]\\\n",
    "                                                                     .median(axis=axc).plot.kde(label=_batch, \n",
    "                                                                                              ax=ax[rdx])            \n",
    "    \n",
    "    data_reduced[_dkey]['ALL']['data'].loc[:,Red_columns].median(axis=0).plot.kde(label='ALL', color='black', ax=ax[rdx])\n",
    "    ax[rdx].set_xlim(xr[0], xr[1])\n",
    "    ax[rdx].set_title(_dkey)\n",
    "    ax[rdx].legend()\n",
    "    fig.suptitle('Median of the principle components over the samples, per cohort')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = Red_columns+['batch_number']\n",
    "data_reduced['LA']['ALL']['data'].loc[:,cols].groupby('batch_number').mean().mean(axis=1).plot.hist(histtype='step', label='LA', figsize=(10,7))\n",
    "data_reduced['LSCC']['ALL']['data'].loc[:,cols].groupby('batch_number').mean().mean(axis=1).plot.hist(histtype='step', label='LSCC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Wasserstein distance internal: distance between samples and median of samples\n",
    "#w1_dist(vi, vm)\n",
    "#w2_dist(vi, vm)\n",
    "\n",
    "# Wasserstein distance external: distance between median of samples and mean of median of cohorts\n",
    "#w1_dist(vm, vmm)\n",
    "#w2_dist(vm, vmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differential methylation between LA and LSCC\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "42"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# KS, MW-U, FDR\n",
    "gc.collect()\n",
    "tmp = _cleaned_ALL.merge(pheno[~pheno.index.duplicated()]['Diagnosis'], how='left', left_index=True, right_index=True)\n",
    "y = tmp.Diagnosis.apply(lambda x: 0 if x=='Lung Adenocarcinoma' else 1).values\n",
    "x = tmp.drop('Diagnosis', axis=1)\n",
    "del tmp \n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "probes_diffs = {}\n",
    "probes_diffs['FDR'] = get_differential_probes(x, y, method='FDR', alpha=0.05) # FDR, FPR, KS, MWU"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "probes_diffs['MWU'] = get_differential_probes(x, y, method='MWU', alpha=0.05) # FDR, FPR, KS, MWU"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "probes_diffs['KS'] = get_differential_probes(x, y, method='KS', alpha=0.05, mode='exact') # FDR, FPR, KS, MWU"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "#probes_diffs['AD'] = get_differential_probes(x, y, method='AD', alpha=0.05) # FDR, FPR, KS, MW"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "probes_diffs['WASS1'] = get_differential_probes(x, y, method='WASS1') \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "probes_diffs['WASS2'] = get_differential_probes(x, y, method='WASS2') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'score'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-36-4cb08b760da0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mtop_probes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0m_key\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mprobes_diffs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0mq\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mprobes_diffs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0m_key\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mscore\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquantile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0.95\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0mprobes_diffs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0m_key\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'score'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m  \u001b[0mprobes_diffs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0m_key\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'score'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mabs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mtop_probes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0m_key\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprobes_diffs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0m_key\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"score>\"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'score'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mascending\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.6/dist-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5055\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5056\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5057\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5058\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5059\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'score'"
     ]
    }
   ],
   "source": [
    "top_probes = {}\n",
    "for _key in probes_diffs.keys():    \n",
    "    q=probes_diffs[_key].score.quantile(0.95)\n",
    "    probes_diffs[_key]['score'] =  probes_diffs[_key]['score'].abs()\n",
    "    top_probes[_key] = probes_diffs[_key].query(\"score>\"+str(q)).sort_values('score', ascending=False)\n",
    "    top_probes[_key] = top_probes[_key].copy().merge(GeneMapping, left_index=True, right_index=True, how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#top_probes[_key].loc[top_probes[_key].Gene.isin(['BRAF', 'MAP2K1', 'KRAS', 'MET', 'RET', 'ROS1', 'EGFR', 'TP53']),:]\n",
    "#top_probes[_key].loc[top_probes[_key].Gene.isin(['KRT5', 'DSC3', 'DSG3', 'TP63', 'MIR205HG', 'CALML3']),:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "FDR/FPR: \n",
    "```BRAF```, ```MAP2K1``` are present which were not differentially expressed in RNA, but not ```KRT5```, ```DSC3```, ```DSG3```, ```TP63```, ```MIR205HG```, ```CALML3``` which **were** strongly differential for the RNA expression.\n",
    "Other genes we might expect are: ```KRAS```, ```MET```, ```RET```, ```ROS1```, ```EGFR```, ```TP53```. \n",
    "\n",
    "MWU: Only ```MAP2K1``` is present.\n",
    "\n",
    "KS: ```MAP2K1``` of the important general cancer genes.  ```CALML3``` and ```TP63``` of the differential RNA expression genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fig, ax = plt.subplots(figsize=(16,10), ncols=3)\n",
    "for idx, _key in enumerate(probes_diffs.keys()):\n",
    "    if _key in['AD', 'WASS1', 'WASS2']:\n",
    "        break\n",
    "    sns.jointplot(x='pvals', y='score', data=probes_diffs[_key], alpha=0.1)\n",
    "    plt.title(_key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get fold change"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc_set = _cleaned_ALL.merge(pheno[~pheno.index.duplicated()]['Diagnosis'], how='left', left_index=True, right_index=True)\n",
    "fc_set['target'] = fc_set.Diagnosis.apply(lambda x: 0 if x=='Lung Adenocarcinoma' else 1)\n",
    "fc_set.drop('Diagnosis', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "abs_fc = (fc_set.query('target==1').mean(axis=0) - fc_set.query('target==0').mean(axis=0))\n",
    "rel_fc = (fc_set.query('target==1').mean(axis=0) - fc_set.query('target==0').mean(axis=0))/fc_set.query('target==0').mean(axis=0)\n",
    "abs_fc_median = (fc_set.query('target==1').median(axis=0) - fc_set.query('target==0').median(axis=0))\n",
    "split_fc = (fc_set.query('target==1').median(axis=0)+fc_set.query('target==0').median(axis=0))/2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "abs_fc_df = pd.DataFrame(abs_fc, columns=['abs_fc'])\n",
    "rel_fc_df = pd.DataFrame(rel_fc, columns=['rel_fc'])\n",
    "abs_fc_median_df = pd.DataFrame(abs_fc_median, columns=['abs_fc_median'])\n",
    "split_fc_df = pd.DataFrame(split_fc, columns=['split_fc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc = pd.concat([abs_fc_df, rel_fc_df, abs_fc_median_df, split_fc_df], axis=1)\n",
    "fc.drop('target', axis=0, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_fc = pd.concat([fc.sort_values(by='rel_fc', ascending=False)[:2500],\n",
    "           fc.sort_values(by='rel_fc', ascending=True)[:2500]],\n",
    "          axis=0)\n",
    "top_fc = top_fc.copy().merge(GeneMapping, left_index=True, right_index=True, how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# top cancer genomes (RNA expression)\n",
    "top_fc.loc[top_fc.Gene.isin(['BRAF', 'MAP2K1', 'KRAS', 'MET', 'RET', 'ROS1', 'EGFR', 'TP53']),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# top differentials (RNA expression)\n",
    "top_fc.loc[top_fc.Gene.isin(['KRT5', 'DSC3', 'DSG3', 'TP63', 'MIR205HG', 'CALML3']),:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Of the differentially expressed RNA only ```DSC3``` and ```MIR205HG``` are present in the top 5000 of fold-changes.\n",
    "Of the important cancer genes only ```RET``` and ```KRAS``` is present."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes_from_meth = ['ARHGEF16', 'AKR7A2', 'LRRC59', 'WFDC2', 'PSRC1', 'NEK9','RUVBL1', \n",
    "                       'COL6A4P2', 'MYO1C', 'PCDHGB5', 'RAPGEFL1', 'PAX6', 'PPP2R2C', 'GPR133', 'EN1', 'PRDM16', \n",
    "                       'PTH2R', 'CLEC2L', 'RPS4Y1', 'DDX43', 'TDRD1', 'CALML3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Sep:\", top_probes['KS'].loc[top_probes['KS'].Gene.isin(top_genes_from_meth),:]['Gene'].unique())\n",
    "\n",
    "print(\"FC:\", top_fc.loc[top_fc.Gene.isin(top_genes_from_meth),:]['Gene'].unique())\n",
    "\n",
    "print(\"+\"*50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Induced differentiability due to imputance ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "LA_nans = Meth_data['LA'].isna().sum(axis=1)\n",
    "LSCC_nans = Meth_data['LSCC'].isna().sum(axis=1)\n",
    "\n",
    "LA_nan_list = LA_nans[np.where(LA_nans>200)[0]].index.tolist()\n",
    "LSCC_nan_list = LSCC_nans[np.where(LSCC_nans>200)[0]].index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"\\t\\tIn the LA set we have {} probes that are NaN for all probes, \\n\\\n",
    "       \\t\\tIn the LSCC set have {} probes that are NaN for all probes. \\n\\\n",
    "      \\t\\tThe overlap between the probes is roughly {}%.\".format(len(LA_nan_list), \n",
    "                                                                 len(LSCC_nan_list),\n",
    "                            len(set(LA_nan_list).intersection(set(LSCC_nan_list)))/len(LA_nan_list)*100))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Which of the reported genes are in the top-list by PwC?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# connect probes to Genes\n",
    "overlap_list = list(set(LA_nan_list).intersection(set(LSCC_nan_list)))\n",
    "la_only_list = list(set(LA_nan_list)-set(LSCC_nan_list))\n",
    "\n",
    "overlap_df = pd.DataFrame(data=np.array(overlap_list), columns=['overlap'])\n",
    "la_only_df = pd.DataFrame(data=np.array(la_only_list), columns=['LA_only'])\n",
    "\n",
    "overlap_df = overlap_df.merge(GeneMapping, left_on='overlap', right_index=True, how='left')\n",
    "la_only_df = la_only_df.merge(GeneMapping, left_on='LA_only', right_index=True, how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Sep:\", overlap_df.loc[overlap_df.Gene.isin(top_genes_from_meth),:]['Gene'].unique())\n",
    "print(\"+\"*50)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hence, at face value, the imputation has had a significant impact on the differential analysis...or has it..?\n",
    "**Obviously** I should look at the probe id's!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_probes_pwc = ['cg1397493', 'cg6117986', 'cg9149410', 'cg4151783', 'cg2694271', 'cg3033748', 'cg4972564',\n",
    "                  'cg7969602', 'cg2317299', 'cg4656271', 'cg0281776', 'cg4717314', 'cg6301283', 'cg1375382',\n",
    "                  'cg3669949', 'cg6600802', 'cg7547703', 'cg7283169', 'cg0691436', 'cg6550453', 'cg0668644',\n",
    "                  'cg4439629', 'cg9506435', 'cg9827780', 'cg2153041', 'cg1301942', 'cg3082523', 'cg3050300',\n",
    "                  'cg2325673', 'cg0129651', 'cg4544105', 'cg6797068', 'cg5533223', 'cg4831879', 'cg0691436',\n",
    "                  'cg7283169', 'cg1375382', 'cg2034222', 'cg4506221', 'cg3669949']\n",
    "\n",
    "top_probes_pwc_nocg = [_top[2:] for _top in top_probes_pwc]\n",
    "top_probes_pwc_0 = ['cg'+'0'+_top[2:] for _top in top_probes_pwc]\n",
    "top_probes_pwc_1 = ['cg'+'1'+_top[2:] for _top in top_probes_pwc]\n",
    "top_probes_pwc_2 = ['cg'+'2'+_top[2:] for _top in top_probes_pwc]\n",
    "top_probes_pwc_3 = ['cg'+'3'+_top[2:] for _top in top_probes_pwc]\n",
    "top_probes_pwc_4 = ['cg'+'4'+_top[2:] for _top in top_probes_pwc]\n",
    "top_probes_pwc_5 = ['cg'+'5'+_top[2:] for _top in top_probes_pwc]\n",
    "\n",
    "top_probes_pwc_correct = top_probes_pwc_0+\\\n",
    "                         top_probes_pwc_1+\\\n",
    "                         top_probes_pwc_2+\\\n",
    "                         top_probes_pwc_3+\\\n",
    "                         top_probes_pwc_4+\\\n",
    "                         top_probes_pwc_5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_overlap = overlap_df.loc[overlap_df.overlap.isin(top_probes_pwc_correct)].shape[0]\n",
    "in_la_only = la_only_df.loc[la_only_df.LA_only.isin(top_probes_pwc_correct)].shape[0]\n",
    "\n",
    "print('NA\\'s in top-list: {} and {}'.format(in_overlap, in_la_only))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perhaps we removed probes unnecessarily, which would mean these probes are also not present in the cleaned df.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_cleaned = (set(_cleaned_ALL.columns.tolist())).intersection(set(top_probes_pwc_correct))\n",
    "Corrected_top = list(in_cleaned)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OK, I have REMOVED perfectly good probes! Why?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_raw = set(Meth_data['LSCC'].index.tolist()).intersection(set(top_probes_pwc_correct))\n",
    "print(list(in_raw))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**What??**, they are also not in the RAW data...?? Where did these probes come from?\n",
    "Are they perhaps in the meta data for the Illumina data?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_illumina1 = set(methylation_meta['IlmnID'].tolist()).intersection(set(top_probes_pwc_correct))\n",
    "print(list(in_illumina1))\n",
    "\n",
    "in_illumina2 = set(methylation_meta['Name'].tolist()).intersection(set(top_probes_pwc_correct))\n",
    "print(list(in_illumina2))\n",
    "\n",
    "in_illumina1 = set(methylation_meta['IlmnID'].tolist()).intersection(set(top_probes_pwc_correct))\n",
    "print(list(in_illumina1))\n",
    "\n",
    "in_illumina2 = set(methylation_meta['Name'].tolist()).intersection(set(top_probes_pwc_correct))\n",
    "print(list(in_illumina2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "They don't exist??? It seems they do..the probe-id's from their list was just missing a character.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Meth_data_T = {}\n",
    "#Meth_data_T['LA'] = Meth_data['LA'].T\n",
    "#Meth_data_T['LSCC'] = Meth_data['LSCC'].T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Given a list of top-probes, how different are their distribution?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(nrows=len(Corrected_top), figsize=(10,40))\n",
    "for i, probeid in enumerate(Corrected_top):\n",
    "    Meth_data_T['LA'][Corrected_top[i]].plot.kde(color='red', ax=ax[i])\n",
    "    Meth_data_T['LSCC'][Corrected_top[i]].plot.kde(color='green', ax=ax[i])\n",
    "    ax[i].set_ylabel('')\n",
    "    #ax[i].set_title(probeid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FDR_top20 = top_probes['FDR'][:20].index.tolist()\n",
    "KS_top20 = top_probes['KS'][:20].index.tolist()\n",
    "MWU_top20 = top_probes['MWU'][:20].index.tolist()\n",
    "#AD_top20 = top_probes['AD'][:20].index.tolist()\n",
    "WASS1_top20 = top_probes['WASS1'][:20].index.tolist()\n",
    "WASS2_top20 = top_probes['WASS2'][:20].index.tolist()\n",
    "LDA_top20 = LDA_scores.abs().sort_values(by='LDA_score', ascending=False)[:20].index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(nrows=len(FDR_top20), figsize=(10,35))\n",
    "for i, probeid in enumerate(FDR_top20):\n",
    "    Meth_data_T['LA'][probeid].plot.kde(color='red', ax=ax[i])\n",
    "    Meth_data_T['LSCC'][probeid].plot.kde(color='green', ax=ax[i])\n",
    "    ax[i].set_ylabel('')\n",
    "    #ax[i].set_title(probeid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(nrows=len(MWU_top20), figsize=(10,35))\n",
    "for i, probeid in enumerate(MWU_top20):\n",
    "    _cleaned_ALL.iloc[np.where(y==1)][probeid].plot.kde(color='red', ax=ax[i])\n",
    "    _cleaned_ALL.iloc[np.where(y==0)][probeid].plot.kde(color='green', ax=ax[i])\n",
    "    ax[i].set_ylabel('')\n",
    "    #ax[i].set_title(probeid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(nrows=len(KS_top20), figsize=(10,35))\n",
    "for i, probeid in enumerate(KS_top20):\n",
    "    Meth_data_T['LA'][probeid].plot.kde(color='red', ax=ax[i])\n",
    "    Meth_data_T['LSCC'][probeid].plot.kde(color='green', ax=ax[i])\n",
    "    ax[i].set_ylabel('')\n",
    "    #ax[i].set_title(probeid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(nrows=len(WASS1_top20), figsize=(10,35))\n",
    "for i, probeid in enumerate(WASS1_top20):\n",
    "    Meth_data_T['LA'][probeid].plot.kde(color='red', ax=ax[i])\n",
    "    Meth_data_T['LSCC'][probeid].plot.kde(color='green', ax=ax[i])\n",
    "    ax[i].set_ylabel('')\n",
    "    #ax[i].set_title(probeid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(nrows=len(LDA_top20), figsize=(10,35))\n",
    "for i, probeid in enumerate(LDA_top20):\n",
    "    Meth_data_T['LA'][probeid].plot.kde(color='red', ax=ax[i])\n",
    "    Meth_data_T['LSCC'][probeid].plot.kde(color='green', ax=ax[i])\n",
    "    ax[i].set_ylabel('')\n",
    "    #ax[i].set_title(probeid)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get overlap of top-probes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for _key in probes_diffs.keys():\n",
    "#    probes_diffs[_key].columns = ['pvals_'+_key, 'score_'+_key]\n",
    "\n",
    "## Only the overlap of the top-20000 probes\n",
    "ppset= None\n",
    "topM = 50000\n",
    "for idx, _key in enumerate(probes_diffs.keys()):\n",
    "    if _key in ['AD', 'WASS2']:\n",
    "        break\n",
    "    else:\n",
    "        # query('pvals_'+_key+'<0.001')\n",
    "        pset = (set(probes_diffs[_key].sort_values(by='score_'+_key, ascending=False)[:topM].index.tolist()))\n",
    "        if idx>0:\n",
    "            ppset = pset.intersection(ppset)\n",
    "        else:\n",
    "            ppset = pset\n",
    "ppset = list(ppset)\n",
    "\n",
    "tot_overlap = probes_diffs['FDR'].merge(probes_diffs['KS'], \n",
    "                              how='inner', \n",
    "                              left_index=True, \n",
    "                              right_index=True, \n",
    "                              suffixes=['_FDR', '_KS'])\n",
    "tot_overlap = tot_overlap.copy().merge(probes_diffs['MWU'], \n",
    "                              how='inner', \n",
    "                              left_index=True, \n",
    "                              right_index=True)\n",
    "tot_overlap = tot_overlap.copy().merge(probes_diffs['WASS1'], \n",
    "                              how='inner', \n",
    "                              left_index=True, \n",
    "                              right_index=True)\n",
    "tot_overlap = tot_overlap.copy().merge(probes_diffs['WASS2'], \n",
    "                              how='inner', \n",
    "                              left_index=True, \n",
    "                              right_index=True)\n",
    "tot_overlap = tot_overlap.loc[ppset]\n",
    "tot_overlap = tot_overlap.copy().merge(fc, left_index=True, right_index=True, how='inner')\n",
    "tot_overlap = tot_overlap.copy().merge(GeneMapping, left_index=True, right_index=True, how='inner')\n",
    "#tot_overlap['score_WASS1'] = tot_overlap['pvals_WASS1']\n",
    "#tot_overlap.drop(['pvals', 'pvals_WASS1', 'pvals_WASS2', 'score_WASS2'], axis=1, inplace=True)\n",
    "print(tot_overlap.shape)\n",
    "gc.collect()\n",
    "print(tot_overlap.columns)\n",
    "tot_overlap.sort_values('score_WASS1', ascending=False, inplace=True)\n",
    "tot_overlap.to_csv('/media/koekiemonster/DATA-FAST/genetic_expression/hackathon_2/Lung/_results/top_methyl_overlap_noPval.csv',\n",
    "          sep=';')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "129"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tot = probes_diffs['FDR'].merge(probes_diffs['KS'], \n",
    "                              how='inner', \n",
    "                              left_index=True, \n",
    "                              right_index=True, \n",
    "                              suffixes=['_FDR', '_KS'])\n",
    "tot = tot.copy().merge(probes_diffs['MWU'].rename(index=str, columns={'pvals': 'pvals_MWU', 'score': 'score_MWU'}), \n",
    "                              how='inner', \n",
    "                              left_index=True, \n",
    "                              right_index=True)\n",
    "tot = tot.copy().merge(probes_diffs['WASS1'].rename(index=str, columns={'score': 'score_WASS1'}), \n",
    "                              how='inner', \n",
    "                              left_index=True, \n",
    "                              right_index=True)\n",
    "tot = tot.copy().merge(probes_diffs['WASS2'].rename(index=str, columns={'score': 'score_WASS2'}), \n",
    "                              how='inner', \n",
    "                              left_index=True, \n",
    "                              right_index=True)\n",
    "tot = tot.copy().merge(fc, left_index=True, right_index=True, how='inner')\n",
    "tot = tot.copy().merge(GeneMapping, left_index=True, right_index=True, how='inner')\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tot.sort_values(by='score_FDR', ascending=False, inplace=True)\n",
    "tot.to_csv('/media/koekiemonster/DATA-FAST/genetic_expression/hackathon_2/Lung/_results/all_fc_scores.csv',\n",
    "          sep=';')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Only the non-mutual set of top-N probes \n",
    "ppset= None\n",
    "topM = 100\n",
    "for idx, _key in enumerate(probes_diffs.keys()):\n",
    "    if _key in ['AD', 'WASS2']:\n",
    "        break\n",
    "    else:\n",
    "        # query('pvals_'+_key+'<0.001')\n",
    "        pset = (set(probes_diffs[_key].sort_values(by='score_'+_key, ascending=False)[:topM].index.tolist()))\n",
    "        if idx>0:\n",
    "            ppset = pset.union(ppset)\n",
    "        else:\n",
    "            ppset = pset\n",
    "ppset = list(ppset)\n",
    "\n",
    "tot_exclusive = probes_diffs['FDR'].merge(probes_diffs['KS'], \n",
    "                              how='inner', \n",
    "                              left_index=True, \n",
    "                              right_index=True, \n",
    "                              suffixes=['_FDR', '_KS'])\n",
    "tot_exclusive = tot_exclusive.copy().merge(probes_diffs['MWU'], \n",
    "                              how='inner', \n",
    "                              left_index=True, \n",
    "                              right_index=True)\n",
    "tot_exclusive = tot_exclusive.copy().merge(probes_diffs['WASS1'], \n",
    "                              how='inner', \n",
    "                              left_index=True, \n",
    "                              right_index=True)\n",
    "tot_exclusive = tot_exclusive.copy().merge(probes_diffs['WASS2'], \n",
    "                              how='inner', \n",
    "                              left_index=True, \n",
    "                              right_index=True)\n",
    "tot_exclusive = tot_exclusive.loc[ppset]\n",
    "tot_exclusive = tot_exclusive.copy().merge(fc, left_index=True, right_index=True, how='inner')\n",
    "tot_exclusive = tot_exclusive.copy().merge(GeneMapping, left_index=True, right_index=True, how='inner')\n",
    "#tot_exclusive.drop(['pvals', 'pvals_WASS1', 'pvals_WASS2', 'score_WASS2'], axis=1, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(tot_exclusive.shape)\n",
    "gc.collect()\n",
    "tot_exclusive.sort_values('score_WASS1', ascending=False, inplace=True)\n",
    "tot_exclusive.to_csv('/media/koekiemonster/DATA-FAST/genetic_expression/hackathon_2/Lung/_results/top_methyl_exclusive_noPval.csv',\n",
    "          sep=';')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_N = 500\n",
    "topD = fc.sort_values(by='abs_fc_median', ascending=False)[:top_N]\n",
    "topA = fc.sort_values(by='abs_fc_median', ascending=True)[:top_N]\n",
    "topFC = pd.concat([topD, topA], axis=0)\n",
    "topFC = topFC.copy().merge(tot, how='inner', left_index=True, right_index=True)\n",
    "\n",
    "#topFC.drop('score_y', axis=1, inplace=True)\n",
    "#topFC.rename(index=str, columns={'score_x': 'score_Wass1'}, inplace=True)\n",
    "topFC.to_csv('/media/koekiemonster/DATA-FAST/genetic_expression/hackathon_2/Lung/_results/top_fc.csv',\n",
    "          sep=';')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_N = 100\n",
    "topD = LDA_scores.sort_values(by='LDA_score', ascending=False)[:top_N]\n",
    "topA = LDA_scores.sort_values(by='LDA_score', ascending=True)[:top_N]\n",
    "topLDA = pd.concat([topD, topA], axis=0)\n",
    "topLDA = topLDA.copy().merge(tot, how='inner', left_index=True, right_index=True)\n",
    "\n",
    "#topFC.drop('score_y', axis=1, inplace=True)\n",
    "#topFC.rename(index=str, columns={'score_x': 'score_Wass1'}, inplace=True)\n",
    "topLDA.to_csv('/media/koekiemonster/DATA-FAST/genetic_expression/hackathon_2/Lung/_results/top_lda.csv',\n",
    "          sep=';')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M = 15\n",
    "fig, ax = plt.subplots(nrows=M, figsize=(10,35))\n",
    "for i, probeid in enumerate(tot_overlap.index.tolist()[:M]):\n",
    "    Meth_data_T['LA'][probeid].plot.kde(color='red', ax=ax[i])\n",
    "    Meth_data_T['LSCC'][probeid].plot.kde(color='green', ax=ax[i])\n",
    "    _cleaned_ALL[probeid].plot.kde(color='black', ax=ax[i])\n",
    "    #ax[i].set_ylabel('')\n",
    "    #ax[i].set_title(probeid)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Unsupervised search for multi-modality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from unidip import UniDip\n",
    "# Use Hartigan's dip test recursively\n",
    "num_modes = []\n",
    "cols = _cleaned_ALL.columns\n",
    "\n",
    "n_procs = 10\n",
    "pool = multiprocessing.Pool(n_procs)\n",
    "\n",
    "def get_dips(i):\n",
    "    _col = cols[i]\n",
    "    dat = _cleaned_ALL[_col].sort_values()\n",
    "    intervals = UniDip(dat, alpha=0.05, ntrials=50, mrg_dst=5).run()\n",
    "    return {'probe': _col, 'num_modes': len(intervals)}\n",
    "\n",
    "num_modes_df = pd.DataFrame(pool.map(get_dips, range(0, len(cols))))\n",
    "num_modes_df.set_index('probe', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "multi_mod_probs = {}\n",
    "multi_mod_probs['1'] = num_modes_df[num_modes_df.num_modes==1].index.tolist()\n",
    "multi_mod_probs['2'] = num_modes_df[num_modes_df.num_modes==2].index.tolist()\n",
    "multi_mod_probs['3'] = num_modes_df[num_modes_df.num_modes==3].index.tolist()\n",
    "multi_mod_probs['4'] = num_modes_df[num_modes_df.num_modes==4].index.tolist()\n",
    "multi_mod_probs['5'] = num_modes_df[num_modes_df.num_modes==5].index.tolist()\n",
    "multi_mod_probs['6'] = num_modes_df[num_modes_df.num_modes==6].index.tolist()\n",
    "multi_mod_probs['7'] = num_modes_df[num_modes_df.num_modes==7].index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "l = '4'\n",
    "k = 1\n",
    "\n",
    "print(\"Gene\", GeneMapping.loc[multi_mod_probs[l][k]], \"\\n\")\n",
    "\n",
    "try: \n",
    "    tot_overlap.loc[multi_mod_probs[l][k]]\n",
    "except KeyError as e:\n",
    "    print(\"Not in index for tot_overlap :( \\n\")\n",
    "    \n",
    "try: \n",
    "    tot_exclusive.loc[multi_mod_probs[l][k]]\n",
    "except KeyError as e:\n",
    "    print(\"Not in index for tot_exclusive :( \\n\")\n",
    "    \n",
    "try: \n",
    "    top_fc.loc[multi_mod_probs[l][k]]\n",
    "except KeyError as e:\n",
    "    print(\"Not in index for top fc :( \\n\")\n",
    "    \n",
    "print(probes_diffs['FDR'].loc[multi_mod_probs[l][k]], \"\\n\")\n",
    "print(probes_diffs['KS'].loc[multi_mod_probs[l][k]], \"\\n\")\n",
    "print(probes_diffs['MWU'].loc[multi_mod_probs[l][k]], \"\\n\")\n",
    "print(probes_diffs['WASS1'].loc[multi_mod_probs[l][k]], \"\\n\")\n",
    "\n",
    "Meth_data_T['LA'].loc[:, multi_mod_probs[l][k]].plot.kde()\n",
    "Meth_data_T['LSCC'].loc[:, multi_mod_probs[l][k]].plot.kde()\n",
    "_cleaned_ALL.loc[:, multi_mod_probs[l][k]].plot.kde(color='black')\n",
    "plt.title(multi_mod_probs[l][k])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With the knowledge of multi-modality per dimension we can effectively filter out the noise to effectively perform unsupervised clustering. \n",
    "\n",
    "```\n",
    "cg00006626, NUP210L\n",
    "cg00001582, ZMIZ1,LOC283050\n",
    "cg00012036, H3F3B\n",
    "cg00087098, G3BP2\n",
    "cg00611101, TCEAL2\n",
    "...\n",
    "etc..\n",
    "```\n",
    "\n",
    "A benefit of this multi-modality search is the robustness in case of non-Gaussian, or non-standard distributions. \n",
    "How to find those probes that **do** seperate the targets but do not fulfill the prerequisites for the unpaired tests?\n",
    "\n",
    "* Multimodality -> paired samples comparison between groups, compared to sample comparison within group\n",
    "* apply supervised learner: LR(Bayesian), RF (random), XGB (gradient boosted), EBM (generalised additive model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Top differential RNA expression\n",
    "* KRT5 \n",
    "* DSC3\n",
    "* DSG3\n",
    "* TP63\n",
    "* MIR205HG\n",
    "* CALML3\n",
    "\n",
    "What are the corresponding probevalues? What are the methylation fold changes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supervised learner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.linear_model import LogisticRegressionCV as LR\n",
    "from sklearn.ensemble import ExtraTreesClassifier as ET\n",
    "from sklearn.ensemble import RandomForestClassifier as RF\n",
    "from xgboost import XGBClassifier as XGB\n",
    "from lightgbm import LGBMClassifier as LGBM\n",
    "from interpret.glassbox import ExplainableBoostingClassifier as EB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# impose some filters \n",
    "# * minimum abs fold change of 0.1 (beta)\n",
    "#filter_1 = fc.loc[(fc.abs_fc>0.1) | (fc.abs_fc<-0.1)].index.tolist() \n",
    "x_filtered = x#.loc[:, filter_1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# vanilla\n",
    "mod_LR = LR(n_jobs=4)\n",
    "mod_ET = ET(n_estimators=7500, n_jobs=4)\n",
    "mod_RF = RF(n_estimators=7500, n_jobs=4)\n",
    "mod_XGB = XGB(n_estimators=500, n_jobs=4)\n",
    "#mod_LGBM = LGBM(n_estimators=150, n_jobs=4)\n",
    "#mod_EB = EB(n_estimators=32, n_jobs=4)\n",
    "\n",
    "#cols = np.array(x.columns.tolist())\n",
    "#lr_coefs = mod_LR.coef_.reshape([-1,1])[:,0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_split.py:2053: FutureWarning:\n",
      "\n",
      "You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "LogisticRegressionCV(Cs=10, class_weight=None, cv='warn', dual=False,\n",
       "           fit_intercept=True, intercept_scaling=1.0, max_iter=100,\n",
       "           multi_class='warn', n_jobs=4, penalty='l2', random_state=None,\n",
       "           refit=True, scoring=None, solver='lbfgs', tol=0.0001, verbose=0)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc.collect()\n",
    "mod_LR.fit(x_filtered,y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc.collect()\n",
    "mod_ET.fit(x_filtered,y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc.collect()\n",
    "mod_RF.fit(x_filtered,y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc.collect()\n",
    "mod_XGB.fit(x_filtered,y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract importances\n",
    "coefs_LR =pd.DataFrame(data=mod_LR.coef_.reshape([-1,1]), index=x_filtered.columns, columns=['LR_coeff'])\n",
    "coefs_LR['LR_coeff_abs'] = coefs_LR.LR_coeff.abs()\n",
    "coefs_LR['LR_coeff_sign'] = np.sign(coefs_LR.LR_coeff)\n",
    "coefs_LR.drop('LR_coeff', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coefs_RF = pd.DataFrame(data=mod_RF.feature_importances_, index=x_filtered.columns,columns=['RF_importance'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coefs_ET = pd.DataFrame(data=mod_ET.feature_importances_, index=x_filtered.columns,columns=['ET_importance'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coefs_XGB = pd.DataFrame(data=mod_XGB.feature_importances_, index=x_filtered.columns,columns=['XGB_importance'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "42"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coefs_list = [coefs_LR] #, coefs_RF, coefs_ET, coefs_XGB]\n",
    "sort_by = 'score_FDR'\n",
    "comb = pd.concat(coefs_list, axis=1)\n",
    "comb = comb.copy().merge(tot, how='left', left_index=True, right_index=True)\n",
    "comb.sort_values(by=sort_by, ascending=False, inplace=True)\n",
    "comb.to_csv('/media/koekiemonster/DATA-FAST/genetic_expression/hackathon_2/Lung/_results/all_res.csv',\n",
    "          sep=';')\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "splt = fc[['abs_fc_median', 'split_fc']].reset_index()\n",
    "tmp=pd.DataFrame(data=x[y==0].std(axis=0), columns=['std_LA'])\n",
    "splt = splt.copy().merge(tmp, how='inner', left_on='index', right_index=True)\n",
    "tmp=pd.DataFrame(data=x[y==1].std(axis=0), columns=['std_LSCC'])\n",
    "splt = splt.copy().merge(tmp, how='inner', left_on='index', right_index=True)\n",
    "splt['adapted_fold_change'] = 2*splt.abs_fc_median/(splt.std_LA+splt.std_LSCC) # correct for large std's\n",
    "splt['half_dist'] = np.abs(0.5-splt.split_fc)\n",
    "# you can introduce the effect of Kurtosis and Skewness as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "def super_simple_model(x, y, splt, min_lscc=0.75, min_la=0.75, max_consecutive_fail=50, adapted_sort=False):\n",
    "    # not optimized, super slow\n",
    "    rnk_var = 'adapted_fold_change' if adapted_sort else 'abs_fc_median'\n",
    "    splt = splt.copy().reindex(splt[rnk_var].abs().sort_values(ascending=False).index)\n",
    "    \n",
    "    x_train = x.copy()\n",
    "    x_train['y'] = y\n",
    "    \n",
    "    fail = 0 \n",
    "    rcount = 0\n",
    "    tcount = 0\n",
    "    rules = []\n",
    "    for row in splt.iterrows():\n",
    "        split_gen = row[1]['index']\n",
    "        fc_split =row[1]['split_fc']\n",
    "        #v1 = x_train.loc[x[split_gen]<fc_split].y.mean()        \n",
    "        #v2 = x_train.loc[x[split_gen]>fc_split].y.mean()\n",
    "        \n",
    "        xg = x[split_gen].values\n",
    "        yv = x_train.y.values\n",
    "        v1 = np.mean(yv[np.where(xg<fc_split)])\n",
    "        v2 = np.mean(yv[np.where(xg>fc_split)])\n",
    "\n",
    "        acc_lscc = np.max([v1, v2])\n",
    "        acc_la = np.max([1-v1,1-v2])\n",
    "       \n",
    "        if v2>v1:\n",
    "            rec_lscc = np.sum(yv[np.where(xg>fc_split)])/373\n",
    "            rec_la = np.sum(1-yv[np.where(xg<fc_split)])/491\n",
    "        else:\n",
    "            rec_lscc = np.sum(yv[np.where(xg<fc_split)])/373\n",
    "            rec_la = np.sum(1-yv[np.where(xg>fc_split)])/491 \n",
    "        tcount +=1\n",
    "        if (acc_lscc>min_lscc) and (acc_la>min_la):\n",
    "            rules.append({'acc_lscc': acc_lscc, \n",
    "                          'acc_la': acc_la, \n",
    "                          'rec_lscc': rec_lscc, \n",
    "                          'rec_la': rec_la,\n",
    "                          'split' : fc_split, \n",
    "                          'dim': split_gen, \n",
    "                          'sign': v2>v1}) \n",
    "            rcount += 1\n",
    "            fail=0\n",
    "            if rcount%500==0:\n",
    "                print(\"Contains:{} rules, processed {} probes\".format(rcount, tcount))\n",
    "        else:\n",
    "            fail+=1\n",
    "        if max_consecutive_fail is not None:\n",
    "            if (fail>max_consecutive_fail) and (max_consecutive_fail!=-1):\n",
    "                return pd.DataFrame(rules)\n",
    "    return pd.DataFrame(rules)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contains:500 rules, processed 500 probes\n",
      "Contains:1000 rules, processed 1000 probes\n",
      "Contains:1500 rules, processed 1500 probes\n",
      "Contains:2000 rules, processed 2000 probes\n",
      "Contains:2500 rules, processed 2500 probes\n",
      "Contains:3000 rules, processed 3000 probes\n",
      "Contains:3500 rules, processed 3501 probes\n",
      "Contains:4000 rules, processed 4001 probes\n",
      "Contains:4500 rules, processed 4502 probes\n",
      "Contains:5000 rules, processed 5003 probes\n",
      "Contains:5500 rules, processed 5505 probes\n",
      "Contains:6000 rules, processed 6006 probes\n",
      "Contains:6500 rules, processed 6509 probes\n",
      "Contains:7000 rules, processed 7010 probes\n",
      "Contains:7500 rules, processed 7519 probes\n",
      "Contains:8000 rules, processed 8028 probes\n",
      "Contains:8500 rules, processed 8541 probes\n",
      "Contains:9000 rules, processed 9057 probes\n",
      "Contains:9500 rules, processed 9572 probes\n",
      "Contains:10000 rules, processed 10093 probes\n",
      "Contains:10500 rules, processed 10624 probes\n",
      "Contains:11000 rules, processed 11163 probes\n",
      "Contains:11500 rules, processed 11690 probes\n",
      "Contains:12000 rules, processed 12228 probes\n",
      "Contains:12500 rules, processed 12757 probes\n",
      "Contains:13000 rules, processed 13297 probes\n",
      "Contains:13500 rules, processed 13852 probes\n",
      "Contains:14000 rules, processed 14418 probes\n",
      "Contains:14500 rules, processed 14996 probes\n",
      "Contains:15000 rules, processed 15585 probes\n",
      "Contains:15500 rules, processed 16177 probes\n",
      "Contains:16000 rules, processed 16788 probes\n",
      "Contains:16500 rules, processed 17399 probes\n",
      "Contains:17000 rules, processed 18045 probes\n",
      "Contains:17500 rules, processed 18681 probes\n",
      "Contains:18000 rules, processed 19353 probes\n",
      "Contains:18500 rules, processed 20018 probes\n",
      "Contains:19000 rules, processed 20719 probes\n",
      "Contains:19500 rules, processed 21423 probes\n",
      "Contains:20000 rules, processed 22171 probes\n",
      "Contains:20500 rules, processed 22956 probes\n",
      "Contains:21000 rules, processed 23725 probes\n",
      "Contains:21500 rules, processed 24575 probes\n",
      "Contains:22000 rules, processed 25470 probes\n",
      "Contains:22500 rules, processed 26353 probes\n",
      "Contains:23000 rules, processed 27335 probes\n",
      "Contains:23500 rules, processed 28317 probes\n",
      "Contains:24000 rules, processed 29452 probes\n",
      "Contains:24500 rules, processed 30558 probes\n",
      "Contains:25000 rules, processed 31847 probes\n",
      "Contains:25500 rules, processed 33267 probes\n",
      "Contains:26000 rules, processed 34722 probes\n",
      "Contains:26500 rules, processed 36330 probes\n",
      "Contains:27000 rules, processed 38216 probes\n",
      "Contains:27500 rules, processed 40142 probes\n",
      "Contains:28000 rules, processed 42485 probes\n",
      "Contains:28500 rules, processed 44895 probes\n",
      "Contains:29000 rules, processed 47482 probes\n",
      "Contains:29500 rules, processed 50168 probes\n",
      "Contains:30000 rules, processed 53053 probes\n",
      "Contains:30500 rules, processed 56440 probes\n",
      "Contains:31000 rules, processed 59969 probes\n",
      "Contains:31500 rules, processed 63878 probes\n",
      "Contains:32000 rules, processed 68047 probes\n",
      "Contains:32500 rules, processed 72300 probes\n",
      "Contains:33000 rules, processed 76938 probes\n",
      "Contains:33500 rules, processed 81849 probes\n",
      "Contains:34000 rules, processed 87054 probes\n",
      "Contains:34500 rules, processed 92974 probes\n",
      "Contains:35000 rules, processed 99673 probes\n",
      "Contains:35500 rules, processed 107509 probes\n",
      "Contains:36000 rules, processed 116430 probes\n",
      "Contains:36500 rules, processed 126232 probes\n",
      "Contains:37000 rules, processed 137110 probes\n",
      "Contains:37500 rules, processed 150145 probes\n",
      "Contains:38000 rules, processed 164890 probes\n",
      "Contains:38500 rules, processed 184349 probes\n",
      "Contains:39000 rules, processed 214521 probes\n"
     ]
    }
   ],
   "source": [
    "split_genes = super_simple_model(x, y, splt, min_lscc=0.85, min_la=0.85, max_consecutive_fail=None, adapted_sort=True);\n",
    "split_genes.set_index('dim', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "#split_genes = split_genes.copy().merge(score_list_all['raw'], how='left', left_index=True, right_index=True)\n",
    "split_genes.rename(index=str, columns={'sign': 'lscc_on_the_rightside_of_split'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge with tot\n",
    "#fc = fc.copy().merge(GeneMapping, how='left', left_index=True, right_index=True) # tot\n",
    "split_genes = split_genes.copy().merge(comb, how='left', left_index=True, right_index=True)\n",
    "split_genes['acc_avg'] = split_genes[['acc_la', 'acc_lscc']].mean(axis=1)\n",
    "split_genes['abs_abs_fc_median'] = split_genes.abs_fc_median.abs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "split_genes.sort_values(by='abs_abs_fc_median', ascending=False, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "split_genes.to_csv(\"/media/koekiemonster/DATA-FAST/genetic_expression/hackathon_2/Lung/_results/Methylation_best_splits_all.csv\", sep=\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "split_genes.query('abs_abs_fc_median>0.1').to_csv(\"/media/koekiemonster/DATA-FAST/genetic_expression/hackathon_2/Lung/_results/Methylation_best_splits.csv\", sep=\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAz0AAAI/CAYAAABd17fKAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAd7UlEQVR4nO3df7Dld13f8debRMoPkR+ypgyhXtSIpa0IrlTHagXEwaEGrBRh1IkOmlqh6mhbVsu02B8zqK1oO7ZjBNtotYDxB9H1RyGiHTsVWISKJNJEDBoEsiKIvwoG3v1jzw6XuMueZc/3nHvf9/GY2dnz/Z5z73nf+83d3We+53y+1d0BAACY6h67HgAAAGBJogcAABhN9AAAAKOJHgAAYDTRAwAAjCZ6AACA0S7f9QDrePCDH9x7e3u7HgMAADigXve61/1Bdx87132HInr29vZy6tSpXY8BAAAcUFX11vPd5+VtAADAaKIHAAAYTfQAAACjiR4AAGA00QMAAIwmegAAgNFEDwAAMJroAQAARhM9AADAaKIHAAAYTfQAAACjiR4AAGA00QMAAIwmegAAgNFEDwAAMJroAQAARhM9AADAaKIHAAAYTfQAAACjiR4AAGA00QMAAIwmegAAgNFEDwAAMJroAQAARhM9AADAaKIHAAAYTfQAAACjiR4AAGA00QMAAIwmegAAgNFEDwAAMJroAQAARhM9AADAaKIHAAAYTfQAAACjiR4AAGA00QMAAIwmegAAgNFEDwAAMJroAQAARhM9AADAaKIHAAAYTfQAAACjiR4AAGA00QMAAIwmegAAgNFEDwAAMJroAQAARhM9AADAaKIHAAAYTfQAAACjiZ4Dbu/EyV2PAAAAh5roAQAARhM9AADAaKIHAAAYTfQAAACjiR4AAGA00QMAAIy2aPRU1QOq6oaq+q2quqWqPqeqHlRVr6iqW1e/P3DJGQAAgKNt6TM935fkF7r705I8KsktSU4kuam7r0py02obAABgEYtFT1XdP8nnJ3lxknT3+7v7PUmekuT61cOuT/LUpWY4rFyQFAAANmfJMz0PT3I6yX+pqtdX1Yuq6r5Jrujut68e844kVyw4AwAAcMQtGT2XJ3lMkv/c3Y9O8qe520vZuruT9Lk+uKqurapTVXXq9OnTC44JAABMtmT03JHkju5+9Wr7hpyJoHdW1UOSZPX7nef64O6+rruPd/fxY8eOLTgmAAAw2WLR093vSPJ7VfWI1a4nJLk5yY1JrlntuybJy5eaAQAA4PKFP/8/TvKjVXXPJG9J8jU5E1ovq6pnJXlrkqcvPAMAAHCELRo93f2GJMfPcdcTlnxeAACAs5a+Tg8AAMBOiR4AAGA00QMAAIwmegAAgNFEDwAAMJroAQAARhM9AADAaKIHAAAYTfQAAACjiZ4DZO/EyV2PAAAA44geAABgNNEDAACMJnoAAIDRRA8AADCa6AEAAEYTPQAAwGiiBwAAGE30AAAAo4keAABgNNEDAACMJnoAAIDRRA8AADCa6AEAAEYTPQAAwGiiBwAAGE30AAAAo4keAABgNNEDAACMJnoAAIDRRA8AADCa6AEAAEYTPQAAwGiiBwAAGE30AAAAo4keAABgNNEDAACMJnoAAIDRRA8AADCa6AEAAEYTPQAAwGiiBwAAGE30AAAAo4keAABgNNEDAACMJnoAAIDRRA8AADCa6AEAAEYTPQAAwGiiBwAAGE30AAAAo4keAABgNNEDAACMJnoAAIDRRA8AADCa6AEAAEYTPQAAwGiiBwAAGE30AAAAo4keAABgNNEDAACMJnoAAIDRRA8AADCa6AEAAEYTPQAAwGiiBwAAGE30AAAAo4keAABgNNEDAACMJnoAAIDRRA8AADCa6AEAAEa7fMlPXlW3J/njJB9Icld3H6+qByV5aZK9JLcneXp3v3vJOQAAgKNrG2d6Htfdn9Hdx1fbJ5Lc1N1XJblptQ0AALCIXby87SlJrl/dvj7JU3cww4G3d+LkrkcAAIARlo6eTvI/qup1VXXtat8V3f321e13JLli4RkAAIAjbOno+Tvd/ZgkX5zk2VX1+fvv7O7OmTD6S6rq2qo6VVWnTp8+vfCYh4OzPwAAcPEWjZ7uftvq9zuT/FSSxyZ5Z1U9JElWv995no+9rruPd/fxY8eOLTkmAAAw2GLRU1X3rar7nb2d5IuS/GaSG5Ncs3rYNUlevtQMAAAASy5ZfUWSn6qqs8/zY939C1X12iQvq6pnJXlrkqcvOAMAAHDELRY93f2WJI86x/53JXnCUs8LAACw3y6WrAYAANga0QMAAIwmegAAgNFEDwAAMJroAQAARhM9AADAaKIHAAAYTfQAAACjiR4AAGA00XPA7J04uesRAABgFNEDAACMJnoAAIDRRA8AADCa6AEAAEYTPQAAwGiiBwAAGE30AAAAo4keAABgNNEDAACMJnoAAIDRRA8AADCa6DnA9k6c3PUIAABw6IkeAABgNNEDAACMJnoAAIDRRA8AADCa6AEAAEYTPQAAwGiiBwAAGE30HBCuyQMAAMsQPQAAwGiiBwAAGE30AAAAo4keAABgNNEDAACMJnoAAIDRRA8AADCa6AEAAEYTPTvmoqQAALAs0QMAAIwmegAAgNFEDwAAMJroAQAARhM9AADAaKLnELDCGwAAfPREDwAAMJroAQAARhM9AADAaKIHAAAYTfQAAACjiR4AAGA00QMAAIwmegAAgNFEDwAAMJroAQAARhM9AADAaKIHAAAYTfQAAACjiR4AAGA00XNI7Z04uesRAADgUBA9AADAaKIHAAAYTfQAAACjiR4AAGA00QMAAIwmegAAgNFEDwAAMJroAQAARhM9AADAaKIHAAAYTfQAAACjLR49VXVZVb2+qn52tf3wqnp1Vd1WVS+tqnsuPQMAAHB0beNMzzcluWXf9ncmeWF3f0qSdyd51hZmOPT2Tpzc9QgAAHAoLRo9VXVlkicnedFqu5I8PskNq4dcn+SpS84AAAAcbUuf6fneJP8syQdX2x+f5D3dfddq+44kD114BgAA4AhbLHqq6u8lubO7X/dRfvy1VXWqqk6dPn16w9MBAABHxZJnej43ydVVdXuSl+TMy9q+L8kDqury1WOuTPK2c31wd1/X3ce7+/ixY8cWHBMAAJhssejp7m/r7iu7ey/JM5L8Und/RZJXJXna6mHXJHn5UjMAAADs4jo9z03yLVV1W868x+fFO5gBAAA4Ii6/8EMuXXf/cpJfXt1+S5LHbuN5p7J8NQAArG8XZ3q4GxEDAADLET0AAMBoogcAABhN9AAAAKOJHgAAYDTRAwAAjCZ6AACA0UQPAAAwmugBAABGEz0AAMBoogcAABhN9AAAAKOJHgAAYDTRAwAAjCZ6AACA0UQPAAAwmugBAABGEz0AAMBoa0VPVf2tpQcBAABYwrpnev5TVb2mqr6hqu6/6EQAAAAbtFb0dPfnJfmKJA9L8rqq+rGqeuKikwEAAGzA2u/p6e5bkzwvyXOT/N0k/6Gqfquq/v5Sw/GX7Z04uesRAADgUFn3PT2fXlUvTHJLkscn+ZLu/uur2y9ccD4AAIBLcvmaj/uPSV6U5Nu7+8/P7uzu36+q5y0yGQAAwAasGz1PTvLn3f2BJKmqeyS5V3f/WXf/yGLTAQAAXKJ139PzyiT33rd9n9U+AACAA23d6LlXd//J2Y3V7fssMxIAAMDmrBs9f1pVjzm7UVWfmeTPP8LjAQAADoR139PzzUl+vKp+P0kl+atJvnyxqQAAADZkrejp7tdW1aclecRq15u7+y+WGwsAAGAz1j3TkySflWRv9TGPqap09w8vMhUAAMCGrBU9VfUjST45yRuSfGC1u5OIHgAA4EBb90zP8SSP7O5echgAAIBNW3f1tt/MmcULAAAADpV1z/Q8OMnNVfWaJO87u7O7r15kKgAAgA1ZN3qev+QQAAAAS1l3yepfqapPTHJVd7+yqu6T5LJlRwMAALh0a72np6q+LskNSX5gteuhSX56qaEAAAA2Zd2FDJ6d5HOTvDdJuvvWJJ+w1FAAAACbsm70vK+73392o6ouz5nr9AAAABxo60bPr1TVtye5d1U9McmPJ/mZ5cYCAADYjHWj50SS00nemOQfJvm5JM9baigAAIBNWXf1tg8m+cHVLwAAgENjreipqt/JOd7D092ftPGJAAAANmjdi5Me33f7Xkn+QZIHbX4cAACAzVrrPT3d/a59v97W3d+b5MkLzwYAAHDJ1n1522P2bd4jZ878rHuWCAAAYGfWDZd/v+/2XUluT/L0jU9zxOydOLnrEQAAYLx1V2973NKDcPH2TpzM7S/wKkMAAPhI1n1527d8pPu7+3s2Mw4AAMBmXczqbZ+V5MbV9pckeU2SW5cYCgAAYFPWjZ4rkzymu/84Sarq+UlOdvdXLjUYAADAJqy1ZHWSK5K8f9/2+1f7AAAADrR1o+eHk7ymqp6/Osvz6iTXLzYVF8UqcAAAcH7rrt72b6vq55N83mrX13T365cbCwAAYDPWPdOTJPdJ8t7u/r4kd1TVwxeaCQAAYGPWip6q+pdJnpvk21a7PibJf1tqKAAAgE1Z90zPlya5OsmfJkl3/36S+y01FAAAwKasGz3v7+5O0klSVfddbiQAAIDNWTd6XlZVP5DkAVX1dUlemeQHlxsLAABgM9Zdve3fVdUTk7w3ySOS/IvufsWikwEAAGzABaOnqi5L8sruflwSoQMAABwqF3x5W3d/IMkHq+r+W5gHAABgo9Z6eVuSP0nyxqp6RVYruCVJd3/jIlMBAABsyLrR85OrXwAAAIfKR4yeqvpr3f273X39tgYCAADYpAu9p+enz96oqp9YeBYAAICNu1D01L7bn7TkIAAAAEu4UPT0eW4DAAAcChdayOBRVfXenDnjc+/V7ay2u7s/btHpAAAALtFHjJ7uvmxbgwAAACzhghcn/WhV1b2q6jVV9X+q6k1V9R2r/Q+vqldX1W1V9dKquudSMwAAACwWPUnel+Tx3f2oJJ+R5ElV9dlJvjPJC7v7U5K8O8mzFpwBAAA44haLnj7jT1abH7P61Uken+SG1f7rkzx1qRkAAACWPNOTqrqsqt6Q5M4kr0jy20ne0913rR5yR5KHLjkDAABwtC0aPd39ge7+jCRXJnlskk9b92Or6tqqOlVVp06fPr3YjAAAwGyLRs9Z3f2eJK9K8jlJHlBVZ1eNuzLJ287zMdd19/HuPn7s2LFtjAkAAAy05Optx6rqAavb907yxCS35Ez8PG31sGuSvHypGQAAAC50cdJL8ZAk11fVZTkTVy/r7p+tqpuTvKSq/k2S1yd58YIzAAAAR9xi0dPdv5Hk0efY/5aceX8PAADA4rbynh4AAIBdET0AAMBoogcAABhN9AAAAKOJHgAAYDTRAwAAjCZ6AACA0UQPAAAwmugBAABGEz0AAMBoogcAABhN9AAAAKOJHgAAYDTRAwAAjCZ6Drm9EyfPeRsAADhD9AAAAKOJHgAAYDTRAwAAjCZ6AACA0UQPAAAwmugBAABGEz0AAMBoogcAABhN9OyAi4gCAMD2iB4AAGA00QMAAIwmenbES9wAAGA7RA8AADCa6AEAAEYTPQAAwGiiBwAAGE30AAAAo4keAABgNNEDAACMJnoAAIDRRA8AADCa6AEAAEYTPUPsnTi56xEAAOBAEj0AAMBoogcAABhN9AAAAKOJHgAAYDTRAwAAjCZ6AACA0UQPAAAwmugBAABGEz0AAMBoogcAABhN9AAAAKOJHgAAYDTRAwAAjCZ6AACA0UQPAAAwmugBAABGEz0AAMBoogcAABhN9AAAAKOJHgAAYDTRAwAAjCZ6AACA0UQPAAAwmugBAABGEz0AAMBoogcAABhN9AAAAKOJHgAAYDTRAwAAjCZ6AACA0UTPUHsnTu56BAAAOBBEDwAAMJroAQAARlsseqrqYVX1qqq6uareVFXftNr/oKp6RVXduvr9gUvNAAAAsOSZnruSfGt3PzLJZyd5dlU9MsmJJDd191VJblptAwAALGKx6Onut3f3r69u/3GSW5I8NMlTkly/etj1SZ661AxHkQUMAADgw23lPT1VtZfk0UleneSK7n776q53JLliGzMAAABH0+LRU1Ufm+Qnknxzd793/33d3Un6PB93bVWdqqpTp0+fXnpMAABgqEWjp6o+JmeC50e7+ydXu99ZVQ9Z3f+QJHee62O7+7ruPt7dx48dO7bkmAAAwGBLrt5WSV6c5Jbu/p59d92Y5JrV7WuSvHypGQAAAC5f8HN/bpKvSvLGqnrDat+3J3lBkpdV1bOSvDXJ0xecAQAAOOIWi57u/tUkdZ67n7DU8wIAAOy3ldXbAAAAdkX0DORaPQAA8CGiBwAAGE30AAAAo4meLfPSMwAA2C7RAwAAjCZ6AACA0UTPFnlpGwAAbJ/oAQAARhM9AADAaKIHAAAYTfQAAACjiR4AAGA00QMAAIwmegAAgNFEDwAAMJroAQAARhM9AADAaKIHAAAYTfQAAACjiR4AAGA00QMAAIwmegAAgNFEDwAAMJroAQAARhM9AADAaKIHAAAYTfQAAACjiR4AAGA00QMAAIwmegAAgNFEDwAAMJroAQAARhM9AADAaKIHAAAYTfQAAACjiZ7B9k6c3PUIAACwc6IHAAAYTfQAAACjiR4AAGA00QMAAIwmegAAgNFEz5ZYSQ0AAHZD9AAAAKOJHgAAYDTRAwAAjCZ6AACA0UQPAAAwmug5AqwcBwDAUSZ6AACA0UQPAAAwmugZzkvbAAA46kQPAAAwmugBAABGEz0AAMBoogcAABhN9AAAAKOJHgAAYDTRAwAAjCZ6AACA0UQPAAAwmugBAABGEz0AAMBoogcAABhN9AAAAKOJHgAAYDTRAwAAjCZ6AACA0UQPAAAwmug5IvZOnNz1CAAAsBOiBwAAGG2x6KmqH6qqO6vqN/fte1BVvaKqbl39/sClnp9zc8YHAICjZskzPf81yZPutu9Ekpu6+6okN622AWB3nn//D/0CYKTFoqe7/2eSP7zb7qckuX51+/okT13q+QEAAJLtv6fniu5+++r2O5JcseXnBwAAjpidLWTQ3Z2kz3d/VV1bVaeq6tTp06e3OBkAADDJtqPnnVX1kCRZ/X7n+R7Y3dd19/HuPn7s2LGtDQgAAMyy7ei5Mck1q9vXJHn5lp8fAAA4YpZcsvq/J/nfSR5RVXdU1bOSvCDJE6vq1iRfuNpmSyxXDQDAUXT5Up+4u595nruesNRzAgAA3N1i0QPAMPuvY/P8P9rdHBfjMM4MwMbtbPU2AACAbRA9AADAaKIHAAAYTfQAAACjiR4AAGA00QMAAIxmyWoADr8llqY+3+dcehlsy2wDbJwzPQAAwGiiBwAAGE30AAAAo4keAABgNNGzBXsnTu56BAAAOLJEDwAAMJroAQAARnOdHoDD7DBe02X/zB+2/wDPf96ZD+H3f7/DPj/AmpzpAQAARhM9AADAaKIHAAAYTfQAAACjiR4AAGA00QMAAIxmyeojau/EySTJ7S948o4ngYEsA7xZ5/t+nm8Z6cNo0tcCcAA503MEnQ0eAAA4CkQPAAAwmugBAABGEz0AAMBoogcAABhN9AAAAKOJngUdllXSDsucAADw0XCdHgAuzUG7LtGlXPNmievlbOr7s871is73+Q/aMQLYMmd6AACA0UTPwrx0DAAAdkv0HHGiDACA6UQPAAAwmugBAABGEz0AAMBolqxmPku1ctCts0zy+ZYo3ubzXuzHXqxt/qwuMf82nW/+i12+GuCIcKYHAAAYTfQAAACjiR4AAGA00QMAAIwmehbiop8AAHAwiB4AAGA00QMAAIzmOj0kOfNyvNtf8ORdj7E7ruXDpbrY66bs6lopF/u8k67psvTXMul7tSvr/Lyss/+gOSxzwmDO9CzA+3kAAODgED0AAMBoogcAABhN9PBhvDQPAIBpRA8AADCa6AEAAEazZPUGHfZln8++tO3s13HYvx4OsSWWd13nc17s825zieLDvhzyppbKPmrfh20+16ZmW2dp9l0t327paDiynOkBAABGEz18RBY2AADgsBM9GyYSAADgYBE9AADAaKKHc3LGCgCAKUQPaxFBAAAcVqIHAAAYzXV6gMNvm9f1OcjXiTkIMx/k7w/bsav/Bta5JtD5Hr/ENXsu5TpJF3udsPP9vK+z/2Jn+7D9B+xaR5v6HjKSMz0bMvnlX5O/NgAA5hM9XJS9EydFEAAAh4roAQAARhM9G3BUznzs/zqPytcMAMDhJ3q4JOIHAICDTvRwyYQPAAAHmSWrYV1HfSnMdb6ei33Mh+3f0BKqS3/spbCcM5yxzs/Cef+sWPjnaK3ZLmFJ+G3+XbL0bBdrnT//l14e+yD8OTzh3weH8GtwpucSHPUzHEf96wcA4HAQPWyM5awBADiIRA8bcb7YEUEAAOzaTqKnqp5UVW+uqtuq6sQuZrhU/jF/fs74AABwkGw9eqrqsiTfn+SLkzwyyTOr6pHbnmNT/OP+/O7+vTm7vf933z8AAJa2i9XbHpvktu5+S5JU1UuSPCXJzTuY5aL5R/rFu9D3bO/Eydz+gid/2GPPbgMAwKXaxcvbHprk9/Zt37HaxxFw97M9d99/933n+rXu5xGoAAAkSXX3dp+w6mlJntTdX7va/qokf7u7n3O3x12b5NrV5iOSvHmDYzw4yR9s8PNxcDnWR4vjfbQ43keL4320ON5HxyaP9Sd297Fz3bGLl7e9LcnD9m1fudr3Ybr7uiTXLTFAVZ3q7uNLfG4OFsf6aHG8jxbH+2hxvI8Wx/vo2Nax3sXL216b5KqqenhV3TPJM5LcuIM5AACAI2DrZ3q6+66qek6SX0xyWZIf6u43bXsOAADgaNjFy9vS3T+X5Od28dwri7xsjgPJsT5aHO+jxfE+Whzvo8XxPjq2cqy3vpABAADANu3iPT0AAABbMzZ6qupJVfXmqrqtqk6c4/6/UlUvXd3/6qra2/6UbMoax/vzq+rXq+qu1bLpHGJrHO9vqaqbq+o3quqmqvrEXczJZqxxvL++qt5YVW+oql+tqkfuYk4240LHe9/jvqyquqqs8HVIrfGz/dVVdXr1s/2GqvraXczJZqzzs11VT1/9/f2mqvqxjT7/xJe3VdVlSf5vkifmzMVPX5vkmd19877HfEOST+/ur6+qZyT50u7+8p0MzCVZ83jvJfm4JP8kyY3dfcP2J2UT1jzej0vy6u7+s6r6R0m+wM/34bTm8f647n7v6vbVSb6hu5+0i3m5NOsc79Xj7pfkZJJ7JnlOd5/a9qxcmjV/tr86yfG7X8uRw2fN431VkpcleXx3v7uqPqG779zUDFPP9Dw2yW3d/Zbufn+SlyR5yt0e85Qk169u35DkCVVVW5yRzbng8e7u27v7N5J8cBcDslHrHO9XdfefrTZ/LWeuB8bhtM7xfu++zfsmmfd/846Odf7+TpJ/neQ7k/y/bQ7HRq17rJlhneP9dUm+v7vfnSSbDJ5kbvQ8NMnv7du+Y7XvnI/p7ruS/FGSj9/KdGzaOsebOS72eD8ryc8vOhFLWut4V9Wzq+q3k3xXkm/c0mxs3gWPd1U9JsnDuvvkNgdj49b9s/zLVi9VvqGqHnaO+zkc1jnen5rkU6vqf1XVr1XVRs/YT40egFTVVyY5nuS7dz0Ly+ru7+/uT07y3CTP2/U8LKOq7pHke5J8665nYSt+Jsled396klfkQ6/QYabLk1yV5AuSPDPJD1bVAzb1yadGz9uS7P+/AVeu9p3zMVV1eZL7J3nXVqZj09Y53syx1vGuqi9M8s+TXN3d79vSbGzexf58vyTJUxediCVd6HjfL8nfTPLLVXV7ks9OcqPFDA6lC/5sd/e79v35/aIkn7ml2di8df4svyNn3nf9F939OznzHqCrNjXA1Oh5bZKrqurhVXXPJM9IcuPdHnNjkmtWt5+W5Jd64qoOR8M6x5s5Lni8q+rRSX4gZ4Jno68JZuvWOd77/1J8cpJbtzgfm/URj3d3/1F3P7i797p7L2fes3e1hQwOpXV+th+yb/PqJLdscT42a51/q/10zpzlSVU9OGde7vaWTQ1w+aY+0UHS3XdV1XOS/GKSy5L8UHe/qar+VZJT3X1jkhcn+ZGqui3JH+bMN59DaJ3jXVWfleSnkjwwyZdU1Xd099/Y4dh8lNb8+f7uJB+b5MdX65P8bndfvbOh+aitebyfszqz9xdJ3p0P/Q8tDpk1jzcDrHmsv3G1IuNdOfNvta/e2cBckjWP9y8m+aKqujnJB5L80+7e2KuwRi5ZDQAAcNbUl7cBAAAkET0AAMBwogcAABhN9AAAAKOJHgAAYDTRAwAAjCZ6AACA0UQPAAAw2v8HQPNgS/leM6sAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1008x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "split_genes.query('abs_abs_fc_median<0.1').abs_abs_fc_median.plot.hist(bins=200, figsize=(14,10), density=True)\n",
    "split_genes.query('abs_abs_fc_median>0.1').abs_abs_fc_median.plot.hist(bins=200, figsize=(14,10), density=True)\n",
    "plt.savefig('/home/koekiemonster/Pictures/methyl_groups.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Paired testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# paired test\n",
    "# KLD\n",
    "from scipy.special import kl_div\n",
    "from scipy.special import rel_entr\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Paired comparison -> probewise\n",
    "i.e. per sample we have a vector of probe values \n",
    "For each target group we have a different set of probe value vectors. \n",
    "\n",
    "This type of analysis can provide us with sample clusters that may be predictive of target groups.\n",
    "\n",
    "Typical methods are: KLD, JSD, spatial distance (abs cosine, spearman, pearson, euclidean..)\n",
    "\n",
    "## Unpaired comparison -> sample wise\n",
    "i.e. per probe we have vector sample values. \n",
    "\n",
    "This type of analysis can provide us with some metric that describes how well the target groups are seperated per probe.\n",
    "\n",
    "Typical methods are: KS, MWU, Wasserstein, FDR, FPR "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Key results\n",
    "\n",
    "* top differential methylation probes and corresponding genes\n",
    "* list of fold changes -> write to file"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
