\documentclass[a4paper,10pt]{article}
\usepackage[utf8]{inputenc}
\usepackage[left=2.54cm, top=3.54cm, right=2.54cm, bottom=2.54cm]{geometry}
\usepackage{amsmath}
\usepackage{graphicx}
\usepackage{subfig}
\usepackage{listings}
\usepackage{hyperref}

%opening
\title{Methods used}
\author{Bram}

\begin{document}

\section{Cohort bias detection}
%
Before we perform cohort bias removal we seek to quantify the presence of such bias.
We have two general approaches: distribution based and pairwise similarity.

Distribution based: 
\begin{itemize}
 \item Wasserstein metrics
 \item Unsupervised non-parametric statistical significance test: Mann-Whitney U, Kolmogorov-Smirnof
 \item Supervised non-parametric statistical significance test: FDR-ANOVA
\end{itemize}
%
Two options for the application:
1. compare cohorts per feature (or reduced dimension) (columnwise)
2. compare cohorts per patients over the features (or reduced dimensions) (rowwise)

Pairwise similarity:
\begin{itemize}
 \item Kullback-Leibler divergence
 \item Distance metrics/correlation
\end{itemize}
%
Here, in general we only have on option for the application which is to compare the cohorts per inter-cohort patient-pair.\\\\
When comparing cohorts we have choose to compare each cohort with eachother, or we can compare each cohort with the overall distribution (minus that specific cohort).
%
Classification based:
\begin{itemize}
 \item separation of biological classes by batch identities
\end{itemize}
%
Variation based:
\begin{itemize}
 \item relations between in-group variance, out-group variance and between-group variances,
 see \href{https://www.biorxiv.org/content/biorxiv/early/2014/12/04/012203.full.pdf}{Hicks}
 \item ANOVA and Kruskall-Wallis to check for significantly different distributions.
\end{itemize}

Is probewise normalisation even appropriate given the large variation of biological groups between the cohorts?
Only apply probewise normalisation to sample groups with little variation (if at all)!

We can use R-quantro to verify..

%
\section{Cohort bias removal}
%
The following bias removal methods are applied
%
\begin{itemize}
\item RNA expression data: L/S adjustment \& cohort based QN
\item Methylation data: 1. cohort correction using ComBat \& cohort based QN. 2. SmoothedQN (color), 4. SubsetQN (type)/SubsetQN (islands)
\end{itemize}

We apply the cohort bias removal to the measurement cohorts. These cohorts indicate measurement batches and the cohort bias removal
reduces any bias that is seemingly related to the cohorts. Arguably we have to apply the bias removal, per cohort, per phenotypical cluster, otherwise
the applicability of the cohort bias removal hinges on the degree of stratification of the phenotypes. This is however prohibited by the sparsity of the data.
The ComBat method uses a combination of L/S normalisation/scaling and empirical Bayes to assess the bias that is introduced by
the cohort. As a reference we apply L/S, and cohort-wise QN.  \\ 
%
We use the same cohort-bias correction for both the RNA expression data and the methylation data. \\ \\
%
Results are evaluated using:
\begin{itemize}
\item distribution of the log10 of the p-values (K-S, each cohort compared to the bulk), for the FDR we use the current cohort versus the rest as the label
\item distribution of median deviation
\item distribution of mean, max, min 
\item distribution of correlation values between PCA1, PCA2, PCA3
\item plots of (PCA1, PCA2, PCA3), colored by cohort and by target.
\item plots of (UMAP1, UMAP2, UMAP3), colored by cohort and by target.
\item clustering of (sample, sample) similarity (HDBSCAN, AP, MC)
\item differential expression
\end{itemize}
%

The basic observation we should be able to make is the following:
prior to cohort-bias correction the cohort-based clusters should be distinctly seperated, 
and the target based clusters should be distinctly seperated as well. After the CBC the cohort-based clusters
should be significantly more similar. 

For the patient-based clustering we should see an increasing seperation of the different patient groups after the CBC based on the different target values.



\subsection{Batch wise normalisation}
%
Location and scale adjustment (L/S):
\begin{equation}
\mbox{Standard}\quad \mathbf{x}^*_k= \frac{\mathbf{x}_k-\overline{\mathbf{x}}_k}{\sigma_k} + \overline{\mathbf{x}}_k,\quad \forall k\in \mathcal{C}
\end{equation}
%
In literate this approach might be referred to as \textit{standardisation}. \\ \\
%
%

ComBat, Bayesian based $\rightarrow$ use \href{http://www.bu.edu/jlab/wp-assets/ComBat/Abstract.html}{1ibrary}, part of Bioconductor's sva package. \\

%
Alternatively: Concordant bias detection, MANCIE, combining CNV data with expression data.
%https://www.nature.com/articles/ncomms11305

%
How are the targets distributed over the batches? How do the phenotypical covariants vary within the 
cohorts and between the cohorts? \\ \\
%
To get rid of bias introduced by demographic variations within the cohorts we ideally have a large
independent data set that relates genetic expression data to a wide range of demographic categories, such that research into
demographic dependency of genetic measurement data is structurally open sourced and applied as common bench marks, see e.g. 
Vi\~{n}uela et al\cite{Vinuela2018}.
%
\subsection{Measurement group bias correction}
%

Methods: 
QN  (R, (methy)lumi), 
SQN (subset quantile normalisation)(R, wateRmelon), 
SWAN (subset-quantile within array normalisation)(R, minfi),
BMIQ (beta-mixture quantile normalisation)(R, wateRmelon)
Smoothed-QN
QN followed by BMIQ
%
BEclear, part of Bioconductor's BEclear package. \\
%
Functional normalisation, part of Bioconductor's minfi package. \\

peak-based correction (PBC), implemented R (wateRmelon/ima/nimbl). \\ 

From Wang et al. \cite{Wang2015}: Quantile normalisation replaces the signal intensity of a probe with the mean intensity of the probes that have the same rank from all studied arrays, and thus makes the distribution of probe intensities from each array the same. We are explicitly interested in variance between the target groups, hence we are fine with probewise bias
as long as it is roughly stratified over the target groups. As an alternative to probewise normalisation it is wiser
to simply split the dataset in seperate datasets per probewise group. \\ \\
%

\section{Methylation plus RNA expression}
%
We have about $60.000$ RNA expression values, and about $400.000$ methylation values, per sample.
To do a full correlation scan of all combinations we need to perform $60.000\times 400.000 \times 1000$ computations, 
or more specifically, we need to perform $60.000\times 400.000=24\,10^9$ in-products on vectors with length $\propto 1000$. \\ \\
%
To make this tractable we can be selective in the gene's by considering the target variable at hand (say the type 
of cancer) and only select the gene's or probe values that seperate the target variables the best based 
on some non-parametric distribution comparison such as Kolmogorov-Smirnof or Mann-Whitney U, or we apply 
a dimension reduction on both data sets and only directly compare  the top components per datasets.

The caveat with all these approaches is the bias we introduce by considering only the individually strong components per data set, 
The only immediate approach at hand to find the strong combinations of components is a brute-force approach?
Another approach we might try is relatively straightforward: we simply append the components to eachother and apply a
dimension reduction based on the variance (PCA) or the separation (LDA) after we can reconstruct what
components co-occur in the reduced dimensions. \\ \\
%
$(1000, 60.000),(1000, 400.000) \rightarrow (1000, 460.000)$
%
\bibliographystyle{plain}
\bibliography{methods}
\end{document}
